

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Genome-Wide Imputation Pipeline &mdash; genipe 1.2.0 documentation</title>
    
    <link rel="stylesheet" href="_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.2.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/bizstyle.js"></script>
    <link rel="top" title="genipe 1.2.0 documentation" href="#" />
    <link rel="next" title="Installation" href="installation.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">genipe 1.2.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Genome-Wide Imputation Pipeline</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#about">About</a></li>
</ul>
</li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href="installation.html"
                        title="next chapter">Installation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="genome-wide-imputation-pipeline">
<h1>Genome-Wide Imputation Pipeline<a class="headerlink" href="#genome-wide-imputation-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> (GENome-wide Imputation PipelinE) module provides an easy
an efficient way of performing genome-wide imputation analysis using the three
commonly used softwares <a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/">PLINK</a>,
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html">SHAPEIT</a> and
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html">IMPUTE2</a>.</p>
<p>A quality metrics report is automatically generated at the end of the
imputation process to easily assess the quality of the analysis. The report is
compiled into a PDF. For information on how to compile the report, refer to the
<a class="reference internal" href="tutorials/tutorial_genipe.html#genipe-tut-compile-report"><span>Compiling the report</span></a> section in the main <a class="reference internal" href="tutorials/tutorial_genipe.html#genipe-tut-page"><span>Pipeline Tutorial</span></a>.</p>
<p>Finally, it also provides a useful standalone tool to perform statistical
analysis on imputed (dosage) data (such as linear, logistic or survival
regressions, or <a class="reference external" href="http://www.hsph.harvard.edu/skat/">SKAT</a> analysis of rare
variants).</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#quick-navigation">Quick navigation</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#virtual-environment">Virtual environment</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#testing-the-installation">Testing the installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#updating-the-package">Updating the package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Short Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorials.html#pipeline-tutorial">Pipeline tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials.html#survival-tutorial">Survival tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials.html#skat-tutorial">SKAT tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials.html#linear-tutorial">Linear tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials.html#logistic-tutorial">Logistic tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials.html#linear-mixed-effects-tutorial">Linear mixed effects tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials.html#site-extraction-tutorial">Site extraction tutorial</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="input_files.html">Input Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="output_files.html">Output Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="implementation.html">Implementation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="module_content/modules.html">genipe</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="usage">
<span id="genipe-usage"></span><h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console"><div class="highlight"><pre><span class="gp">$</span> genipe-launcher --help
<span class="go">usage: genipe-launcher [-h] [-v] [--debug] [--thread THREAD] --bfile PREFIX</span>
<span class="go">                       [--reference FILE] [--output-dir DIR] [--bgzip]</span>
<span class="go">                       [--use-drmaa] [--drmaa-config FILE] [--preamble FILE]</span>
<span class="go">                       [--shapeit-bin BINARY] [--shapeit-thread INT]</span>
<span class="go">                       [--plink-bin BINARY] [--impute2-bin BINARY]</span>
<span class="go">                       [--segment-length BP] --hap-template TEMPLATE</span>
<span class="go">                       --legend-template TEMPLATE --map-template TEMPLATE</span>
<span class="go">                       --sample-file FILE [--filtering-rules RULE [RULE ...]]</span>
<span class="go">                       [--probability FLOAT] [--completion FLOAT]</span>
<span class="go">                       [--info FLOAT] [--report-number NB]</span>
<span class="go">                       [--report-title TITLE] [--report-author AUTHOR]</span>

<span class="go">Execute the genome-wide imputation pipeline. This script is part of the</span>
<span class="go">&#39;genipe&#39; package, version 1.2.0.</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>
<span class="go">  --debug               set the logging level to debug</span>
<span class="go">  --thread THREAD       number of threads [1]</span>

<span class="go">Input Options:</span>
<span class="go">  --bfile PREFIX        The prefix of the binary pedfiles (input data).</span>
<span class="go">  --reference FILE      The human reference to perform an initial strand check</span>
<span class="go">                        (useful for genotyped markers not in the IMPUTE2</span>
<span class="go">                        reference files) (optional).</span>

<span class="go">Output Options:</span>
<span class="go">  --output-dir DIR      The name of the output directory. [genipe]</span>
<span class="go">  --bgzip               Use bgzip to compress the impute2 files.</span>

<span class="go">HPC Options:</span>
<span class="go">  --use-drmaa           Launch tasks using DRMAA.</span>
<span class="go">  --drmaa-config FILE   The configuration file for tasks (use this option when</span>
<span class="go">                        launching tasks using DRMAA). This file should</span>
<span class="go">                        describe the walltime and the number of</span>
<span class="go">                        nodes/processors to use for each task.</span>
<span class="go">  --preamble FILE       This option should be used when using DRMAA on a HPC</span>
<span class="go">                        to load required module and set environment variables.</span>
<span class="go">                        The content of the file will be added between the</span>
<span class="go">                        &#39;shebang&#39; line and the tool command.</span>

<span class="go">SHAPEIT Options:</span>
<span class="go">  --shapeit-bin BINARY  The SHAPEIT binary if it&#39;s not in the path.</span>
<span class="go">  --shapeit-thread INT  The number of thread for phasing. [1]</span>

<span class="go">Plink Options:</span>
<span class="go">  --plink-bin BINARY    The Plink binary if it&#39;s not in the path.</span>

<span class="go">IMPUTE2 Options:</span>
<span class="go">  --impute2-bin BINARY  The IMPUTE2 binary if it&#39;s not in the path.</span>
<span class="go">  --segment-length BP   The length of a single segment for imputation. [5e+06]</span>
<span class="go">  --hap-template TEMPLATE</span>
<span class="go">                        The template for IMPUTE2&#39;s haplotype files (replace</span>
<span class="go">                        the chromosome number by &#39;{chrom}&#39;, e.g.</span>
<span class="go">                        &#39;1000GP_Phase3_chr{chrom}.hap.gz&#39;).</span>
<span class="go">  --legend-template TEMPLATE</span>
<span class="go">                        The template for IMPUTE2&#39;s legend files (replace the</span>
<span class="go">                        chromosome number by &#39;{chrom}&#39;, e.g.</span>
<span class="go">                        &#39;1000GP_Phase3_chr{chrom}.legend.gz&#39;).</span>
<span class="go">  --map-template TEMPLATE</span>
<span class="go">                        The template for IMPUTE2&#39;s map files (replace the</span>
<span class="go">                        chromosome number by &#39;{chrom}&#39;, e.g.</span>
<span class="go">                        &#39;genetic_map_chr{chrom}_combined_b37.txt&#39;).</span>
<span class="go">  --sample-file FILE    The name of IMPUTE2&#39;s sample file.</span>
<span class="go">  --filtering-rules RULE [RULE ...]</span>
<span class="go">                        IMPUTE2 filtering rules (optional).</span>

<span class="go">IMPUTE2 Merger Options:</span>
<span class="go">  --probability FLOAT   The probability threshold for no calls. [&lt;0.9]</span>
<span class="go">  --completion FLOAT    The completion rate threshold for site exclusion.</span>
<span class="go">                        [&lt;0.98]</span>
<span class="go">  --info FLOAT          The measure of the observed statistical information</span>
<span class="go">                        associated with the allele frequency estimate</span>
<span class="go">                        threshold for site exclusion. [&lt;0.00]</span>

<span class="go">Automatic Report Options:</span>
<span class="go">  --report-number NB    The report number. [genipe automatic report]</span>
<span class="go">  --report-title TITLE  The report title. [genipe: Automatic genome-wide</span>
<span class="go">                        imputation]</span>
<span class="go">  --report-author AUTHOR</span>
<span class="go">                        The report author. [Automatically generated by genipe]</span>
</pre></div>
</div>
</div>
<div class="section" id="about">
<h2>About<a class="headerlink" href="#about" title="Permalink to this headline">¶</a></h2>
<p>This project was initiated at the
<a class="reference external" href="http://www.pharmacogenomics.ca">Beaulieu-Saucier Pharmacogenomics Centre</a> of
the <a class="reference external" href="https://www.icm-mhi.org">Montreal Heart Institute</a>. The aim was to speed
up (and automatize) the imputation process for the whole genome.</p>
<a class="reference internal image-reference" href="_images/logo_pgx.png"><img alt="_images/logo_pgx.png" class="align-center" src="_images/logo_pgx.png" style="width: 350px;" /></a>
<a class="reference internal image-reference" href="_images/logo_statgen.png"><img alt="_images/logo_statgen.png" class="align-center" src="_images/logo_statgen.png" style="width: 200px;" /></a>
<a class="reference internal image-reference" href="_images/logo_icm.jpg"><img alt="_images/logo_icm.jpg" class="align-center" src="_images/logo_icm.jpg" style="width: 350px;" /></a>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">genipe 1.2.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>