

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>gwip.pipeline &mdash; gwip 0.1 documentation</title>
    
    <link rel="stylesheet" href="../../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../_static/bizstyle.js"></script>
    <link rel="top" title="gwip 0.1 documentation" href="../../index.html" />
    <link rel="up" title="gwip" href="../gwip.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">gwip 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../gwip.html" accesskey="U">gwip</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for gwip.pipeline</h1><div class="highlight"><pre>
<span class="c"># This file is part of gwip.</span>
<span class="c">#</span>
<span class="c"># This work is licensed under the Creative Commons Attribution-NonCommercial</span>
<span class="c"># 4.0 International License. To view a copy of this license, visit</span>
<span class="c"># http://creativecommons.org/licenses/by-nc/4.0/ or send a letter to Creative</span>
<span class="c"># Commons, PO Box 1866, Mountain View, CA 94042, USA.</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>
<span class="kn">from</span> <span class="nn">math</span> <span class="kn">import</span> <span class="n">floor</span>
<span class="kn">from</span> <span class="nn">shutil</span> <span class="kn">import</span> <span class="n">which</span><span class="p">,</span> <span class="n">copyfile</span>
<span class="kn">from</span> <span class="nn">urllib.request</span> <span class="kn">import</span> <span class="n">urlopen</span>
<span class="kn">from</span> <span class="nn">subprocess</span> <span class="kn">import</span> <span class="n">Popen</span><span class="p">,</span> <span class="n">PIPE</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">.db</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">__version__</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">chromosomes</span>
<span class="kn">from</span> <span class="nn">.task</span> <span class="kn">import</span> <span class="n">launcher</span>
<span class="kn">from</span> <span class="nn">.error</span> <span class="kn">import</span> <span class="n">ProgramError</span>
<span class="kn">from</span> <span class="nn">.config</span> <span class="kn">import</span> <span class="n">parse_drmaa_config</span>
<span class="kn">from</span> <span class="nn">.reporting</span> <span class="kn">import</span> <span class="n">generate_report</span>


<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">matplotlib</span> <span class="kn">as</span> <span class="nn">mpl</span>
    <span class="n">mpl</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s">&quot;Agg&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
    <span class="kn">import</span> <span class="nn">matplotlib.patches</span> <span class="kn">as</span> <span class="nn">mpatches</span>
    <span class="n">HAS_MATPLOTLIB</span> <span class="o">=</span> <span class="bp">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">HAS_MATPLOTLIB</span> <span class="o">=</span> <span class="bp">False</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">pyfaidx</span> <span class="kn">import</span> <span class="n">Fasta</span>
    <span class="n">HAS_PYFAIDX</span> <span class="o">=</span> <span class="bp">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">HAS_PYFAIDX</span> <span class="o">=</span> <span class="bp">False</span>


<span class="n">__author__</span> <span class="o">=</span> <span class="s">&quot;Louis-Philippe Lemieux Perreault&quot;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s">&quot;Copyright 2014, Beaulieu-Saucier Pharmacogenomics Centre&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s">&quot;Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)&quot;</span>


<span class="n">_complement</span> <span class="o">=</span> <span class="p">{</span><span class="s">&quot;A&quot;</span><span class="p">:</span> <span class="s">&quot;T&quot;</span><span class="p">,</span> <span class="s">&quot;T&quot;</span><span class="p">:</span> <span class="s">&quot;A&quot;</span><span class="p">,</span> <span class="s">&quot;C&quot;</span><span class="p">:</span> <span class="s">&quot;G&quot;</span><span class="p">,</span> <span class="s">&quot;G&quot;</span><span class="p">:</span> <span class="s">&quot;C&quot;</span><span class="p">}</span>


<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.main">[docs]</a>    <span class="sd">&quot;&quot;&quot;The main function of the genome-wide imputation pipeline.</span>

<span class="sd">    These are the pipeline steps:</span>

<span class="sd">    1.  Find markers with strand issues and flip them (Plink)</span>
<span class="sd">    2.  Exclude markers that have ambiguous alleles [A/T or C/G] (Plink)</span>
<span class="sd">    3.  Exclude duplicated markers (only keep one) (Plink)</span>
<span class="sd">    4.  Split the dataset by chromosome (Plink)</span>
<span class="sd">    5.  Find markers that need to be flipped (strand problem) (SHAPEIT)</span>
<span class="sd">    6.  Flip those markers (Plink)</span>
<span class="sd">    7.  Find markers with strand problem (SHAPEIT)</span>
<span class="sd">    8.  Exclude markers with strand problem (Plink)</span>
<span class="sd">    9.  Phase using SHAPEIT</span>
<span class="sd">    10. Impute using IMPUTE2</span>
<span class="sd">    11. Merge IMPUTE2 files</span>
<span class="sd">    12. If asked by the user, compress the final IMPUTE2 files (bgzip)</span>

<span class="sd">    At the end of the pipeline, a report (LaTeX) is automatically generated. It</span>
<span class="sd">    includes different quality statistics and imputation metrics.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Creating the option parser</span>
    <span class="n">desc</span> <span class="o">=</span> <span class="p">(</span><span class="s">&quot;Execute the genome-wide imputation pipeline. This script is part &quot;</span>
            <span class="s">&quot;of the &#39;gwip&#39; package, version {}.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>

    <span class="c"># We run the script</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c"># Parsing the options</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">)</span>

        <span class="c"># Creating the output directory if it doesn&#39;t exist</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">)</span>

        <span class="c"># Adding the logging capability</span>
        <span class="n">log_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;gwip.log&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
            <span class="n">format</span><span class="o">=</span><span class="s">&quot;[</span><span class="si">%(asctime)s</span><span class="s"> </span><span class="si">%(levelname)s</span><span class="s">] </span><span class="si">%(message)s</span><span class="s">&quot;</span><span class="p">,</span>
            <span class="n">datefmt</span><span class="o">=</span><span class="s">&quot;%Y-%m-</span><span class="si">%d</span><span class="s"> %H:%M:%S&quot;</span><span class="p">,</span>
            <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">debug</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
            <span class="n">handlers</span><span class="o">=</span><span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(),</span>
                      <span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s">&quot;w&quot;</span><span class="p">)]</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Logging everything into &#39;{}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">log_file</span><span class="p">))</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Program arguments: &#39;{}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])))</span>

        <span class="c"># Checking the options</span>
        <span class="n">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

        <span class="c"># Getting the task options</span>
        <span class="n">args</span><span class="o">.</span><span class="n">task_options</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Task will be launched with DRMAA&quot;</span><span class="p">)</span>
            <span class="n">args</span><span class="o">.</span><span class="n">task_options</span> <span class="o">=</span> <span class="n">parse_drmaa_config</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">drmaa_config</span><span class="p">)</span>
            <span class="n">args</span><span class="o">.</span><span class="n">preamble</span> <span class="o">=</span> <span class="n">read_preamble</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>

        <span class="c"># Creating the database</span>
        <span class="n">db_name</span> <span class="o">=</span> <span class="n">create_task_db</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">)</span>

        <span class="c"># Creating the output directories</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
            <span class="n">chr_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">chr_dir</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">chr_dir</span><span class="p">)</span>

        <span class="c"># Creating a data structure to gather run information</span>
        <span class="n">run_information</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c"># Getting shapeit version</span>
        <span class="n">run_information</span><span class="p">[</span><span class="s">&quot;shapeit_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_shapeit_version</span><span class="p">(</span>
            <span class="s">&quot;shapeit&quot;</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">shapeit_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">args</span><span class="o">.</span><span class="n">shapeit_bin</span>
        <span class="p">)</span>

        <span class="c"># Getting impute2 version</span>
        <span class="n">run_information</span><span class="p">[</span><span class="s">&quot;impute2_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_impute2_version</span><span class="p">(</span>
            <span class="s">&quot;impute2&quot;</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">impute2_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">args</span><span class="o">.</span><span class="n">impute2_bin</span>
        <span class="p">)</span>

        <span class="c"># Getting Plink version</span>
        <span class="n">run_information</span><span class="p">[</span><span class="s">&quot;plink_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_plink_version</span><span class="p">(</span>
            <span class="s">&quot;plink&quot;</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">plink_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">args</span><span class="o">.</span><span class="n">plink_bin</span>
        <span class="p">)</span>

        <span class="c"># Excluding markers prior to phasing (ambiguous markers [A/T and [G/C]</span>
        <span class="c"># and duplicated markers and finds markers to flip if there is a</span>
        <span class="c"># reference genome available</span>
        <span class="n">numbers</span> <span class="o">=</span> <span class="n">exclude_markers_before_phasing</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bfile</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">numbers</span><span class="p">)</span>

        <span class="c"># Computing the marker missing rate</span>
        <span class="n">missing_rate</span> <span class="o">=</span> <span class="n">compute_marker_missing_rate</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bfile</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="c"># Checking the strand</span>
        <span class="n">numbers</span> <span class="o">=</span> <span class="n">check_strand</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">),</span>
            <span class="s">&quot;_1&quot;</span><span class="p">,</span>
            <span class="n">db_name</span><span class="p">,</span>
            <span class="n">args</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">numbers</span><span class="p">)</span>

        <span class="c"># Flipping the markers</span>
        <span class="n">flip_markers</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">),</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}.to_flip&quot;</span><span class="p">),</span>
            <span class="n">db_name</span><span class="p">,</span>
            <span class="n">args</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c"># Checking the strand</span>
        <span class="n">numbers</span> <span class="o">=</span> <span class="n">check_strand</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}.flipped&quot;</span><span class="p">),</span>
            <span class="s">&quot;_2&quot;</span><span class="p">,</span>
            <span class="n">db_name</span><span class="p">,</span>
            <span class="n">args</span><span class="p">,</span>
            <span class="n">exclude</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">numbers</span><span class="p">)</span>

        <span class="c"># The final marker exclusion</span>
        <span class="n">numbers</span> <span class="o">=</span> <span class="n">final_exclusion</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}.flipped&quot;</span><span class="p">),</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}.to_exclude&quot;</span><span class="p">),</span>
            <span class="n">db_name</span><span class="p">,</span>
            <span class="n">args</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">numbers</span><span class="p">)</span>

        <span class="c"># Phasing the data</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="n">phase_markers</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}.final&quot;</span><span class="p">),</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                         <span class="s">&quot;chr{chrom}.final.phased&quot;</span><span class="p">),</span>
            <span class="n">db_name</span><span class="p">,</span>
            <span class="n">args</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c"># Gathering the chromosome length from Ensembl REST API</span>
        <span class="n">chromosome_length</span> <span class="o">=</span> <span class="n">get_chromosome_length</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">)</span>

        <span class="c"># Performs the imputation</span>
        <span class="n">impute_markers</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                                    <span class="s">&quot;chr{chrom}.final.phased.haps&quot;</span><span class="p">),</span>
                       <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                                    <span class="s">&quot;chr{chrom}.{start}_{end}.impute2&quot;</span><span class="p">),</span>
                       <span class="n">chromosome_length</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="c"># Getting the weighed average for cross-validation</span>
        <span class="n">numbers</span> <span class="o">=</span> <span class="n">get_cross_validation_results</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                         <span class="s">&quot;chr{chrom}.*.impute2_summary&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">numbers</span><span class="p">)</span>

        <span class="c"># Merging the impute2 files</span>
        <span class="n">merge_impute2_files</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                                         <span class="s">&quot;chr{chrom}.*.impute2&quot;</span><span class="p">),</span>
                            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                                         <span class="s">&quot;final_impute2&quot;</span><span class="p">,</span>
                                         <span class="s">&quot;chr{chrom}.imputed&quot;</span><span class="p">),</span>
                            <span class="n">args</span><span class="o">.</span><span class="n">probability</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">completion</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="c"># If required, zipping the impute2 files</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">bgzip</span><span class="p">:</span>
            <span class="n">compress_impute2_files</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                                                <span class="s">&quot;final_impute2&quot;</span><span class="p">,</span>
                                                <span class="s">&quot;chr{chrom}.imputed.impute2&quot;</span><span class="p">),</span>
                                   <span class="n">db_name</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="c"># Gathering the imputation statistics</span>
        <span class="n">numbers</span> <span class="o">=</span> <span class="n">gather_imputation_stats</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">probability</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">completion</span><span class="p">,</span>
                                          <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span> <span class="n">missing_rate</span><span class="p">,</span>
                                          <span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">numbers</span><span class="p">)</span>

        <span class="c"># Gathering the MAF statistics</span>
        <span class="n">numbers</span> <span class="o">=</span> <span class="n">gather_maf_stats</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">numbers</span><span class="p">)</span>

        <span class="c"># Checking that the number of good sites equals the number of sites</span>
        <span class="c"># with MAF (for internal validation)</span>
        <span class="n">nb_good_sites</span> <span class="o">=</span> <span class="n">run_information</span><span class="p">[</span><span class="s">&quot;nb_good_sites&quot;</span><span class="p">]</span>
        <span class="n">nb_sites_with_maf</span> <span class="o">=</span> <span class="n">run_information</span><span class="p">[</span><span class="s">&quot;nb_marker_with_maf&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">nb_good_sites</span> <span class="o">!=</span> <span class="n">nb_sites_with_maf</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;All {} (good) imputed sites should have a MAF, &quot;</span>
                            <span class="s">&quot;but only {} with MAF&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_good_sites</span><span class="p">,</span>
                                                          <span class="n">nb_sites_with_maf</span><span class="p">))</span>

        <span class="c"># Gathering the execution time</span>
        <span class="n">exec_time</span> <span class="o">=</span> <span class="n">gather_execution_time</span><span class="p">(</span><span class="n">db_name</span><span class="p">)</span>
        <span class="n">run_information</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">exec_time</span><span class="p">)</span>

        <span class="c"># Creating the output directory for the report (if it doesn&#39;t exits)</span>
        <span class="n">report_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;report&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">report_dir</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">report_dir</span><span class="p">)</span>

        <span class="c"># Generating the report</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Generating automatic report&quot;</span><span class="p">)</span>
        <span class="n">generate_report</span><span class="p">(</span><span class="n">report_dir</span><span class="p">,</span> <span class="n">args</span><span class="p">,</span> <span class="n">run_information</span><span class="p">)</span>

    <span class="c"># Catching the Ctrl^C</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="c"># Catching the ProgramError</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">raise</span>


<span class="k">def</span> <span class="nf">phase_markers</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">o_prefix</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span></div>
<div class="viewcode-block" id="phase_markers"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.phase_markers">[docs]</a>    <span class="sd">&quot;&quot;&quot;Phase markers using shapeit.</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix (str): the prefix template of the input files</span>
<span class="sd">        o_prefix (str): the prefix template of the output files</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: the list of all samples that were phased</span>

<span class="sd">    A template contains the string ``{chrom}``, which will be replaced by the</span>
<span class="sd">    chromosome number (e.g. ``gwip/chr{chrom}/chr{chrom}.final`` will be</span>
<span class="sd">    replaced by ``gwip/chr1/chr1.final``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;shapeit&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">shapeit_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">options</span><span class="o">.</span><span class="n">shapeit_bin</span><span class="p">,</span>
        <span class="s">&quot;-phase&quot;</span><span class="p">,</span>
        <span class="s">&quot;--thread&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">shapeit_thread</span><span class="p">),</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The current output prefix</span>
        <span class="n">c_prefix</span> <span class="o">=</span> <span class="n">o_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s">&quot;-B&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;-M&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">map_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;-O&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">,</span>
            <span class="s">&quot;-L&quot;</span><span class="p">,</span> <span class="n">c_prefix</span> <span class="o">+</span> <span class="s">&quot;.log&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;shapeit_phase_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;SHAPEIT phase chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
            <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
            <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="n">ext</span> <span class="k">for</span> <span class="n">ext</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;.haps&quot;</span><span class="p">,</span> <span class="s">&quot;.sample&quot;</span><span class="p">)],</span>
        <span class="p">})</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Phasing markers&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done phasing markers&quot;</span><span class="p">)</span>

    <span class="c"># Checking that all the sample files are the same</span>
    <span class="n">compare_with</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">o_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span> <span class="o">+</span> <span class="s">&quot;.sample&quot;</span>
        <span class="n">compare_to</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">compare_to</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">compare_with</span> <span class="o">=</span> <span class="n">compare_to</span>

        <span class="k">if</span> <span class="n">compare_with</span> <span class="o">!=</span> <span class="n">compare_to</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;phased sample files are different...&quot;</span><span class="p">)</span>

    <span class="c"># Returning the samples</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">compare_with</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()[</span><span class="mi">2</span><span class="p">:]]</span>


<span class="k">def</span> <span class="nf">impute_markers</span><span class="p">(</span><span class="n">phased_haplotypes</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">,</span> <span class="n">chrom_length</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span></div>
<div class="viewcode-block" id="impute_markers"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.impute_markers">[docs]</a>                   <span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Imputes the markers using IMPUTE2.</span>

<span class="sd">    Args:</span>
<span class="sd">        phased_haplotypes (str): the template for the haplotype files</span>
<span class="sd">        out_prefix (str): the prefix template of the output files</span>
<span class="sd">        chrom_length (dict): the length of each chromosome</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    A template contains the string ``{chrom}``, which will be replaced by the</span>
<span class="sd">    chromosome number (e.g. ``gwip/chr{chrom}/chr{chrom}.final`` will be</span>
<span class="sd">    replaced by ``gwip/chr1/chr1.final``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Are we skipping DRMAA options?</span>
    <span class="n">skip_drmaa_config</span> <span class="o">=</span> <span class="bp">False</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">task_options</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">skip_drmaa_config</span> <span class="o">=</span> <span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="s">&quot;skip_drmaa_config&quot;</span><span class="p">,</span>
            <span class="bp">False</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;impute2&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">impute2_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">options</span><span class="o">.</span><span class="n">impute2_bin</span><span class="p">,</span>
        <span class="s">&quot;-use_prephased_g&quot;</span><span class="p">,</span>
        <span class="s">&quot;-Ne&quot;</span><span class="p">,</span> <span class="s">&quot;20000&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c"># Are there any filtering rules?</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">filtering_rules</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">base_command</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&quot;-filt_rules_l&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">rule</span> <span class="ow">in</span> <span class="n">options</span><span class="o">.</span><span class="n">filtering_rules</span><span class="p">:</span>
            <span class="n">base_command</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rule</span><span class="p">)</span>

    <span class="c"># Each chromosome have multiple segments</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="k">assert</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="ow">in</span> <span class="n">chrom_length</span>

        <span class="n">length</span> <span class="o">=</span> <span class="n">chrom_length</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">length</span><span class="p">:</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">floor</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">segment_length</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>

            <span class="c"># The current output prefix</span>
            <span class="n">c_prefix</span> <span class="o">=</span> <span class="n">out_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">)</span>

            <span class="c"># The task ID</span>
            <span class="n">task_id</span> <span class="o">=</span> <span class="s">&quot;impute2_chr{}_{}_{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>

            <span class="c"># The command for this segment</span>
            <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
                <span class="s">&quot;-known_haps_g&quot;</span><span class="p">,</span> <span class="n">phased_haplotypes</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
                <span class="s">&quot;-h&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">hap_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
                <span class="s">&quot;-l&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">legend_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
                <span class="s">&quot;-m&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">map_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
                <span class="s">&quot;-int&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">end</span><span class="p">),</span>
                <span class="s">&quot;-o&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">,</span>
            <span class="p">]</span>
            <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="n">task_id</span><span class="p">,</span>
                <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;IMPUTE2 chr{} from {} to {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span>
                                                             <span class="n">end</span><span class="p">),</span>
                <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
                <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
                <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="s">&quot;_summary&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">],</span>
            <span class="p">})</span>

            <span class="c"># The new starting position</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">end</span> <span class="o">+</span> <span class="mi">1</span>

            <span class="c"># Adding the walltime for this particular task_id</span>
            <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">skip_drmaa_config</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">task_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">:</span>
                    <span class="c"># Sending the chromosome specific instead</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">[</span><span class="s">&quot;impute2_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
                    <span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">[</span><span class="n">task_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

    <span class="c"># Executing the commands</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Imputing markers&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done imputing markers&quot;</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">merge_impute2_files</span><span class="p">(</span><span class="n">in_glob</span><span class="p">,</span> <span class="n">o_prefix</span><span class="p">,</span> <span class="n">probability_t</span><span class="p">,</span> <span class="n">completion_t</span><span class="p">,</span></div>
<div class="viewcode-block" id="merge_impute2_files"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.merge_impute2_files">[docs]</a>                        <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Merges impute2 files.</span>

<span class="sd">    Args:</span>
<span class="sd">        in_glob (str): the template that will be used to find files with the</span>
<span class="sd">                       :py:mod:`glob` module</span>
<span class="sd">        o_prefix (str): the prefix template of the output files</span>
<span class="sd">        probability_t (float): the probability threshold to use</span>
<span class="sd">        completion_t (float): the completion threshold to use</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    A template contains the string ``{chrom}``, which will be replaced by the</span>
<span class="sd">    chromosome number (e.g. ``gwip/chr{chrom}/chr{chrom}.final`` will be</span>
<span class="sd">    replaced by ``gwip/chr1/chr1.final``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;impute2-merger&quot;</span><span class="p">,</span>
        <span class="s">&quot;--probability&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">probability_t</span><span class="p">),</span>
        <span class="s">&quot;--completion&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">completion_t</span><span class="p">),</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The current output prefix</span>
        <span class="n">c_prefix</span> <span class="o">=</span> <span class="n">o_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="c"># Checking that the output directory exists</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">c_prefix</span><span class="p">)):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">c_prefix</span><span class="p">))</span>

        <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s">&quot;--prefix&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">,</span>
            <span class="s">&quot;--chr&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;-i&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">filenames</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="n">in_glob</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)),</span> <span class="n">key</span><span class="o">=</span><span class="n">file_sorter</span><span class="p">)</span>
        <span class="n">remaining_command</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span>
        <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;merge_impute2_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;Merge imputed chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
            <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
            <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="n">ext</span> <span class="k">for</span> <span class="n">ext</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;.alleles&quot;</span><span class="p">,</span>
                                                   <span class="s">&quot;.completion_rates&quot;</span><span class="p">,</span>
                                                   <span class="s">&quot;.good_sites&quot;</span><span class="p">,</span>
                                                   <span class="s">&quot;.impute2&quot;</span><span class="p">,</span>
                                                   <span class="s">&quot;.imputed_sites&quot;</span><span class="p">,</span>
                                                   <span class="s">&quot;.map&quot;</span><span class="p">,</span>
                                                   <span class="s">&quot;.maf&quot;</span><span class="p">)],</span>
        <span class="p">})</span>

        <span class="c"># Getting the name of the sample file</span>
        <span class="n">sample_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">in_glob</span><span class="p">),</span>
            <span class="s">&quot;chr{chrom}.final.phased.sample&quot;</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="c"># Checking if the file exists</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">sample_file</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample_file</span><span class="p">))</span>

        <span class="c"># Copying the file</span>
        <span class="n">copyfile</span><span class="p">(</span><span class="n">sample_file</span><span class="p">,</span> <span class="n">c_prefix</span> <span class="o">+</span> <span class="s">&quot;.sample&quot;</span><span class="p">)</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Merging impute2 files&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done merging reports&quot;</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">compress_impute2_files</span><span class="p">(</span><span class="n">filename_template</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span></div>
<div class="viewcode-block" id="compress_impute2_files"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.compress_impute2_files">[docs]</a>    <span class="sd">&quot;&quot;&quot;Merges impute2 files.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename_template (str): the template for the final IMPUTE2 file</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    A template contains the string ``{chrom}``, which will be replaced by the</span>
<span class="sd">    chromosome number (e.g. ``gwip/chr{chrom}/chr{chrom}.final`` will be</span>
<span class="sd">    replaced by ``gwip/chr1/chr1.final``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;bgzip&quot;</span><span class="p">,</span> <span class="s">&quot;-f&quot;</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The current output prefix</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">filename_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">filename</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;bgzip_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;Compress chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
            <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
            <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">filename</span> <span class="o">+</span> <span class="s">&quot;.gz&quot;</span><span class="p">],</span>
        <span class="p">})</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Compressing impute2 files&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done compressing impute2 files&quot;</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">file_sorter</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span></div>
<div class="viewcode-block" id="file_sorter"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.file_sorter">[docs]</a>    <span class="sd">&quot;&quot;&quot;Helps in filename sorting.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename (str): the name of the file to compare while sorting</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: a tuple containing three elements: chromosome (int), start (int)</span>
<span class="sd">               and end (int)</span>

<span class="sd">    Using a regular expression, finds the chromosome along with the starting</span>
<span class="sd">    and ending position of an imputed segment. The file</span>
<span class="sd">    ``chr22.1_50000.impute2`` should return ``(22, 1, 50000)``.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&quot;chr(\d+)\.(\d+)_(\d+)\.impute2&quot;</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)),</span> <span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)),</span> <span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">)))</span>


<span class="k">def</span> <span class="nf">get_chromosome_length</span><span class="p">(</span><span class="n">out_dir</span><span class="p">):</span></div>
<div class="viewcode-block" id="get_chromosome_length"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.get_chromosome_length">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gets the chromosome length from Ensembl REST API.</span>

<span class="sd">    Args:</span>
<span class="sd">        out_dir (str): the output directory</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: the chromosome length (chr -&gt; length)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chrom_length</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chromosome_lengths.txt&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Gathering chromosome length (Ensembl, GRCh37)&quot;</span><span class="p">)</span>

        <span class="c"># The URL</span>
        <span class="n">url</span> <span class="o">=</span> <span class="p">(</span><span class="s">&quot;http://grch37.rest.ensembl.org/info/assembly/homo_sapiens&quot;</span>
               <span class="s">&quot;?content-type=application/json&quot;</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">urlopen</span><span class="p">(</span><span class="n">url</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>

        <span class="c"># Checking the build</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="p">[</span><span class="s">&quot;assembly_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;GRCh37&quot;</span><span class="p">)</span> <span class="ow">or</span> \
           <span class="n">result</span><span class="p">[</span><span class="s">&quot;default_coord_system_version&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&quot;GRCh37&quot;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: wrong &quot;</span>
                               <span class="s">&quot;build&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s">&quot;assembly_name&quot;</span><span class="p">]))</span>

        <span class="c"># Gathering the chromosome length</span>
        <span class="n">chrom_length</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">req_chrom</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">23</span><span class="p">)}</span> <span class="o">|</span> <span class="p">{</span><span class="s">&quot;X&quot;</span><span class="p">}</span>
        <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s">&quot;top_level_region&quot;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">region</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">req_chrom</span><span class="p">:</span>
                <span class="n">chrom_length</span><span class="p">[</span><span class="n">region</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">region</span><span class="p">[</span><span class="s">&quot;length&quot;</span><span class="p">]</span>

        <span class="c"># Saving to file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o_file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">chrom_length</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                <span class="k">print</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">chrom_length</span><span class="p">[</span><span class="n">chrom</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c"># Gathering from file</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Gathering chromosome length ({})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">chrom_length</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">chrom_length</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

    <span class="c"># Checking we have all the required data</span>
    <span class="n">required_chrom</span> <span class="o">=</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">}</span>
    <span class="k">if</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">chrom_length</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">required_chrom</span><span class="p">)</span> <span class="o">!=</span> <span class="n">required_chrom</span><span class="p">:</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="s">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">required_chrom</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">chrom_length</span><span class="p">)))</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;missing chromosomes: {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">missing</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">chrom_length</span>


<span class="k">def</span> <span class="nf">check_strand</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">id_suffix</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">,</span> <span class="n">exclude</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span></div>
<div class="viewcode-block" id="check_strand"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.check_strand">[docs]</a>    <span class="sd">&quot;&quot;&quot;Checks the strand using SHAPEIT2.</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix (str): the prefix template of the input files</span>
<span class="sd">        id_suffix (str): the suffix of the task</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>
<span class="sd">        exclude (bool): should markers be excluded or flipped (default is</span>
<span class="sd">                        flipped)</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: statistics about the task (number of markers that will be flipped</span>
<span class="sd">              or excluded)</span>

<span class="sd">    A template contains the string ``{chrom}``, which will be replaced by the</span>
<span class="sd">    chromosome number (e.g. ``gwip/chr{chrom}/chr{chrom}.final`` will be</span>
<span class="sd">    replaced by ``gwip/chr1/chr1.final``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Creating the command to launch</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;shapeit&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">shapeit_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">options</span><span class="o">.</span><span class="n">shapeit_bin</span><span class="p">,</span>
        <span class="s">&quot;-check&quot;</span>
    <span class="p">]</span>

    <span class="c"># The output suffix</span>
    <span class="n">suffix</span> <span class="o">=</span> <span class="s">&quot;alignments&quot;</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">exclude</span><span class="p">:</span>
        <span class="c"># This is for exclusion</span>
        <span class="n">suffix</span> <span class="o">=</span> <span class="s">&quot;to_exclude.alignments&quot;</span>

    <span class="c"># The output file prefix</span>
    <span class="n">o_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                            <span class="s">&quot;chr{chrom}.&quot;</span> <span class="o">+</span> <span class="n">suffix</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The current output prefix</span>
        <span class="n">c_prefix</span> <span class="o">=</span> <span class="n">o_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s">&quot;-B&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;-M&quot;</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">map_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;--input-ref&quot;</span><span class="p">,</span>
            <span class="n">options</span><span class="o">.</span><span class="n">hap_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="n">options</span><span class="o">.</span><span class="n">legend_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="n">options</span><span class="o">.</span><span class="n">sample_file</span><span class="p">,</span>
            <span class="s">&quot;--output-log&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;shapeit_check_chr{}{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">id_suffix</span><span class="p">),</span>
            <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;SHAPEIT check strand chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
            <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
            <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="s">&quot;.snp.strand&quot;</span><span class="p">,</span> <span class="p">],</span>
        <span class="p">})</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Checking strand of markers&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">check_rc</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                          <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done checking strand of markers&quot;</span><span class="p">)</span>

    <span class="c"># The output suffix</span>
    <span class="n">o_suffix</span> <span class="o">=</span> <span class="s">&quot;to_flip&quot;</span>
    <span class="n">what</span> <span class="o">=</span> <span class="s">&quot;flip&quot;</span>
    <span class="k">if</span> <span class="n">exclude</span><span class="p">:</span>
        <span class="n">o_suffix</span> <span class="o">=</span> <span class="s">&quot;to_exclude&quot;</span>
        <span class="n">what</span> <span class="o">=</span> <span class="s">&quot;exclude&quot;</span>

    <span class="c"># The name of the files</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="n">o_prefix</span> <span class="o">+</span> <span class="s">&quot;.snp.strand&quot;</span>
    <span class="n">o_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                              <span class="s">&quot;chr{chrom}.{o_suffix}&quot;</span><span class="p">)</span>

    <span class="c"># For each chromosome, we find markers to change strand</span>
    <span class="n">nb_total</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The SNP to print in the output file</span>
        <span class="n">to_write</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="n">chrom_filename</span> <span class="o">=</span> <span class="n">filename</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>
        <span class="n">chrom_o_filename</span> <span class="o">=</span> <span class="n">o_filename</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">o_suffix</span><span class="o">=</span><span class="n">o_suffix</span><span class="p">)</span>

        <span class="c"># Checking the input file exists</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">chrom_filename</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom_filename</span><span class="p">))</span>

        <span class="c"># Markers to flip</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">chrom_filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="c"># Reading the header</span>
            <span class="n">header</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\r\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="n">header</span> <span class="o">=</span> <span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">header</span><span class="p">)}</span>

            <span class="c"># Checking header</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;type&quot;</span><span class="p">,</span> <span class="s">&quot;main_id&quot;</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no column named &quot;</span>
                                       <span class="s">&quot;{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom_filename</span><span class="p">,</span> <span class="n">name</span><span class="p">))</span>

            <span class="c"># Reading the file</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\r\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="n">header</span><span class="p">[</span><span class="s">&quot;type&quot;</span><span class="p">]]</span> <span class="o">==</span> <span class="s">&quot;Strand&quot;</span><span class="p">:</span>
                    <span class="n">to_write</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">header</span><span class="p">[</span><span class="s">&quot;main_id&quot;</span><span class="p">]])</span>

        <span class="c"># The number of markers to flip</span>
        <span class="n">to_flip</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">to_write</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">chrom_o_filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o_file</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="o">*</span><span class="n">to_write</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>

        <span class="n">nb_total</span> <span class="o">+=</span> <span class="n">to_flip</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;chr{}: {:,d} markers to {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">to_flip</span><span class="p">,</span> <span class="n">what</span><span class="p">))</span>

    <span class="c"># Logging the last one</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;After strand check: {:,d} markers &quot;</span>
                 <span class="s">&quot;to {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_total</span><span class="p">,</span> <span class="n">what</span><span class="p">))</span>

    <span class="k">return</span> <span class="p">{</span><span class="s">&quot;nb_{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">what</span><span class="p">):</span> <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_total</span><span class="p">)}</span>


<span class="k">def</span> <span class="nf">flip_markers</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">to_flip</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span></div>
<div class="viewcode-block" id="flip_markers"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.flip_markers">[docs]</a>    <span class="sd">&quot;&quot;&quot;Flip markers.</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix (str): the prefix template of the input files</span>
<span class="sd">        to_flip (str): the name of the file containing markers to flip</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    A template contains the string ``{chrom}``, which will be replaced by the</span>
<span class="sd">    chromosome number (e.g. ``gwip/chr{chrom}/chr{chrom}.final`` will be</span>
<span class="sd">    replaced by ``gwip/chr1/chr1.final``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The commands to run</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;plink&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span><span class="p">,</span>
        <span class="s">&quot;--noweb&quot;</span><span class="p">,</span>
        <span class="s">&quot;--make-bed&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c"># The output prefix</span>
    <span class="n">o_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span>
                            <span class="s">&quot;chr{chrom}.flipped&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The current output prefix</span>
        <span class="n">c_prefix</span> <span class="o">=</span> <span class="n">o_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;--flip&quot;</span><span class="p">,</span> <span class="n">to_flip</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;--out&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;plink_flip_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;plink flip chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
            <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
            <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="n">ext</span> <span class="k">for</span> <span class="n">ext</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;.bed&quot;</span><span class="p">,</span> <span class="s">&quot;.bim&quot;</span><span class="p">,</span> <span class="s">&quot;.fam&quot;</span><span class="p">)],</span>
        <span class="p">})</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Flipping markers&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done flipping markers&quot;</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">final_exclusion</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span></div>
<div class="viewcode-block" id="final_exclusion"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.final_exclusion">[docs]</a>    <span class="sd">&quot;&quot;&quot;Flip markers.</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix (str): the prefix template of the input files</span>
<span class="sd">        to_exclude (str): the name of the file containing the markers to</span>
<span class="sd">                          exclude</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: the number of remaining markers for phasing</span>

<span class="sd">    A template contains the string ``{chrom}``, which will be replaced by the</span>
<span class="sd">    chromosome number (e.g. ``gwip/chr{chrom}/chr{chrom}.final`` will be</span>
<span class="sd">    replaced by ``gwip/chr1/chr1.final``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The commands to run</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;plink&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span><span class="p">,</span>
        <span class="s">&quot;--noweb&quot;</span><span class="p">,</span>
        <span class="s">&quot;--make-bed&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c"># The output prefix</span>
    <span class="n">o_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}.final&quot;</span><span class="p">)</span>

    <span class="c"># The output files (for statistics)</span>
    <span class="n">bims</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The current output prefix</span>
        <span class="n">c_prefix</span> <span class="o">=</span> <span class="n">o_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;--exclude&quot;</span><span class="p">,</span> <span class="n">to_exclude</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;--out&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;plink_final_exclude_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;plink final exclude chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
            <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
            <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="n">ext</span> <span class="k">for</span> <span class="n">ext</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;.bed&quot;</span><span class="p">,</span> <span class="s">&quot;.bim&quot;</span><span class="p">,</span> <span class="s">&quot;.fam&quot;</span><span class="p">)],</span>
        <span class="p">})</span>

        <span class="n">bims</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="s">&quot;.bim&quot;</span><span class="p">)</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Final marker exclusion&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done final marker exclusion&quot;</span><span class="p">)</span>

    <span class="n">nb_markers</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">bim</span> <span class="ow">in</span> <span class="n">bims</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">bim</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
                <span class="n">nb_markers</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="p">{</span><span class="s">&quot;nb_phasing_markers&quot;</span><span class="p">:</span> <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_markers</span><span class="p">)}</span>


<span class="k">def</span> <span class="nf">compute_marker_missing_rate</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span></div>
<div class="viewcode-block" id="compute_marker_missing_rate"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.compute_marker_missing_rate">[docs]</a>    <span class="sd">&quot;&quot;&quot;Compute (using Plink) marker missing rate.</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix (str): the prefix of the input file</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: the missing rate for each site (results from Plink)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The output prefix</span>
    <span class="n">o_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;missing&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">o_prefix</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">o_prefix</span><span class="p">)</span>
    <span class="n">o_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">o_prefix</span><span class="p">,</span> <span class="s">&quot;missing&quot;</span><span class="p">)</span>

    <span class="c"># The command to run</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;plink&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span><span class="p">,</span>
        <span class="s">&quot;--noweb&quot;</span><span class="p">,</span>
        <span class="s">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span>
        <span class="s">&quot;--missing&quot;</span><span class="p">,</span>
        <span class="s">&quot;--out&quot;</span><span class="p">,</span> <span class="n">o_prefix</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
        <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;plink_missing_rate&quot;</span><span class="p">,</span>
        <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;plink missing rate&quot;</span><span class="p">,</span>
        <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">command</span><span class="p">,</span>
        <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
        <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">o_prefix</span> <span class="o">+</span> <span class="n">ext</span> <span class="k">for</span> <span class="n">ext</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;.lmiss&quot;</span><span class="p">,</span> <span class="s">&quot;.imiss&quot;</span><span class="p">)],</span>
    <span class="p">})</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Computing missing rate&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done computing missing rate&quot;</span><span class="p">)</span>

    <span class="c"># Getting the missing rate</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Reading the missing rate&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">o_prefix</span> <span class="o">+</span> <span class="s">&quot;.lmiss&quot;</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">exclude_markers_before_phasing</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="n">db_name</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span></div>
<div class="viewcode-block" id="exclude_markers_before_phasing"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.exclude_markers_before_phasing">[docs]</a>    <span class="sd">&quot;&quot;&quot;Finds and excludes ambiguous markers (A/T and G/C) or duplicated.</span>

<span class="sd">    Args:</span>
<span class="sd">        prefix (str): the prefix of the input files</span>
<span class="sd">        db_name (str): the name of the DB saving tasks&#39; information</span>
<span class="sd">        options (argparse.Namespace): the pipeline options</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: information about the data set</span>

<span class="sd">    If required, the function uses :py:func:`is_reversed` to check if a marker</span>
<span class="sd">    is on the reverse strand and needs flipping.</span>

<span class="sd">    The information returned includes the initial number of markers and</span>
<span class="sd">    samples, the number of ambiguous, duplicated, special (not on autosomal</span>
<span class="sd">    chromosomes) markers, along with the number of markers to flip if the</span>
<span class="sd">    reference was checked.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The ambiguous genotypes</span>
    <span class="n">ambiguous_genotypes</span> <span class="o">=</span> <span class="p">{</span><span class="s">&quot;AT&quot;</span><span class="p">,</span> <span class="s">&quot;TA&quot;</span><span class="p">,</span> <span class="s">&quot;GC&quot;</span><span class="p">,</span> <span class="s">&quot;CG&quot;</span><span class="p">}</span>

    <span class="c"># The positions that have been written to file</span>
    <span class="n">kept_positions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">nb_ambiguous</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nb_kept</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nb_dup</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">to_flip</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="n">reference</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">chrom_encoding</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">reference</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">reference</span> <span class="o">=</span> <span class="n">Fasta</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">reference</span><span class="p">,</span> <span class="n">as_raw</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="n">chrom_encoding</span> <span class="o">=</span> <span class="n">get_chrom_encoding</span><span class="p">(</span><span class="n">reference</span><span class="p">)</span>

    <span class="c"># Logging</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Finding markers to exclude&quot;</span> <span class="o">+</span>
                 <span class="p">(</span><span class="s">&quot; and to flip&quot;</span> <span class="k">if</span> <span class="n">reference</span> <span class="k">else</span> <span class="s">&quot;&quot;</span><span class="p">))</span>

    <span class="c"># Counting the total number of markers and samples</span>
    <span class="n">nb_markers</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nb_samples</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nb_special_markers</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;.fam&quot;</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">nb_samples</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">o_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;markers_to_exclude.txt&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">o_filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o_file</span><span class="p">,</span> \
            <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;.bim&quot;</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">nb_markers</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\r\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>

            <span class="c"># The marker information</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">a1</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
            <span class="n">a2</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>

            <span class="n">is_special</span> <span class="o">=</span> <span class="bp">False</span>
            <span class="k">if</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="p">{</span><span class="s">&quot;23&quot;</span><span class="p">,</span> <span class="s">&quot;24&quot;</span><span class="p">,</span> <span class="s">&quot;25&quot;</span><span class="p">,</span> <span class="s">&quot;26&quot;</span><span class="p">}:</span>
                <span class="n">nb_special_markers</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">is_special</span> <span class="o">=</span> <span class="bp">True</span>

            <span class="c"># Checking the alleles</span>
            <span class="k">if</span> <span class="n">a1</span> <span class="o">+</span> <span class="n">a2</span> <span class="ow">in</span> <span class="n">ambiguous_genotypes</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">is_special</span><span class="p">:</span>
                    <span class="n">nb_ambiguous</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s">&quot;  - {}: {}: &quot;</span>
                              <span class="s">&quot;ambiguous&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">a1</span> <span class="o">+</span> <span class="n">a2</span><span class="p">))</span>
                <span class="k">continue</span>

            <span class="c"># Checking if we already have this marker</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">)</span> <span class="ow">in</span> <span class="n">kept_positions</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">is_special</span><span class="p">:</span>
                    <span class="n">nb_dup</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s">&quot;  - {}: duplicated&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">))</span>
                <span class="k">continue</span>

            <span class="c"># Checking strand if required</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_special</span> <span class="ow">and</span> <span class="p">(</span><span class="n">reference</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">is_reversed</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">),</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span>
                               <span class="n">chrom_encoding</span><span class="p">):</span>
                    <span class="n">to_flip</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>

            <span class="c"># We keep this marker</span>
            <span class="n">kept_positions</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_special</span><span class="p">:</span>
                <span class="n">nb_kept</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c"># Closing the reference</span>
    <span class="k">if</span> <span class="n">reference</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">reference</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c"># Logging</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} special markers&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_special_markers</span><span class="p">))</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} ambiguous markers removed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_ambiguous</span><span class="p">))</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} duplicated markers removed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_dup</span><span class="p">))</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} markers kept&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_kept</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">reference</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} markers to flip&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">to_flip</span><span class="p">)))</span>

    <span class="c"># Needs to flip?</span>
    <span class="n">o_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;markers_to_flip.txt&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">to_flip</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">o_filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o_file</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="o">*</span><span class="n">to_flip</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>

    <span class="c"># The commands to run</span>
    <span class="n">commands_info</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">base_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s">&quot;plink&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">options</span><span class="o">.</span><span class="n">plink_bin</span><span class="p">,</span>
        <span class="s">&quot;--noweb&quot;</span><span class="p">,</span>
        <span class="s">&quot;--exclude&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;markers_to_exclude.txt&quot;</span><span class="p">),</span>
        <span class="s">&quot;--make-bed&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">to_flip</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">base_command</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span>
            <span class="s">&quot;--flip&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;markers_to_flip.txt&quot;</span><span class="p">)</span>
        <span class="p">])</span>

    <span class="c"># The output prefix</span>
    <span class="n">o_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># The current output prefix</span>
        <span class="n">c_prefix</span> <span class="o">=</span> <span class="n">o_prefix</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>

        <span class="n">remaining_command</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s">&quot;--bfile&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span>
            <span class="s">&quot;--chr&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;--out&quot;</span><span class="p">,</span> <span class="n">c_prefix</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">commands_info</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s">&quot;task_id&quot;</span><span class="p">:</span> <span class="s">&quot;plink_exclude_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;plink exclude chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span>
            <span class="s">&quot;command&quot;</span><span class="p">:</span> <span class="n">base_command</span> <span class="o">+</span> <span class="n">remaining_command</span><span class="p">,</span>
            <span class="s">&quot;task_db&quot;</span><span class="p">:</span> <span class="n">db_name</span><span class="p">,</span>
            <span class="s">&quot;o_files&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">c_prefix</span> <span class="o">+</span> <span class="n">ext</span> <span class="k">for</span> <span class="n">ext</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;.bed&quot;</span><span class="p">,</span> <span class="s">&quot;.bim&quot;</span><span class="p">,</span> <span class="s">&quot;.fam&quot;</span><span class="p">)],</span>
        <span class="p">})</span>

    <span class="c"># Executing command</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Excluding and splitting markers&quot;</span><span class="p">)</span>
    <span class="n">launcher</span><span class="o">.</span><span class="n">launch_tasks</span><span class="p">(</span><span class="n">commands_info</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">thread</span><span class="p">,</span> <span class="n">hpc</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">,</span>
                          <span class="n">hpc_options</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">task_options</span><span class="p">,</span>
                          <span class="n">out_dir</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">preamble</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">preamble</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Done excluding and splitting markers&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s">&quot;initial_nb_markers&quot;</span><span class="p">:</span> <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_markers</span><span class="p">),</span>
        <span class="s">&quot;initial_nb_samples&quot;</span><span class="p">:</span> <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_samples</span><span class="p">),</span>
        <span class="s">&quot;nb_ambiguous&quot;</span><span class="p">:</span>       <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_ambiguous</span><span class="p">),</span>
        <span class="s">&quot;nb_duplicates&quot;</span><span class="p">:</span>      <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_dup</span><span class="p">),</span>
        <span class="s">&quot;nb_special_markers&quot;</span><span class="p">:</span> <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_special_markers</span><span class="p">),</span>
        <span class="s">&quot;nb_flip_reference&quot;</span><span class="p">:</span>  <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">to_flip</span><span class="p">)),</span>
        <span class="s">&quot;reference_checked&quot;</span><span class="p">:</span>  <span class="n">options</span><span class="o">.</span><span class="n">reference</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span>
    <span class="p">}</span>


<span class="k">def</span> <span class="nf">get_chrom_encoding</span><span class="p">(</span><span class="n">reference</span><span class="p">):</span></div>
<div class="viewcode-block" id="get_chrom_encoding"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.get_chrom_encoding">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gets the chromosome&#39;s encoding (e.g. 1 vs chr1, X vs 23, etc).</span>

<span class="sd">    Args:</span>
<span class="sd">        reference (pyfaidx.Fasta): the reference</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: the chromosome encoding</span>

<span class="sd">    The encoding is a dictionary where numerical chromosomes from 1 to 22 are</span>
<span class="sd">    the keys, and the encoded chromosomes (the one present in the reference)</span>
<span class="sd">    are the values. An example would be ``1 -&gt; chr1``.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The encoding</span>
    <span class="n">encoding</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c"># The autosomes</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="ow">in</span> <span class="n">reference</span><span class="p">:</span>
            <span class="n">encoding</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
        <span class="k">elif</span> <span class="s">&quot;chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="ow">in</span> <span class="n">reference</span><span class="p">:</span>
            <span class="n">encoding</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span> <span class="o">=</span> <span class="s">&quot;chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>

    <span class="c"># The X and Y chromosome</span>
    <span class="k">for</span> <span class="n">num_chrom</span><span class="p">,</span> <span class="n">char_chrom</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="s">&quot;23&quot;</span><span class="p">,</span> <span class="s">&quot;24&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s">&quot;X&quot;</span><span class="p">,</span> <span class="s">&quot;Y&quot;</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">num_chrom</span> <span class="ow">in</span> <span class="n">reference</span><span class="p">:</span>
            <span class="n">encoding</span><span class="p">[</span><span class="n">num_chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">num_chrom</span>
        <span class="k">elif</span> <span class="n">char_chrom</span> <span class="ow">in</span> <span class="n">reference</span><span class="p">:</span>
            <span class="n">encoding</span><span class="p">[</span><span class="n">num_chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">char_chrom</span>
        <span class="k">elif</span> <span class="s">&quot;chr&quot;</span> <span class="o">+</span> <span class="n">char_chrom</span> <span class="ow">in</span> <span class="n">reference</span><span class="p">:</span>
            <span class="n">encoding</span><span class="p">[</span><span class="n">num_chrom</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;chr&quot;</span> <span class="o">+</span> <span class="n">char_chrom</span>
        <span class="k">elif</span> <span class="s">&quot;chr&quot;</span> <span class="o">+</span> <span class="n">num_chrom</span> <span class="ow">in</span> <span class="n">reference</span><span class="p">:</span>
            <span class="n">encoding</span><span class="p">[</span><span class="n">num_chrom</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;chr&quot;</span> <span class="o">+</span> <span class="n">num_chrom</span>

    <span class="c"># The mitochondrial chromosome</span>
    <span class="n">possibilities</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;26&quot;</span><span class="p">,</span> <span class="s">&quot;M&quot;</span><span class="p">,</span> <span class="s">&quot;MT&quot;</span><span class="p">,</span> <span class="s">&quot;chrM&quot;</span><span class="p">,</span> <span class="s">&quot;chrMT&quot;</span><span class="p">,</span> <span class="s">&quot;chr26&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">possibility</span> <span class="ow">in</span> <span class="n">possibilities</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">possibility</span> <span class="ow">in</span> <span class="n">reference</span><span class="p">:</span>
            <span class="n">encoding</span><span class="p">[</span><span class="s">&quot;26&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">possibility</span>
            <span class="k">break</span>

    <span class="c"># Checking if we have everything</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">27</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">chrom</span> <span class="o">!=</span> <span class="mi">25</span> <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">encoding</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;{}: chromosome not in reference&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">encoding</span>


<span class="k">def</span> <span class="nf">is_reversed</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">encoding</span><span class="p">):</span></div>
<div class="viewcode-block" id="is_reversed"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.is_reversed">[docs]</a>    <span class="sd">&quot;&quot;&quot;Checks the strand using a reference, returns False if problem.</span>

<span class="sd">    Args:</span>
<span class="sd">        chrom (str): the chromosome</span>
<span class="sd">        pos (int): the position</span>
<span class="sd">        a1 (str): the first allele</span>
<span class="sd">        a2 (str): the second allele</span>
<span class="sd">        reference (pyfaidx.Fasta): the reference</span>
<span class="sd">        encoding (dict): the chromosome encoding in the reference</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: ``True`` if it&#39;s the complement, ``False`` otherwise (also</span>
<span class="sd">              returns ``False`` if there was a problem)</span>

<span class="sd">    The encoding (used to encode chromosome to search in the reference) is the</span>
<span class="sd">    dictionary returned by the :py:func:`get_chrom_encoding` function.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Upper!</span>
    <span class="n">a1</span> <span class="o">=</span> <span class="n">a1</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="n">a2</span> <span class="o">=</span> <span class="n">a2</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

    <span class="c"># Some Illumina chip has I/D (so if not A/C/G/T, we cannot check)</span>
    <span class="k">if</span> <span class="n">a1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">_complement</span> <span class="ow">or</span> <span class="n">a2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">_complement</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">False</span>

    <span class="c"># If the chromosome is not present, we suppose no strand problem</span>
    <span class="k">if</span> <span class="n">chrom</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">encoding</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">False</span>

    <span class="c"># Getting the nucleotide at this position</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">reference</span><span class="p">[</span><span class="n">encoding</span><span class="p">[</span><span class="n">chrom</span><span class="p">]][</span><span class="n">pos</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>

    <span class="c"># Is there one (if not, we suppose no strand problem)</span>
    <span class="k">if</span> <span class="n">ref</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">False</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">ref</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

    <span class="c"># Is REF a valid nucleotide?</span>
    <span class="k">if</span> <span class="n">ref</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">_complement</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">False</span>

    <span class="c"># If either a1 or a2 equals to ref, no strand problem</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">ref</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">a2</span> <span class="o">==</span> <span class="n">ref</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">False</span>

    <span class="c"># If the complement equals, then incorrect strand</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">_complement</span><span class="p">[</span><span class="n">a1</span><span class="p">]</span> <span class="o">==</span> <span class="n">ref</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">_complement</span><span class="p">[</span><span class="n">a2</span><span class="p">]</span> <span class="o">==</span> <span class="n">ref</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">True</span>

    <span class="c"># If nothing works, raising an exception</span>
    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;chr{}: {}: {}: {}/{}: &quot;</span>
                       <span class="s">&quot;invalid&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">get_cross_validation_results</span><span class="p">(</span><span class="n">glob_pattern</span><span class="p">):</span></div>
<div class="viewcode-block" id="get_cross_validation_results"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.get_cross_validation_results">[docs]</a>    <span class="sd">&quot;&quot;&quot;Creates a weighted mean for each chromosome for cross-validation.</span>

<span class="sd">    Args:</span>
<span class="sd">        glob_pattern (str): the pattern used to find files using :py:mod:`glob`</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: weighted mean by chromosome (cross-validation)</span>

<span class="sd">    The returned information includes the total number of genotyped tested, the</span>
<span class="sd">    total number of genotyped by chromosome, the two summary tables produced by</span>
<span class="sd">    IMPUTE2 (but using weighted means) for all chromosomes, and the two summary</span>
<span class="sd">    tables for each chromosome.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Gathering cross-validation statistics&quot;</span><span class="p">)</span>

    <span class="c"># The regular expressions used here</span>
    <span class="n">nb_genotypes_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">r&quot;^In the current analysis, IMPUTE2 masked, &quot;</span>
                                 <span class="s">r&quot;imputed, and evaluated (\d+) genotypes&quot;</span><span class="p">)</span>
    <span class="n">table_header_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">r&quot;Interval\s+#Genotypes\s+%Concordance\s+&quot;</span>
                                 <span class="s">r&quot;Interval\s+%Called\s+%Concordance&quot;</span><span class="p">)</span>
    <span class="n">split_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">r&quot;\s+&quot;</span><span class="p">)</span>

    <span class="c"># The intervals</span>
    <span class="n">table_1_intervals</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;[0.0-0.1]&quot;</span><span class="p">,</span> <span class="s">&quot;[0.1-0.2]&quot;</span><span class="p">,</span> <span class="s">&quot;[0.2-0.3]&quot;</span><span class="p">,</span> <span class="s">&quot;[0.3-0.4]&quot;</span><span class="p">,</span>
                         <span class="s">&quot;[0.4-0.5]&quot;</span><span class="p">,</span> <span class="s">&quot;[0.5-0.6]&quot;</span><span class="p">,</span> <span class="s">&quot;[0.6-0.7]&quot;</span><span class="p">,</span> <span class="s">&quot;[0.7-0.8]&quot;</span><span class="p">,</span>
                         <span class="s">&quot;[0.8-0.9]&quot;</span><span class="p">,</span> <span class="s">&quot;[0.9-1.0]&quot;</span><span class="p">]</span>
    <span class="n">table_2_intervals</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;[&gt;=0.0]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.1]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.2]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.3]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.4]&quot;</span><span class="p">,</span>
                         <span class="s">&quot;[&gt;=0.5]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.6]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.7]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.8]&quot;</span><span class="p">,</span> <span class="s">&quot;[&gt;=0.9]&quot;</span><span class="p">]</span>

    <span class="c"># The final data per chromosome</span>
    <span class="n">per_chrom_table_1</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">per_chrom_table_2</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">chrom_nb_geno</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c"># The final data for all the dataset</span>
    <span class="n">tot_concordance</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">tot_nb_geno</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">tot_weight</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">tot_cumm_called</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">tot_cumm_concordance</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">tot_cumm_weight</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">final_nb_genotypes</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c"># For each chromosome</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="n">filenames</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="n">glob_pattern</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">))</span>

        <span class="c"># The total number of genotypes for this chromosome</span>
        <span class="n">tot_chrom_nb_genotypes</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c"># The numbers for table 1</span>
        <span class="n">tot_chrom_concordance</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">tot_chrom_nb_geno</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">tot_chrom_weight</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="c"># The numbers for table 2</span>
        <span class="n">tot_chrom_cumm_called</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">tot_chrom_cumm_concordance</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">tot_chrom_cumm_weight</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                <span class="n">nb_genotypes</span> <span class="o">=</span> <span class="n">nb_genotypes_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">while</span> <span class="p">(</span><span class="n">line</span> <span class="o">!=</span> <span class="s">&quot;&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">nb_genotypes</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">):</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                    <span class="n">nb_genotypes</span> <span class="o">=</span> <span class="n">nb_genotypes_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

                <span class="c"># Check if we have data</span>
                <span class="k">if</span> <span class="n">nb_genotypes</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c"># We have the number of genotypes</span>
                <span class="n">nb_genotypes</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">nb_genotypes</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">nb_genotypes</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c"># Increasing the total number of genotypes for this chromosome</span>
                <span class="n">tot_chrom_nb_genotypes</span> <span class="o">+=</span> <span class="n">nb_genotypes</span>
                <span class="n">final_nb_genotypes</span> <span class="o">+=</span> <span class="n">nb_genotypes</span>

                <span class="c"># Looking for the headers of the two tables</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                <span class="n">table_header</span> <span class="o">=</span> <span class="n">table_header_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">while</span> <span class="p">(</span><span class="n">line</span> <span class="o">!=</span> <span class="s">&quot;&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">table_header</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">):</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                    <span class="n">table_header</span> <span class="o">=</span> <span class="n">table_header_re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

                <span class="c"># We now should have data</span>
                <span class="k">if</span> <span class="n">table_header</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;Problem with {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>

                <span class="c"># Now reading the data inside the tables</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
                    <span class="c"># Getting the value for the tables</span>
                    <span class="n">table</span> <span class="o">=</span> <span class="n">split_re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>

                    <span class="c"># The first table</span>
                    <span class="n">interval</span><span class="p">,</span> <span class="n">nb_geno</span><span class="p">,</span> <span class="n">concordance</span> <span class="o">=</span> <span class="n">table</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span>

                    <span class="c"># Casting to values</span>
                    <span class="n">nb_geno</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">nb_geno</span><span class="p">)</span>
                    <span class="n">concordance</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">concordance</span><span class="p">)</span>

                    <span class="c"># The total number of genotypes for this interval</span>
                    <span class="n">tot_nb_geno</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_geno</span>
                    <span class="n">tot_chrom_nb_geno</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_geno</span>

                    <span class="c"># The concordance for this interval</span>
                    <span class="n">tot_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="p">(</span><span class="n">concordance</span> <span class="o">*</span> <span class="n">nb_geno</span><span class="p">)</span>
                    <span class="n">tot_chrom_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="p">(</span><span class="n">concordance</span> <span class="o">*</span> <span class="n">nb_geno</span><span class="p">)</span>

                    <span class="c"># The weight for this interval</span>
                    <span class="n">tot_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_geno</span>
                    <span class="n">tot_chrom_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_geno</span>

                    <span class="c"># The second table</span>
                    <span class="n">table</span> <span class="o">=</span> <span class="n">table</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
                    <span class="n">interval</span> <span class="o">=</span> <span class="s">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">table</span><span class="p">[:</span><span class="mi">3</span><span class="p">])</span>
                    <span class="n">pc_called</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">table</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">])</span>
                    <span class="n">pc_concordance</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">table</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

                    <span class="c"># The interval number of genotypes</span>
                    <span class="n">nb_called</span> <span class="o">=</span> <span class="n">pc_called</span> <span class="o">*</span> <span class="n">nb_genotypes</span> <span class="o">/</span> <span class="mi">100</span>
                    <span class="n">tot_chrom_cumm_called</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_called</span>
                    <span class="n">tot_cumm_called</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_called</span>

                    <span class="c"># The concordance for this interval</span>
                    <span class="n">weighted_value</span> <span class="o">=</span> <span class="n">nb_called</span> <span class="o">*</span> <span class="n">pc_concordance</span>
                    <span class="n">tot_chrom_cumm_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">weighted_value</span>
                    <span class="n">tot_cumm_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">weighted_value</span>

                    <span class="c"># The weight for this interval</span>
                    <span class="n">tot_chrom_cumm_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_called</span>
                    <span class="n">tot_cumm_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span> <span class="o">+=</span> <span class="n">nb_called</span>

        <span class="c"># Computing the weighted average for the first table</span>
        <span class="n">table_1_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">table_1_intervals</span><span class="p">:</span>
            <span class="n">nb_geno</span> <span class="o">=</span> <span class="n">tot_chrom_nb_geno</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
            <span class="n">concordance</span> <span class="o">=</span> <span class="n">tot_chrom_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
            <span class="n">weight</span> <span class="o">=</span> <span class="n">tot_chrom_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>

            <span class="c"># Computing the weighted concordance</span>
            <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">weight</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="n">concordance</span> <span class="o">/</span> <span class="n">weight</span>

            <span class="c"># Saving the data for the current chromosome</span>
            <span class="n">table_1_data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span>
                <span class="n">interval</span><span class="p">,</span>
                <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_geno</span><span class="p">),</span>
                <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">weighted_concordance</span><span class="p">),</span>
            <span class="p">])</span>

        <span class="c"># Computing the weighed average for the second table</span>
        <span class="n">table_2_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">table_2_intervals</span><span class="p">:</span>
            <span class="n">nb_called</span> <span class="o">=</span> <span class="n">tot_chrom_cumm_called</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
            <span class="n">concordance</span> <span class="o">=</span> <span class="n">tot_chrom_cumm_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
            <span class="n">weight</span> <span class="o">=</span> <span class="n">tot_chrom_cumm_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>

            <span class="c"># Computing the weighted concordance</span>
            <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">weight</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="n">concordance</span> <span class="o">/</span> <span class="n">weight</span>

            <span class="c"># Saving the data for the current chromosome</span>
            <span class="n">table_2_data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span>
                <span class="n">interval</span><span class="p">,</span>
                <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_called</span> <span class="o">/</span> <span class="n">tot_chrom_nb_genotypes</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
                <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">weighted_concordance</span><span class="p">),</span>
            <span class="p">])</span>

        <span class="c"># Saving the tables</span>
        <span class="n">per_chrom_table_1</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">table_1_data</span>
        <span class="n">per_chrom_table_2</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">table_2_data</span>
        <span class="n">chrom_nb_geno</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">tot_chrom_nb_genotypes</span>

    <span class="c"># The final table 1</span>
    <span class="n">table_1_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">table_1_intervals</span><span class="p">:</span>
        <span class="n">nb_geno</span> <span class="o">=</span> <span class="n">tot_nb_geno</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
        <span class="n">concordance</span> <span class="o">=</span> <span class="n">tot_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
        <span class="n">weight</span> <span class="o">=</span> <span class="n">tot_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>

        <span class="c"># Computing the weighted concordance</span>
        <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">weight</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="n">concordance</span> <span class="o">/</span> <span class="n">weight</span>

        <span class="n">table_1_data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span>
            <span class="n">interval</span><span class="p">,</span>
            <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_geno</span><span class="p">),</span>
            <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">weighted_concordance</span><span class="p">),</span>
        <span class="p">])</span>

    <span class="c"># The final table 2</span>
    <span class="n">table_2_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">interval</span> <span class="ow">in</span> <span class="n">table_2_intervals</span><span class="p">:</span>
        <span class="n">nb_called</span> <span class="o">=</span> <span class="n">tot_cumm_called</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
        <span class="n">concordance</span> <span class="o">=</span> <span class="n">tot_cumm_concordance</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>
        <span class="n">weight</span> <span class="o">=</span> <span class="n">tot_cumm_weight</span><span class="p">[</span><span class="n">interval</span><span class="p">]</span>

        <span class="c"># Computing the weighted concordance</span>
        <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">weight</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">weighted_concordance</span> <span class="o">=</span> <span class="n">concordance</span> <span class="o">/</span> <span class="n">weight</span>

        <span class="c"># Saving the data for all the chromosome</span>
        <span class="n">table_2_data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span>
            <span class="n">interval</span><span class="p">,</span>
            <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_called</span> <span class="o">/</span> <span class="n">final_nb_genotypes</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
            <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">weighted_concordance</span><span class="p">),</span>
        <span class="p">])</span>

    <span class="c"># Returning the data</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="s">&quot;cross_validation_final_nb_genotypes&quot;</span><span class="p">:</span> <span class="n">final_nb_genotypes</span><span class="p">,</span>
        <span class="s">&quot;cross_validation_nb_genotypes_chrom&quot;</span><span class="p">:</span> <span class="n">chrom_nb_geno</span><span class="p">,</span>
        <span class="s">&quot;cross_validation_table_1&quot;</span><span class="p">:</span>            <span class="n">table_1_data</span><span class="p">,</span>
        <span class="s">&quot;cross_validation_table_2&quot;</span><span class="p">:</span>            <span class="n">table_2_data</span><span class="p">,</span>
        <span class="s">&quot;cross_validation_table_1_chrom&quot;</span><span class="p">:</span>      <span class="n">per_chrom_table_1</span><span class="p">,</span>
        <span class="s">&quot;cross_validation_table_2_chrom&quot;</span><span class="p">:</span>      <span class="n">per_chrom_table_2</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">def</span> <span class="nf">gather_imputation_stats</span><span class="p">(</span><span class="n">prob_t</span><span class="p">,</span> <span class="n">completion_t</span><span class="p">,</span> <span class="n">nb_samples</span><span class="p">,</span> <span class="n">missing</span><span class="p">,</span> <span class="n">o_dir</span><span class="p">):</span></div>
<div class="viewcode-block" id="gather_imputation_stats"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.gather_imputation_stats">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gathers imputation statistics from the merged dataset.</span>

<span class="sd">    Args:</span>
<span class="sd">        prob_t (float): the probability threshold (&gt;= t)</span>
<span class="sd">        completion_t (float): the completion threshold (&gt;= t)</span>
<span class="sd">        nb_samples (int): the number of samples</span>
<span class="sd">        missing (pandas.DataFrame): the missing rate</span>
<span class="sd">        o_dir (str): the output directory</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: imputation statistics</span>

<span class="sd">    The information returned includes the following (all values are formated in</span>
<span class="sd">    strings):</span>

<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | Information                    | Description                            |</span>
<span class="sd">    +================================+========================================+</span>
<span class="sd">    | ``prob_threshold``             | the probability threshold used (&gt;= t)  |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_imputed``                 | the number of imputed sites            |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``average_comp_rate``          | the average completion rate            |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``rate_threshold``             | the completion rate threshold (&gt;= t)   |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_good_sites``              | the number of &quot;good&quot; sites (that pass  |</span>
<span class="sd">    |                                | the different probability and          |</span>
<span class="sd">    |                                | completion thresholds)                 |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``pct_good_sites``             | the percentage of &quot;good&quot; sites (that   |</span>
<span class="sd">    |                                | pass the different probability and     |</span>
<span class="sd">    |                                | completion thresholds)                 |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``average_comp_rate_cleaned``  | the average completion rate of &quot;good&quot;  |</span>
<span class="sd">    |                                | sites                                  |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``mean_missing``               | the mean number of missing genotypes   |</span>
<span class="sd">    |                                | per sample (according to the           |</span>
<span class="sd">    |                                | probability and completion thresholds) |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_samples``                 | the total number of samples            |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_genotyped``               | the number of genotyped sites in the   |</span>
<span class="sd">    |                                | original dataset that are included in  |</span>
<span class="sd">    |                                | the imputed dataset                    |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_genotyped_not_complete``  | The number of original markers with a  |</span>
<span class="sd">    |                                | completion rate lower than 100%        |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``pct_genotyped_not_complete`` | The percentage of original markers     |</span>
<span class="sd">    |                                | with a completion rate lower than 100% |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_geno_now_complete``       | The number of original (genotyped)     |</span>
<span class="sd">    |                                | markers which are now 100% complete    |</span>
<span class="sd">    |                                | after imputation                       |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_missing_geno``            | The number of missing genotypes        |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``pct_geno_now_complete``      | The percentage of now complete         |</span>
<span class="sd">    |                                | genotypes (according to the number of  |</span>
<span class="sd">    |                                | previously missing ones)               |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>
<span class="sd">    | ``nb_site_now_complete``       | The number of previously genotyped     |</span>
<span class="sd">    |                                | markers that are now 100% complete     |</span>
<span class="sd">    |                                | after imputation                       |</span>
<span class="sd">    +--------------------------------+----------------------------------------+</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Gathering imputation statistics&quot;</span><span class="p">)</span>

    <span class="c"># The stats we want to gather</span>
    <span class="c"># For all the sites</span>
    <span class="n">tot_nb_sites</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">sum_rates</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c"># For the best sites</span>
    <span class="n">tot_good_sites</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">sum_good_rates</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c"># All imputed sites</span>
    <span class="n">genotyped_sites</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c"># For each chromosome, get the statistics</span>
    <span class="n">filename_template</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">o_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;final_impute2&quot;</span><span class="p">,</span>
                                     <span class="s">&quot;chr{chrom}.imputed.{suffix}&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - chromosome {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>

        <span class="c"># First, we read the imputed sites</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">filename_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
                                            <span class="n">suffix</span><span class="o">=</span><span class="s">&quot;imputed_sites&quot;</span><span class="p">)</span>
        <span class="n">imputed_sites</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">imputed_sites</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">i_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()}</span>

        <span class="c"># Then, we read the completion using a DataFrame</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">filename_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
                                            <span class="n">suffix</span><span class="o">=</span><span class="s">&quot;completion_rates&quot;</span><span class="p">)</span>
        <span class="n">completion_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>

        <span class="c"># Saving the stats for all sites</span>
        <span class="n">tot_nb_sites</span> <span class="o">+=</span> <span class="n">completion_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sum_rates</span> <span class="o">+=</span> <span class="n">completion_data</span><span class="o">.</span><span class="n">completion_rate</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

        <span class="c"># Saving the stats for good sites</span>
        <span class="n">good_sites</span> <span class="o">=</span> <span class="n">completion_data</span><span class="o">.</span><span class="n">completion_rate</span> <span class="o">&gt;=</span> <span class="n">completion_t</span>
        <span class="n">tot_good_sites</span> <span class="o">+=</span> <span class="n">completion_data</span><span class="p">[</span><span class="n">good_sites</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sum_good_rates</span> <span class="o">+=</span> <span class="n">completion_data</span><span class="p">[</span><span class="n">good_sites</span><span class="p">]</span><span class="o">.</span><span class="n">completion_rate</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

        <span class="c"># Saving the imputed sites call rates</span>
        <span class="n">imputed_sites</span> <span class="o">=</span> <span class="n">completion_data</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">imputed_sites</span><span class="p">)</span>
        <span class="n">genotyped_sites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">completion_data</span><span class="p">[</span><span class="o">~</span><span class="n">imputed_sites</span><span class="p">])</span>

    <span class="c"># Concatenating the completion rates for the genotyped sites</span>
    <span class="n">genotyped_sites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">genotyped_sites</span><span class="p">)</span>
    <span class="n">now_incomplete_sites</span> <span class="o">=</span> <span class="n">genotyped_sites</span><span class="o">.</span><span class="n">nb_missing</span> <span class="o">&gt;</span> <span class="mi">0</span>

    <span class="c"># Getting the original missing rates for the genotyped sites</span>
    <span class="n">sites</span> <span class="o">=</span> <span class="n">missing</span><span class="o">.</span><span class="n">SNP</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genotyped_sites</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="n">incomplete_sites</span> <span class="o">=</span> <span class="n">missing</span><span class="o">.</span><span class="n">N_MISS</span> <span class="o">&gt;</span> <span class="mi">0</span>

    <span class="c"># Gathering genotyped sites statistics</span>
    <span class="n">nb_genotyped_sites</span> <span class="o">=</span> <span class="n">genotyped_sites</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">nb_incomplete_sites</span> <span class="o">=</span> <span class="n">missing</span><span class="p">[</span><span class="n">sites</span> <span class="o">&amp;</span> <span class="n">incomplete_sites</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">pct_incomplete_sites</span> <span class="o">=</span> <span class="n">nb_incomplete_sites</span> <span class="o">/</span> <span class="n">nb_genotyped_sites</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">nb_sites_now_complete</span> <span class="o">=</span> <span class="n">genotyped_sites</span><span class="p">[</span><span class="o">~</span><span class="n">now_incomplete_sites</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">nb_missing_geno</span> <span class="o">=</span> <span class="n">missing</span><span class="p">[</span><span class="n">sites</span> <span class="o">&amp;</span> <span class="n">incomplete_sites</span><span class="p">]</span><span class="o">.</span><span class="n">N_MISS</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">now_nb_missing_geno</span> <span class="o">=</span> <span class="n">genotyped_sites</span><span class="p">[</span><span class="n">now_incomplete_sites</span><span class="p">]</span>
    <span class="n">now_nb_missing_geno</span> <span class="o">=</span> <span class="n">now_nb_missing_geno</span><span class="o">.</span><span class="n">nb_missing</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">nb_geno_now_complete</span> <span class="o">=</span> <span class="n">nb_missing_geno</span> <span class="o">-</span> <span class="n">now_nb_missing_geno</span>
    <span class="n">pct_geno_now_complete</span> <span class="o">=</span> <span class="n">nb_geno_now_complete</span> <span class="o">/</span> <span class="n">nb_missing_geno</span> <span class="o">*</span> <span class="mi">100</span>

    <span class="c"># Computing the rates</span>
    <span class="n">comp_rate</span> <span class="o">=</span> <span class="n">sum_rates</span> <span class="o">/</span> <span class="n">tot_nb_sites</span>
    <span class="n">good_comp_rate</span> <span class="o">=</span> <span class="n">sum_good_rates</span> <span class="o">/</span> <span class="n">tot_good_sites</span>

    <span class="c"># Other statistics</span>
    <span class="n">pct_good_sites</span> <span class="o">=</span> <span class="n">tot_good_sites</span> <span class="o">/</span> <span class="n">tot_nb_sites</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">mean_missing</span> <span class="o">=</span> <span class="n">nb_samples</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">good_comp_rate</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s">&quot;prob_threshold&quot;</span><span class="p">:</span>             <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">prob_t</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
        <span class="s">&quot;nb_imputed&quot;</span><span class="p">:</span>                 <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tot_nb_sites</span><span class="p">),</span>
        <span class="s">&quot;average_comp_rate&quot;</span><span class="p">:</span>          <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">comp_rate</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
        <span class="s">&quot;rate_threshold&quot;</span><span class="p">:</span>             <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">completion_t</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
        <span class="s">&quot;nb_good_sites&quot;</span><span class="p">:</span>              <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tot_good_sites</span><span class="p">),</span>
        <span class="s">&quot;pct_good_sites&quot;</span><span class="p">:</span>             <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_good_sites</span><span class="p">),</span>
        <span class="s">&quot;average_comp_rate_cleaned&quot;</span><span class="p">:</span>  <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">good_comp_rate</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
        <span class="s">&quot;mean_missing&quot;</span><span class="p">:</span>               <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mean_missing</span><span class="p">),</span>
        <span class="s">&quot;nb_samples&quot;</span><span class="p">:</span>                 <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_samples</span><span class="p">),</span>
        <span class="s">&quot;nb_genotyped&quot;</span><span class="p">:</span>               <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_genotyped_sites</span><span class="p">),</span>
        <span class="s">&quot;nb_genotyped_not_complete&quot;</span><span class="p">:</span>  <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_incomplete_sites</span><span class="p">),</span>
        <span class="s">&quot;pct_genotyped_not_complete&quot;</span><span class="p">:</span> <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_incomplete_sites</span><span class="p">),</span>
        <span class="s">&quot;nb_geno_now_complete&quot;</span><span class="p">:</span>       <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_geno_now_complete</span><span class="p">),</span>
        <span class="s">&quot;nb_missing_geno&quot;</span><span class="p">:</span>            <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_missing_geno</span><span class="p">),</span>
        <span class="s">&quot;pct_geno_now_complete&quot;</span><span class="p">:</span>      <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_geno_now_complete</span><span class="p">),</span>
        <span class="s">&quot;nb_site_now_complete&quot;</span><span class="p">:</span>       <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_sites_now_complete</span><span class="p">)</span>
    <span class="p">}</span>


<span class="k">def</span> <span class="nf">gather_maf_stats</span><span class="p">(</span><span class="n">o_dir</span><span class="p">):</span></div>
<div class="viewcode-block" id="gather_maf_stats"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.gather_maf_stats">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gather minor allele frequencies from imputation.</span>

<span class="sd">    Args:</span>
<span class="sd">        o_dir (str): the output directory</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: frequency information</span>

<span class="sd">    The following information about minor allele frequencies (MAF) are</span>
<span class="sd">    returned (all values are formatted in strings):</span>

<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | Information              | Description                                  |</span>
<span class="sd">    +==========================+==============================================+</span>
<span class="sd">    | ``nb_maf_nan``           | the number of markers with invalid MAF       |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``nb_marker_with_maf``   | the number of markers with valid MAF         |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``nb_maf_geq_01``        | the number of markers with MAF &gt;= 1%         |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``nb_maf_geq_05``        | the number of markers with MAF &gt;= 5%         |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``nb_maf_lt_05``         | the number of markers with MAF &lt; 5%          |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``nb_maf_lt_01``         | the number of markers with MAF &lt; 1%          |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``nb_maf_geq_01_lt_05``  | the number of markers with 1% &gt;= MAF &lt; 5%    |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``pct_maf_geq_01``       | the percentage of markers with MAF &gt;= 1%     |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``pct_maf_geq_05``       | the percentage of markers with MAF &gt;= 5%     |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``pct_maf_lt_05``        | the percentage of markers with MAF &lt; 5%      |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``pct_maf_lt_01``        | the percentage of markers with MAF &lt; 1%      |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``pct_maf_geq_01_lt_05`` | the percentage of markers with               |</span>
<span class="sd">    |                          | 1% &gt;= MAF &lt; 5%                               |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>
<span class="sd">    | ``frequency_pie``        | the name of the file containing the pie      |</span>
<span class="sd">    |                          | chart (if it was created)                    |</span>
<span class="sd">    +--------------------------+----------------------------------------------+</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Gathering frequency statistics&quot;</span><span class="p">)</span>

    <span class="c"># The statistics we want to gather</span>
    <span class="n">nb_marker_with_maf</span> <span class="o">=</span> <span class="mi">0</span>   <span class="c"># The number of markers for which we have a MAF</span>
    <span class="n">nb_maf_geq_01</span> <span class="o">=</span> <span class="mi">0</span>        <span class="c"># The number of markers with MAF &gt;= 0.01</span>
    <span class="n">nb_maf_geq_05</span> <span class="o">=</span> <span class="mi">0</span>        <span class="c"># The number of markers with MAF &gt;= 0.05</span>
    <span class="n">nb_maf_lt_05</span> <span class="o">=</span> <span class="mi">0</span>         <span class="c"># The number of markers with MAF &lt; 0.05</span>
    <span class="n">nb_maf_lt_01</span> <span class="o">=</span> <span class="mi">0</span>         <span class="c"># The number of markers with MAF &lt; 0.01</span>
    <span class="n">nb_maf_geq_01_lt_05</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c"># The number of markers with 0.01 &lt;= MAF &lt; 0.05</span>
    <span class="n">nb_maf_nan</span> <span class="o">=</span> <span class="mi">0</span>           <span class="c"># The number of markers without MAF</span>

    <span class="c"># For each chromosome, get the MAF statistics</span>
    <span class="n">filename_template</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">o_dir</span><span class="p">,</span> <span class="s">&quot;chr{chrom}&quot;</span><span class="p">,</span> <span class="s">&quot;final_impute2&quot;</span><span class="p">,</span>
                                     <span class="s">&quot;chr{chrom}.imputed.{suffix}&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - chromosome {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>

        <span class="c"># The name of the file</span>
        <span class="n">maf_filename</span> <span class="o">=</span> <span class="n">filename_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s">&quot;maf&quot;</span><span class="p">)</span>
        <span class="n">good_sites_filename</span> <span class="o">=</span> <span class="n">filename_template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
                                                       <span class="n">suffix</span><span class="o">=</span><span class="s">&quot;good_sites&quot;</span><span class="p">)</span>

        <span class="c"># Checking the file exists</span>
        <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="p">[</span><span class="n">maf_filename</span><span class="p">,</span> <span class="n">good_sites_filename</span><span class="p">]:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>

        <span class="c"># Reading the list of good sites</span>
        <span class="n">good_sites</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">good_sites_filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">good_sites</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">i_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>

        <span class="c"># Reading the file using pandas</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">maf_filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>

        <span class="c"># Keeping only the good sites</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="n">maf</span><span class="p">[</span><span class="n">maf</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">good_sites</span><span class="p">)]</span>

        <span class="c"># Excluding sites with no MAF (NaN values)</span>
        <span class="n">null_maf</span> <span class="o">=</span> <span class="n">maf</span><span class="o">.</span><span class="n">maf</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span>
        <span class="n">nb_nan</span> <span class="o">=</span> <span class="n">null_maf</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">maf</span> <span class="o">=</span> <span class="n">maf</span><span class="p">[</span><span class="o">~</span><span class="n">null_maf</span><span class="p">]</span>

        <span class="c"># There should not be any NaN sites...</span>
        <span class="k">if</span> <span class="n">nb_nan</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;chr{}: good sites with invalid MAF &quot;</span>
                            <span class="s">&quot;(NaN)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>

        <span class="c"># Checking we have MAF (and not just frequencies)</span>
        <span class="n">maf_description</span> <span class="o">=</span> <span class="n">maf</span><span class="o">.</span><span class="n">maf</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">maf_description</span><span class="p">[</span><span class="s">&quot;max&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="n">bad</span> <span class="o">=</span> <span class="n">maf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">maf</span><span class="o">.</span><span class="n">maf</span><span class="o">.</span><span class="n">idxmax</span><span class="p">(),</span> <span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;maf&quot;</span><span class="p">]]</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: invalid MAF&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">bad</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">]),</span>
                                                            <span class="nb">round</span><span class="p">(</span><span class="n">bad</span><span class="o">.</span><span class="n">maf</span><span class="p">,</span> <span class="mi">3</span><span class="p">)))</span>
        <span class="k">if</span> <span class="n">maf_description</span><span class="p">[</span><span class="s">&quot;max&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">bad</span> <span class="o">=</span> <span class="n">maf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">maf</span><span class="o">.</span><span class="n">maf</span><span class="o">.</span><span class="n">idxmin</span><span class="p">(),</span> <span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;maf&quot;</span><span class="p">]]</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: invalid MAF&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">bad</span><span class="p">[</span><span class="s">&quot;name&quot;</span><span class="p">]),</span>
                                                            <span class="nb">round</span><span class="p">(</span><span class="n">bad</span><span class="o">.</span><span class="n">maf</span><span class="p">,</span> <span class="mi">3</span><span class="p">)))</span>

        <span class="c"># Some of the true/false we need to keep (to not compute multiple time)</span>
        <span class="n">maf_geq_01</span> <span class="o">=</span> <span class="n">maf</span><span class="o">.</span><span class="n">maf</span> <span class="o">&gt;=</span> <span class="mf">0.01</span>
        <span class="n">maf_lt_05</span> <span class="o">=</span> <span class="n">maf</span><span class="o">.</span><span class="n">maf</span> <span class="o">&lt;</span> <span class="mf">0.05</span>

        <span class="c"># Updating the statistics</span>
        <span class="n">nb_marker_with_maf</span> <span class="o">+=</span> <span class="n">maf</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">nb_maf_nan</span> <span class="o">+=</span> <span class="n">nb_nan</span>
        <span class="n">nb_maf_geq_01</span> <span class="o">+=</span> <span class="n">maf_geq_01</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">nb_maf_geq_05</span> <span class="o">+=</span> <span class="p">(</span><span class="n">maf</span><span class="o">.</span><span class="n">maf</span> <span class="o">&gt;=</span> <span class="mf">0.05</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">nb_maf_lt_05</span> <span class="o">+=</span> <span class="n">maf_lt_05</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">nb_maf_lt_01</span> <span class="o">+=</span> <span class="p">(</span><span class="n">maf</span><span class="o">.</span><span class="n">maf</span> <span class="o">&lt;</span> <span class="mf">0.01</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">nb_maf_geq_01_lt_05</span> <span class="o">+=</span> <span class="p">(</span><span class="n">maf_geq_01</span> <span class="o">&amp;</span> <span class="n">maf_lt_05</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

    <span class="c"># Checking</span>
    <span class="n">nb_total</span> <span class="o">=</span> <span class="n">nb_maf_lt_01</span> <span class="o">+</span> <span class="n">nb_maf_geq_01_lt_05</span> <span class="o">+</span> <span class="n">nb_maf_geq_05</span>
    <span class="k">if</span> <span class="n">nb_total</span> <span class="o">!=</span> <span class="n">nb_marker_with_maf</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;something went wrong&quot;</span><span class="p">)</span>

    <span class="c"># Computing the percentages</span>
    <span class="n">pct_maf_geq_01</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">pct_maf_geq_05</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">pct_maf_lt_05</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">pct_maf_lt_01</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">pct_maf_geq_01_lt_05</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="n">nb_marker_with_maf</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">pct_maf_geq_01</span> <span class="o">=</span> <span class="n">nb_maf_geq_01</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="n">pct_maf_geq_05</span> <span class="o">=</span> <span class="n">nb_maf_geq_05</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="n">pct_maf_lt_05</span> <span class="o">=</span> <span class="n">nb_maf_lt_05</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="n">pct_maf_lt_01</span> <span class="o">=</span> <span class="n">nb_maf_lt_01</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="n">pct_maf_geq_01_lt_05</span> <span class="o">=</span> <span class="n">nb_maf_geq_01_lt_05</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;There were no marker with MAF (something went wrong)&quot;</span><span class="p">)</span>

    <span class="c"># Generating a pie chart if matplotlib is installed</span>
    <span class="n">frequency_pie</span> <span class="o">=</span> <span class="s">&quot;&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">nb_marker_with_maf</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">and</span> <span class="n">HAS_MATPLOTLIB</span><span class="p">:</span>
        <span class="c"># Creating a figure and axe</span>
        <span class="n">figure</span><span class="p">,</span> <span class="n">axe</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>

        <span class="c"># The colors</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;#0099CC&quot;</span><span class="p">,</span> <span class="s">&quot;#669900&quot;</span><span class="p">,</span> <span class="s">&quot;#FF8800&quot;</span><span class="p">]</span>

        <span class="c"># The data for the pie chart</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s">&quot;{:.1f}%&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_lt_01</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
            <span class="s">&quot;{:.1f}%&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_geq_01_lt_05</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
            <span class="s">&quot;{:.1f}%&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_geq_05</span> <span class="o">/</span> <span class="n">nb_marker_with_maf</span> <span class="o">*</span> <span class="mi">100</span><span class="p">),</span>
        <span class="p">]</span>
        <span class="n">sizes</span> <span class="o">=</span> <span class="p">[</span><span class="n">nb_maf_lt_01</span><span class="p">,</span> <span class="n">nb_maf_geq_01_lt_05</span><span class="p">,</span> <span class="n">nb_maf_geq_05</span><span class="p">]</span>
        <span class="n">explode</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">)</span>
        <span class="n">wedges</span><span class="p">,</span> <span class="n">texts</span> <span class="o">=</span> <span class="n">axe</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">sizes</span><span class="p">,</span> <span class="n">explode</span><span class="o">=</span><span class="n">explode</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span>
                                <span class="n">colors</span><span class="o">=</span><span class="n">colors</span><span class="p">,</span> <span class="n">startangle</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span>
                                <span class="n">wedgeprops</span><span class="o">=</span><span class="p">{</span><span class="s">&quot;linewidth&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">},</span>
                                <span class="n">textprops</span><span class="o">=</span><span class="p">{</span><span class="s">&quot;fontsize&quot;</span><span class="p">:</span> <span class="mi">12</span><span class="p">,</span> <span class="s">&quot;weight&quot;</span><span class="p">:</span> <span class="s">&quot;bold&quot;</span><span class="p">})</span>

        <span class="c"># Changing the label parameters</span>
        <span class="k">for</span> <span class="n">text</span> <span class="ow">in</span> <span class="n">texts</span><span class="p">:</span>
            <span class="n">text</span><span class="o">.</span><span class="n">set_bbox</span><span class="p">({</span><span class="s">&quot;boxstyle&quot;</span><span class="p">:</span> <span class="s">&quot;round&quot;</span><span class="p">,</span> <span class="s">&quot;fc&quot;</span><span class="p">:</span> <span class="s">&quot;#C0C0C0&quot;</span><span class="p">,</span>
                           <span class="s">&quot;ec&quot;</span><span class="p">:</span> <span class="s">&quot;#C0C0C0&quot;</span><span class="p">})</span>

        <span class="c"># Shrink current axis by 50%</span>
        <span class="n">bbox</span> <span class="o">=</span> <span class="n">axe</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
        <span class="n">axe</span><span class="o">.</span><span class="n">set_position</span><span class="p">([</span><span class="n">bbox</span><span class="o">.</span><span class="n">x0</span><span class="p">,</span> <span class="n">bbox</span><span class="o">.</span><span class="n">y0</span><span class="p">,</span> <span class="n">bbox</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="mf">0.5</span><span class="p">,</span>
                         <span class="n">bbox</span><span class="o">.</span><span class="n">height</span> <span class="o">*</span> <span class="mf">0.5</span><span class="p">])</span>

        <span class="c"># If no restriction was performed while imputing, there will be a high</span>
        <span class="c"># majority of ultra rare variants... We need to move the text box if</span>
        <span class="c"># they touch</span>
        <span class="c"># Getting the renderer and the bboxes</span>
        <span class="n">renderer</span> <span class="o">=</span> <span class="n">figure</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">get_renderer</span><span class="p">()</span>
        <span class="n">bbox_1</span> <span class="o">=</span> <span class="n">texts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_window_extent</span><span class="p">(</span><span class="n">renderer</span><span class="o">=</span><span class="n">renderer</span><span class="p">)</span>
        <span class="n">bbox_2</span> <span class="o">=</span> <span class="n">texts</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get_window_extent</span><span class="p">(</span><span class="n">renderer</span><span class="o">=</span><span class="n">renderer</span><span class="p">)</span>
        <span class="k">while</span> <span class="n">bbox_1</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">bbox_2</span><span class="p">):</span>
            <span class="c"># Moving the first label a bit</span>
            <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">texts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
            <span class="n">texts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_x</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mf">0.006</span><span class="p">)</span>
            <span class="n">bbox_1</span> <span class="o">=</span> <span class="n">texts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_window_extent</span><span class="p">(</span><span class="n">renderer</span><span class="o">=</span><span class="n">renderer</span><span class="p">)</span>

            <span class="c"># Moving the second label (MAF &gt;= 0.05)</span>
            <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">texts</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
            <span class="n">texts</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_x</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="mf">0.036</span><span class="p">)</span>
            <span class="n">bbox_2</span> <span class="o">=</span> <span class="n">texts</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get_window_extent</span><span class="p">(</span><span class="n">renderer</span><span class="o">=</span><span class="n">renderer</span><span class="p">)</span>

        <span class="c"># Adding a legend in the margin with custom patches</span>
        <span class="n">ultra_rare</span> <span class="o">=</span> <span class="n">mpatches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">rare</span> <span class="o">=</span> <span class="n">mpatches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">common</span> <span class="o">=</span> <span class="n">mpatches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">axe</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span>
            <span class="p">[</span><span class="n">ultra_rare</span><span class="p">,</span> <span class="n">rare</span><span class="p">,</span> <span class="n">common</span><span class="p">],</span>
            <span class="p">[</span><span class="s">r&quot;$MAF &lt; 1\%$&quot;</span><span class="p">,</span> <span class="s">r&quot;$1\% \leq MAF &lt; 5\%$&quot;</span><span class="p">,</span> <span class="s">r&quot;$MAF \geq 5\%$&quot;</span><span class="p">],</span>
            <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.3</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
            <span class="n">loc</span><span class="o">=</span><span class="s">&quot;upper left&quot;</span><span class="p">,</span>
            <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">frameon</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
            <span class="n">fontsize</span><span class="o">=</span><span class="mi">21</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c"># Setting the axis to equal size</span>
        <span class="n">axe</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s">&quot;equal&quot;</span><span class="p">)</span>

        <span class="c"># Saving and closing the figure</span>
        <span class="n">frequency_pie</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">o_dir</span><span class="p">,</span> <span class="s">&quot;frequency_pie.pdf&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">frequency_pie</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s">&quot;tight&quot;</span><span class="p">,</span> <span class="n">figure</span><span class="o">=</span><span class="n">figure</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">figure</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s">&quot;nb_maf_nan&quot;</span><span class="p">:</span>           <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_nan</span><span class="p">),</span>
        <span class="s">&quot;nb_marker_with_maf&quot;</span><span class="p">:</span>   <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_marker_with_maf</span><span class="p">),</span>
        <span class="s">&quot;nb_maf_geq_01&quot;</span><span class="p">:</span>        <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_geq_01</span><span class="p">),</span>
        <span class="s">&quot;nb_maf_geq_05&quot;</span><span class="p">:</span>        <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_geq_05</span><span class="p">),</span>
        <span class="s">&quot;nb_maf_lt_05&quot;</span><span class="p">:</span>         <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_lt_05</span><span class="p">),</span>
        <span class="s">&quot;nb_maf_lt_01&quot;</span><span class="p">:</span>         <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_lt_01</span><span class="p">),</span>
        <span class="s">&quot;nb_maf_geq_01_lt_05&quot;</span><span class="p">:</span>  <span class="s">&quot;{:,d}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_maf_geq_01_lt_05</span><span class="p">),</span>
        <span class="s">&quot;pct_maf_geq_01&quot;</span><span class="p">:</span>       <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_maf_geq_01</span><span class="p">),</span>
        <span class="s">&quot;pct_maf_geq_05&quot;</span><span class="p">:</span>       <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_maf_geq_05</span><span class="p">),</span>
        <span class="s">&quot;pct_maf_lt_05&quot;</span><span class="p">:</span>        <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_maf_lt_05</span><span class="p">),</span>
        <span class="s">&quot;pct_maf_lt_01&quot;</span><span class="p">:</span>        <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_maf_lt_01</span><span class="p">),</span>
        <span class="s">&quot;pct_maf_geq_01_lt_05&quot;</span><span class="p">:</span> <span class="s">&quot;{:.1f}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pct_maf_geq_01_lt_05</span><span class="p">),</span>
        <span class="s">&quot;frequency_pie&quot;</span><span class="p">:</span>        <span class="n">frequency_pie</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">def</span> <span class="nf">gather_execution_time</span><span class="p">(</span><span class="n">db_name</span><span class="p">):</span></div>
<div class="viewcode-block" id="gather_execution_time"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.gather_execution_time">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gather all the execution times.</span>

<span class="sd">    Args:</span>
<span class="sd">        db_name (str): the name of the DB</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: the execution time for all tasks</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Getting all the execution time from the DB</span>
    <span class="n">exec_time</span> <span class="o">=</span> <span class="n">get_all_runtimes</span><span class="p">(</span><span class="n">db_name</span><span class="p">)</span>

    <span class="c"># Getting the execution time for the steps</span>
    <span class="n">plink_exclude_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">shapeit_check_1_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">plink_flip_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">shapeit_check_2_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">plink_final_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">shapeit_phase_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">impute2_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">merge_impute2_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">bgzip_exec_time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
        <span class="c"># Getting the time for &#39;plink_exclude&#39;</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;plink_exclude_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">plink_exclude_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

        <span class="c"># Getting the time for &#39;shapeit_check_1&#39;</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;shapeit_check_chr{}_1&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">shapeit_check_1_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

        <span class="c"># Getting the time for &#39;plink_flip&#39;</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;plink_flip_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">plink_flip_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

        <span class="c"># Getting the time for &#39;shapeit_check_2&#39;</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;shapeit_check_chr{}_2&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">shapeit_check_2_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

        <span class="c"># Getting the time for &#39;plink_final_exclude&#39;</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;plink_final_exclude_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">plink_final_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

        <span class="c"># Getting the time for &#39;shapeit_phase&#39;</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;shapeit_phase_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">shapeit_phase_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

        <span class="c"># Getting the execution times for the imputation step</span>
        <span class="n">chr_imputation_tasks</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">exec_time</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;impute2_chr{}_&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>
        <span class="p">]</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">exec_time</span><span class="p">[</span><span class="n">task_name</span><span class="p">]</span> <span class="k">for</span> <span class="n">task_name</span> <span class="ow">in</span> <span class="n">chr_imputation_tasks</span>
        <span class="p">]</span>
        <span class="n">impute2_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span>
            <span class="n">chrom</span><span class="p">,</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">chr_imputation_tasks</span><span class="p">),</span>
            <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">seconds</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">seconds</span><span class="p">),</span> <span class="mi">0</span><span class="p">)),</span>
            <span class="nb">max</span><span class="p">(</span><span class="n">seconds</span><span class="p">),</span>
        <span class="p">])</span>

        <span class="c"># Getting the time for &#39;merge_impute2&#39;</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;merge_impute2_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">)]</span>
        <span class="n">merge_impute2_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

        <span class="c"># Getting the time for &#39;bgzip&#39; if required</span>
        <span class="n">seconds</span> <span class="o">=</span> <span class="n">exec_time</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;bgzip_chr{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span> <span class="bp">None</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">seconds</span><span class="p">:</span>
            <span class="n">bgzip_exec_time</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="n">seconds</span><span class="p">])</span>

    <span class="c"># Getting the execution time for the second step (plink missing)</span>
    <span class="n">plink_missing_exec_time</span> <span class="o">=</span> <span class="n">exec_time</span><span class="p">[</span><span class="s">&quot;plink_missing_rate&quot;</span><span class="p">]</span>

    <span class="c"># Returning the time</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="s">&quot;plink_exclude_exec_time&quot;</span><span class="p">:</span>   <span class="n">plink_exclude_exec_time</span><span class="p">,</span>
        <span class="s">&quot;shapeit_check_1_exec_time&quot;</span><span class="p">:</span> <span class="n">shapeit_check_1_exec_time</span><span class="p">,</span>
        <span class="s">&quot;plink_missing_exec_time&quot;</span><span class="p">:</span>   <span class="n">plink_missing_exec_time</span><span class="p">,</span>
        <span class="s">&quot;plink_flip_exec_time&quot;</span><span class="p">:</span>      <span class="n">plink_flip_exec_time</span><span class="p">,</span>
        <span class="s">&quot;shapeit_check_2_exec_time&quot;</span><span class="p">:</span> <span class="n">shapeit_check_2_exec_time</span><span class="p">,</span>
        <span class="s">&quot;plink_final_exec_time&quot;</span><span class="p">:</span>     <span class="n">plink_final_exec_time</span><span class="p">,</span>
        <span class="s">&quot;shapeit_phase_exec_time&quot;</span><span class="p">:</span>   <span class="n">shapeit_phase_exec_time</span><span class="p">,</span>
        <span class="s">&quot;merge_impute2_exec_time&quot;</span><span class="p">:</span>   <span class="n">merge_impute2_exec_time</span><span class="p">,</span>
        <span class="s">&quot;impute2_exec_time&quot;</span><span class="p">:</span>         <span class="n">impute2_exec_time</span><span class="p">,</span>
        <span class="s">&quot;bgzip_exec_time&quot;</span><span class="p">:</span>           <span class="n">bgzip_exec_time</span><span class="p">,</span>
    <span class="p">}</span>


<span class="k">def</span> <span class="nf">read_preamble</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span></div>
<div class="viewcode-block" id="read_preamble"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.read_preamble">[docs]</a>    <span class="sd">&quot;&quot;&quot;Reads the preamble file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename (str): the name of the preamble file</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the preamble</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="s">&quot;&quot;</span>

    <span class="n">preamble</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="n">preamble</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

    <span class="k">while</span> <span class="ow">not</span> <span class="n">preamble</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n\n</span><span class="s">&quot;</span><span class="p">):</span>
        <span class="n">preamble</span> <span class="o">+=</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">preamble</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">):</span>
        <span class="n">preamble</span> <span class="o">=</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">+</span> <span class="n">preamble</span>

    <span class="k">return</span> <span class="n">preamble</span>


<span class="k">def</span> <span class="nf">get_shapeit_version</span><span class="p">(</span><span class="n">binary</span><span class="p">):</span></div>
<div class="viewcode-block" id="get_shapeit_version"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.get_shapeit_version">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gets the SHAPEIT version from the binary.</span>

<span class="sd">    Args:</span>
<span class="sd">        binary (str): the name of the SHAPEIT binary</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the version of the SHAPEIT software</span>

<span class="sd">    This function uses :py:class:`subprocess.Popen` to gather the version of</span>
<span class="sd">    the SHAPEIT binary.</span>

<span class="sd">    Warning</span>
<span class="sd">    -------</span>
<span class="sd">        This function only works as long as the version is returned as</span>
<span class="sd">        ``Version: NNN`` (where ``NNN`` is the version), since we use regular</span>
<span class="sd">        expression to extract the version number from the standard output of</span>
<span class="sd">        the software.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Running the command</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">binary</span><span class="p">,</span> <span class="s">&quot;--version&quot;</span><span class="p">]</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">Popen</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">)</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">communicate</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>

    <span class="c"># Finding the version</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&quot;Version : ([\S]+)&quot;</span><span class="p">,</span> <span class="n">output</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">version</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;unknown&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">version</span> <span class="o">=</span> <span class="n">version</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Will be using SHAPEIT version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">version</span>


<span class="k">def</span> <span class="nf">get_impute2_version</span><span class="p">(</span><span class="n">binary</span><span class="p">):</span></div>
<div class="viewcode-block" id="get_impute2_version"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.get_impute2_version">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gets the IMPUTE2 version from the binary.</span>

<span class="sd">    Args:</span>
<span class="sd">        binary (str): the name of the IMPUTE2 binary</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the version of the IMPUTE2 software</span>

<span class="sd">    This function uses :py:class:`subprocess.Popen` to gather the version of</span>
<span class="sd">    the IMPUTE2 binary. Since executing the software to gather the version</span>
<span class="sd">    creates output files, they are deleted.</span>

<span class="sd">    Warning</span>
<span class="sd">    -------</span>
<span class="sd">        This function only works as long as the version is returned as</span>
<span class="sd">        ``IMPUTE version NNN`` (where ``NNN`` is the version), since we use</span>
<span class="sd">        regular expression to extract the version number from the standard</span>
<span class="sd">        output of the software.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Running the command</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">binary</span><span class="p">]</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">Popen</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">)</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">communicate</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>

    <span class="c"># Deleting the output files automatically created by IMPUTE2</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&quot;test.impute2_summary&quot;</span><span class="p">,</span> <span class="s">&quot;test.impute2_warnings&quot;</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

    <span class="c"># Finding the version</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&quot;IMPUTE version ([\S]+)&quot;</span><span class="p">,</span> <span class="n">output</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">version</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;unknown&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">version</span> <span class="o">=</span> <span class="n">version</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Will be using IMPUTE2 version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">version</span>


<span class="k">def</span> <span class="nf">get_plink_version</span><span class="p">(</span><span class="n">binary</span><span class="p">):</span></div>
<div class="viewcode-block" id="get_plink_version"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.get_plink_version">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gets the Plink version from the binary.</span>

<span class="sd">    Args:</span>
<span class="sd">        binary (str): the name of the Plink binary</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the version of the Plink software</span>

<span class="sd">    This function uses :py:class:`subprocess.Popen` to gather the version of</span>
<span class="sd">    the Plink binary. Since executing the software to gather the version</span>
<span class="sd">    creates an output file, it is deleted.</span>

<span class="sd">    Warning</span>
<span class="sd">    -------</span>
<span class="sd">        This function only works as long as the version is returned as</span>
<span class="sd">        ``| PLINK! | NNN |`` (where, ``NNN`` is the version), since we use</span>
<span class="sd">        regular expresion to extract the version number from the standard</span>
<span class="sd">        output of the software.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Running the command</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">binary</span><span class="p">,</span> <span class="s">&quot;--noweb&quot;</span><span class="p">]</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">Popen</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">PIPE</span><span class="p">)</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">communicate</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>

    <span class="c"># Deleting the output file automatically created by Plink</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s">&quot;plink.log&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s">&quot;plink.log&quot;</span><span class="p">)</span>

    <span class="c"># Finding the version</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&quot;\|\s+PLINK!\s+\|\s+(\S+)\s+\|&quot;</span><span class="p">,</span> <span class="n">output</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">version</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">version</span> <span class="o">=</span> <span class="s">&quot;unknown&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">version</span> <span class="o">=</span> <span class="n">version</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Will be using Plink version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">version</span>


<span class="k">def</span> <span class="nf">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">):</span></div>
<div class="viewcode-block" id="check_args"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.check_args">[docs]</a>    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    Args:</span>
<span class="sd">        args (argparse.Namespace): the arguments and options</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: `True` if everything is OK</span>

<span class="sd">    If an option is invalid, a :py:class:`gwip.error.ProgramError` is raised.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Checking the presence of the BED, BIM and BAM files</span>
    <span class="k">for</span> <span class="n">suffix</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&quot;.bed&quot;</span><span class="p">,</span> <span class="s">&quot;.bim&quot;</span><span class="p">,</span> <span class="s">&quot;.fam&quot;</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bfile</span> <span class="o">+</span> <span class="n">suffix</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bfile</span> <span class="o">+</span> <span class="n">suffix</span><span class="p">))</span>

    <span class="c"># Checking the thread</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">thread</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;thread should be one or more&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">shapeit_thread</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;thread should be one or more&quot;</span><span class="p">)</span>

    <span class="c"># Checking IMPUTE2&#39;s files</span>
    <span class="k">for</span> <span class="n">template</span> <span class="ow">in</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">hap_template</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">legend_template</span><span class="p">,</span>
                     <span class="n">args</span><span class="o">.</span><span class="n">map_template</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
            <span class="c"># Checking the haplotype file</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">template</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">sample_file</span><span class="p">):</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">sample_file</span><span class="p">))</span>

    <span class="c"># Checking if bgzip is installed, if asking for compression</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">bgzip</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">which</span><span class="p">(</span><span class="s">&quot;bgzip&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;bgzip: no installed&quot;</span><span class="p">)</span>

    <span class="c"># Checking the SHAPEIT binary if required</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">shapeit_bin</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">shapeit_bin</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">shapeit_bin</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">which</span><span class="p">(</span><span class="s">&quot;shapeit&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;shapeit: not in the path (use --shapeit-bin)&quot;</span><span class="p">)</span>

    <span class="c"># Checking the IMPUTE2 binary if required</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">impute2_bin</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">impute2_bin</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">impute2_bin</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">which</span><span class="p">(</span><span class="s">&quot;impute2&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;impute2: not in the path (use --impute2-bin)&quot;</span><span class="p">)</span>

    <span class="c"># Checking that Plink is in the path</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">plink_bin</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">plink_bin</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">plink_bin</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">which</span><span class="p">(</span><span class="s">&quot;plink&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;plink: not in the path (use --plink-bin)&quot;</span><span class="p">)</span>

    <span class="c"># Checking the segment length</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">segment_length</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: invalid segment &quot;</span>
                           <span class="s">&quot;length&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">segment_length</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">segment_length</span> <span class="o">&lt;</span> <span class="mf">1e3</span><span class="p">:</span>
        <span class="c"># This is too small.. We continue with a warning</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;segment length ({:g} bp) is too &quot;</span>
                        <span class="s">&quot;small&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">segment_length</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">segment_length</span> <span class="o">&gt;</span> <span class="mf">5e6</span><span class="p">:</span>
        <span class="c"># This is too big... We continue with a warning</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;segment length ({:g} bp) is more than &quot;</span>
                        <span class="s">&quot;5Mb&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">segment_length</span><span class="p">))</span>

    <span class="c"># Checking the preamble file (if required)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">preamble</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">preamble</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">preamble</span><span class="p">))</span>

    <span class="c"># Checking the DRMAA configuration file</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">use_drmaa</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">drmaa_config</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;DRMAA configuration file was not provided &quot;</span>
                               <span class="s">&quot;(--drmaa-config), but DRMAA is used &quot;</span>
                               <span class="s">&quot;(--use-drmaa)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">drmaa_config</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">drmaa_config</span><span class="p">))</span>

    <span class="c"># Checking the reference file (if required)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">reference</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">HAS_PYFAIDX</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;pyfaidx is not installed, can not perform &quot;</span>
                            <span class="s">&quot;initial strand check&quot;</span><span class="p">)</span>
            <span class="n">args</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">reference</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">reference</span><span class="p">))</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">reference</span> <span class="o">+</span> <span class="s">&quot;.fai&quot;</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: should be indexed using &quot;</span>
                                   <span class="s">&quot;FAIDX&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">reference</span><span class="p">))</span>

    <span class="k">return</span> <span class="bp">True</span>


<span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">):</span></div>
<div class="viewcode-block" id="parse_args"><a class="viewcode-back" href="../../module_content/gwip.html#gwip.pipeline.parse_args">[docs]</a>    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        parser (argparse.ArgumentParser): the parser object</span>

<span class="sd">    Returns:</span>
<span class="sd">        argparse.Namespace: the parsed options and arguments</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;-v&quot;</span><span class="p">,</span>
        <span class="s">&quot;--version&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;version&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%(prog)s</span><span class="s"> {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--debug&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;set the logging level to debug&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--thread&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;number of threads [</span><span class="si">%(default)d</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The input files</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Input Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--bfile&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;PREFIX&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The prefix of the binary pedfiles (input data).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--reference&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The human reference to perform an initial strand check (useful &quot;</span>
             <span class="s">&quot;for genotyped markers not in the IMPUTE2 reference files) &quot;</span>
             <span class="s">&quot;(optional).&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The output options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Output Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--output-dir&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;DIR&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;gwip&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s">&quot;out_dir&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The name of the output directory. [</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--bgzip&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Use bgzip to compress the impute2 files.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The HPC options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;HPC Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--use-drmaa&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Launch tasks using DRMAA.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--drmaa-config&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The configuration file for tasks (use this option when &quot;</span>
             <span class="s">&quot;launching tasks using DRMAA). This file should describe the &quot;</span>
             <span class="s">&quot;walltime and the number of nodes/processors to use for each &quot;</span>
             <span class="s">&quot;task.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--preamble&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;This option should be used when using DRMAA on a HPC to load &quot;</span>
             <span class="s">&quot;required module and set environment variables. The content of &quot;</span>
             <span class="s">&quot;the file will be added between the &#39;shebang&#39; line and the tool &quot;</span>
             <span class="s">&quot;command.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The SHAPEIT software options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;SHAPEIT Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--shapeit-bin&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;BINARY&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The SHAPEIT binary if it&#39;s not in the path.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--shapeit-thread&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;INT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The number of thread for phasing. [</span><span class="si">%(default)d</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The Plink option</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Plink Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--plink-bin&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;BINARY&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The Plink binary if it&#39;s not in the path.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The IMPUTE2 software options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;IMPUTE2 Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--impute2-bin&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;BINARY&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The IMPUTE2 binary if it&#39;s not in the path.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--segment-length&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;BP&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">5e6</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The length of a single segment for imputation. [</span><span class="si">%(default).1g</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--hap-template&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;TEMPLATE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The template for IMPUTE2&#39;s haplotype files (replace the &quot;</span>
             <span class="s">&quot;chromosome number by &#39;{chrom}&#39;, e.g. &quot;</span>
             <span class="s">&quot;&#39;1000GP_Phase3_chr{chrom}.hap.gz&#39;).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--legend-template&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;TEMPLATE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The template for IMPUTE2&#39;s legend files (replace the chromosome &quot;</span>
             <span class="s">&quot;number by &#39;{chrom}&#39;, e.g. &quot;</span>
             <span class="s">&quot;&#39;1000GP_Phase3_chr{chrom}.legend.gz&#39;).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--map-template&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;TEMPLATE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The template for IMPUTE2&#39;s map files (replace the chromosome &quot;</span>
             <span class="s">&quot;number by &#39;{chrom}&#39;, e.g. &quot;</span>
             <span class="s">&quot;&#39;genetic_map_chr{chrom}_combined_b37.txt&#39;).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--sample-file&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The name of IMPUTE2&#39;s sample file.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--filtering-rules&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;RULE&quot;</span><span class="p">,</span>
        <span class="n">nargs</span><span class="o">=</span><span class="s">&quot;+&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;IMPUTE2 filtering rules (optional).&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The impute2 file merger options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;IMPUTE2 Merger Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--probability&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The probability threshold for no calls. [&lt;</span><span class="si">%(default).1f</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--completion&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.98</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The completion rate threshold for site exclusion. &quot;</span>
             <span class="s">&quot;[&lt;</span><span class="si">%(default).2f</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The automatic report options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Automatic Report Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--report-number&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NB&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;GWIP automatic report&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The report number. [</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--report-title&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;TITLE&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;GWIP: Automatic genome-wide imputation&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The report title. [</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--report-author&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;AUTHOR&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;Automatically generated by GWIP&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The report author. [</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span></div>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">gwip 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../gwip.html" >gwip</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>