

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Pipeline Tutorial &mdash; genipe 1.2.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/bizstyle.js"></script>
    <link rel="top" title="genipe 1.2.1 documentation" href="../index.html" />
    <link rel="up" title="Short Tutorials" href="../tutorials.html" />
    <link rel="next" title="Survival Tutorial" href="tutorial_cox.html" />
    <link rel="prev" title="Short Tutorials" href="../tutorials.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_cox.html" title="Survival Tutorial"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../tutorials.html" title="Short Tutorials"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.2.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" accesskey="U">Short Tutorials</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Pipeline Tutorial</a><ul>
<li><a class="reference internal" href="#quick-navigation">Quick navigation</a></li>
<li><a class="reference internal" href="#genome-wide-imputation-pipeline">Genome-wide imputation pipeline</a><ul>
<li><a class="reference internal" href="#required-softwares">Required softwares</a></li>
<li><a class="reference internal" href="#input-files">Input files</a><ul>
<li><a class="reference internal" href="#genotypes">Genotypes</a></li>
<li><a class="reference internal" href="#reference-panels">Reference panels</a></li>
<li><a class="reference internal" href="#human-reference-optional">Human reference (optional)</a></li>
<li><a class="reference internal" href="#drmaa-configuration-optional">DRMAA configuration (optional)</a></li>
<li><a class="reference internal" href="#preamble-optional">Preamble (optional)</a></li>
<li><a class="reference internal" href="#input-file-summary">Input file summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#executing-the-pipeline">Executing the pipeline</a></li>
<li><a class="reference internal" href="#compiling-the-report">Compiling the report</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#genipe-directory"><code class="docutils literal"><span class="pre">genipe</span></code> directory</a></li>
<li><a class="reference internal" href="#genipe-chrn-directories"><code class="docutils literal"><span class="pre">genipe/chrN</span></code> directories</a></li>
<li><a class="reference internal" href="#genipe-chrn-final-impute2-directories"><code class="docutils literal"><span class="pre">genipe/chrN/final_impute2</span></code> directories</a></li>
<li><a class="reference internal" href="#genipe-missing-directory"><code class="docutils literal"><span class="pre">genipe/missing</span></code> directory</a></li>
<li><a class="reference internal" href="#genipe-report-directory"><code class="docutils literal"><span class="pre">genipe/report</span></code> directory</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../tutorials.html"
                        title="previous chapter">Short Tutorials</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_cox.html"
                        title="next chapter">Survival Tutorial</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/tutorial_genipe.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pipeline-tutorial">
<span id="genipe-tut-page"></span><h1>Pipeline Tutorial<a class="headerlink" href="#pipeline-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="quick-navigation">
<h2>Quick navigation<a class="headerlink" href="#quick-navigation" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><a class="reference internal" href="#genipe-tut-softwares"><span>Required softwares</span></a></li>
<li><a class="reference internal" href="#genipe-tut-input-files"><span>Input files</span></a></li>
<li><a class="reference internal" href="#genipe-tut-execute"><span>Executing the pipeline</span></a></li>
<li><a class="reference internal" href="#genipe-tut-compile-report"><span>Compiling the report</span></a></li>
<li><a class="reference internal" href="#genipe-tut-output-files"><span>Output files</span></a></li>
</ol>
</div>
<div class="section" id="genome-wide-imputation-pipeline">
<h2>Genome-wide imputation pipeline<a class="headerlink" href="#genome-wide-imputation-pipeline" title="Permalink to this headline">¶</a></h2>
<p>High-throughput genotyping platforms can assess up to five million variations
in thousands of samples. Depending on the genotyped data and the reference
panel used, genome-wide imputation tools can infer genotypes for more than 38
million variants (single nucleotides, insertions and deletions). Unfortunately,
genome-wide imputation requires high computation power and multiple data
processing steps.</p>
<p>The <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> pipeline automates the different steps for pre-phasing and
imputation for genome-wide data. The pipeline follows the guideline described
by IMPUTE2&#8217;s best practices when analyzing genome-wide data (described
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#prephasing">by IMPUTE2</a>
and <a class="reference external" href="http://www.shapeit.fr/pages/m03_phasing/imputation.html">by SHAPEIT</a>).</p>
<p>First, we will create a project directory where all the analysis will be
performed:</p>
<div class="highlight-bash"><div class="highlight"><pre>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial
</pre></div>
</div>
<div class="section" id="required-softwares">
<span id="genipe-tut-softwares"></span><h3>Required softwares<a class="headerlink" href="#required-softwares" title="Permalink to this headline">¶</a></h3>
<p>The main <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> pipeline requires three external tools:
<a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/">Plink</a>,
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html">IMPUTE2</a> and
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html">SHAPEIT</a>.
These tools are not required to be located in your <code class="docutils literal"><span class="pre">PATH</span></code> variable, since you
can specify each of their location at runtime.</p>
<p>If some of the binaries are missing in you <code class="docutils literal"><span class="pre">PATH</span></code> variable, we will create a
directory where the missing binaries will go:</p>
<div class="highlight-bash"><div class="highlight"><pre>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial/bin
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/bin
</pre></div>
</div>
<p>Plink is freely available. Go to
<a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/download.shtml">Plink download page</a>,
read the license, and download the most recent version. Extract the binary and
copy it to the <code class="docutils literal"><span class="pre">bin</span></code> directory previously created.</p>
<p>IMPUTE2 is freely available for academic use. Go to
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#download">IMPUTE2 download page</a>,
read the license, and download the most recent version. Extract the binary and
copy it to the <code class="docutils literal"><span class="pre">bin</span></code> directory previously created.</p>
<p>SHAPEIT is freely available for academic use. Go to
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#download">SHAPEIT download page</a>,
read the license, and download the most recent version. Extract the binary and
copy it to the <code class="docutils literal"><span class="pre">bin</span></code> directory previously created.</p>
<p>The <code class="docutils literal"><span class="pre">bin</span></code> directory should now contain the missing binaries: <code class="docutils literal"><span class="pre">impute2</span></code>,
<code class="docutils literal"><span class="pre">plink</span></code> and/or <code class="docutils literal"><span class="pre">shapeit</span></code>.</p>
</div>
<div class="section" id="input-files">
<span id="genipe-tut-input-files"></span><h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h3>
<div class="section" id="genotypes">
<h4>Genotypes<a class="headerlink" href="#genotypes" title="Permalink to this headline">¶</a></h4>
<p>The genotypes of the study cohort have to be in Plink&#8217;s binary pedfile format.
This format allows for time and disk space optimization. Each dataset is
comprised of three files: the <code class="docutils literal"><span class="pre">.bed</span></code> (binary file, genotype information),
<code class="docutils literal"><span class="pre">.fam</span></code> (text file, sample information) and <code class="docutils literal"><span class="pre">.bim</span></code> (text file, marker
information) files. Those file are generated by Plink.</p>
<p>Genotype data are provided for this tutorial. They are a subset from the files
provided by Plink on its
<a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/res.shtml">resources page</a>.
The build was lifted over from <em>GRCh36</em> to <em>GRCh37</em> using
<a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">USCS liftOver</a>. Only CEU samples
were kept, along with markers having a completion rate higher or equal to 0.98.
To download the data for this tutorial, execute the following command:</p>
<div class="highlight-bash"><div class="highlight"><pre>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial/data
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/data

wget http://pgxcentre.github.io/genipe/_static/tutorial/hapmap_CEU_r23a_hg19.tar.bz2
tar -jxf hapmap_CEU_r23a_hg19.tar.bz2
rm hapmap_CEU_r23a_hg19.tar.bz2
</pre></div>
</div>
</div>
<div class="section" id="reference-panels">
<h4>Reference panels<a class="headerlink" href="#reference-panels" title="Permalink to this headline">¶</a></h4>
<p>IMPUTE2 can use publicly available reference datasets. They provide such
dataset on their website. Go to IMPUTE2&#8217;s
<a class="reference external" href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#reference">reference page</a>,
and download the most recent reference data (which is over 12Gb). Once the
reference is downloaded, extract it in the working directory
(<code class="docutils literal"><span class="pre">$HOME/genipe_tutorial</span></code>).</p>
</div>
<div class="section" id="human-reference-optional">
<h4>Human reference (optional)<a class="headerlink" href="#human-reference-optional" title="Permalink to this headline">¶</a></h4>
<p>The pipeline include a step to check for strand alignment with the reference
panel (using <em>SHAPEIT</em>). The drawback of this method is that it is impossible
to verify the strand of markers which are absent from the <em>IMPUTE2</em>&#8216;s
reference. We have introduce a way to check the strand using the reference
genome (in <em>fasta</em> format, indexed using <em>faidx</em>).</p>
<p>It is possible to download the human reference (<em>fasta</em> format) from
<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/">UCSC&#8217;s download page</a>
(the file named <code class="docutils literal"><span class="pre">chromFa.tar.gz</span></code>).</p>
<div class="highlight-bash"><div class="highlight"><pre>mkdir -p <span class="nv">$HOME</span>/genipe_tutorial/hg19
<span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/hg19

wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz
tar -zxf chromFa.tar.gz
<span class="k">for</span> chrom in <span class="k">$(</span>seq <span class="m">1</span> 22<span class="k">)</span> X Y M<span class="p">;</span> <span class="k">do</span> cat chr<span class="si">${</span><span class="nv">chrom</span><span class="si">}</span>.fa<span class="p">;</span> <span class="k">done</span> &gt; hg19.fasta
rm -f *.fa chromFa.tar.gz
</pre></div>
</div>
<p>To index the <em>fasta</em> file, <em>samtools</em> is required at
<a class="reference external" href="http://www.htslib.org/download/">htslib&#8217;s download page</a>. Once <em>samtools</em> is
installed, perform the following command:</p>
<div class="highlight-bash"><div class="highlight"><pre>samtools faidx hg19.fasta
</pre></div>
</div>
<p>You should now have two files in the <code class="docutils literal"><span class="pre">$HOME/genipe_tutorial/hg19</span></code> directory:
<code class="docutils literal"><span class="pre">hg19.fasta</span></code> and <code class="docutils literal"><span class="pre">hg19.fasta.fai</span></code>.</p>
</div>
<div class="section" id="drmaa-configuration-optional">
<h4>DRMAA configuration (optional)<a class="headerlink" href="#drmaa-configuration-optional" title="Permalink to this headline">¶</a></h4>
<p>If the pipeline is to be launch on a computing server, the <code class="docutils literal"><span class="pre">--use-drmaa</span></code>
option should be used. This will launch each step on the server using the DRMAA
api. On some cluster, supplemental information is required for each task
(<em>i.e.</em> execution time, number of nodes/processes to reserve). This
parametrization is done using a configuration (<em>ini</em>) file, describing these
parameters for each step.</p>
<p>When providing an empty <em>ini</em> file, the default walltime and number of
nodes/processes will be 15 minutes and 1/1, respectively. Otherwise, different
parameters can be used for each step. For example, the following configuration
will increase the walltime for all phasing tasks from 15 minutes to 3 hours. It
will also run each phasing tasks on one node using 12 processes.</p>
<div class="highlight-ini"><div class="highlight"><pre><span class="k">[shapeit_phase]</span>
<span class="na">walltime</span> <span class="o">=</span> <span class="s">03:00:00</span>
<span class="na">nodes</span>    <span class="o">=</span> <span class="s">1</span>
<span class="na">ppn</span>      <span class="o">=</span> <span class="s">12</span>
</pre></div>
</div>
<p>The following example has the same configuration as the previous example, but
will increase the walltime for chromosome 2 to 4 hours, with 1 node and 24
processes.</p>
<div class="highlight-ini"><div class="highlight"><pre><span class="k">[shapeit_phase]</span>
<span class="na">walltime</span> <span class="o">=</span> <span class="s">03:00:00</span>
<span class="na">nodes</span>    <span class="o">=</span> <span class="s">1</span>
<span class="na">ppn</span>      <span class="o">=</span> <span class="s">12</span>

<span class="na">chr2_walltime</span> <span class="o">=</span> <span class="s">04:00:00</span>
<span class="na">chr2_nodes</span>    <span class="o">=</span> <span class="s">1</span>
<span class="na">chr2_ppn</span>      <span class="o">=</span> <span class="s">24</span>
</pre></div>
</div>
<p>Since imputation is performed on segments for each chromosome, it is possible
to modify the parameters for a single segment. This is usefull when a segment
doesn&#8217;t have time to finish and its imputation requires a rerun. For example,
the following parameters will increase the walltime from 15 minutes to 3.5
hours for segment 10,000,001-15,000,000 on chromosome 1. Also, all segments
located on chromosome 2 will have a walltime of 4 hours.</p>
<div class="highlight-ini"><div class="highlight"><pre><span class="k">[impute2]</span>
<span class="na">chr1_10000001_15000000_walltime</span> <span class="o">=</span> <span class="s">03:30:00</span>

<span class="na">chr2_walltime</span> <span class="o">=</span> <span class="s">04:00:00</span>
</pre></div>
</div>
<p>We provide a
<a class="reference external" href="http://pgxcentre.github.io/genipe/_static/tutorial/config_example.ini">configuration example</a>
including all possible section. Also, here is a list of all possible section
(<em>i.e</em> pipeline step) that can be parametrized.</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">plink_exclude</span></code></li>
<li><code class="docutils literal"><span class="pre">plink_missing_rate</span></code></li>
<li><code class="docutils literal"><span class="pre">shapeit_check_1</span></code></li>
<li><code class="docutils literal"><span class="pre">plink_flip</span></code></li>
<li><code class="docutils literal"><span class="pre">shapeit_check_2</span></code></li>
<li><code class="docutils literal"><span class="pre">plink_final_exclude</span></code></li>
<li><code class="docutils literal"><span class="pre">shapeit_phase</span></code></li>
<li><code class="docutils literal"><span class="pre">impute2</span></code></li>
<li><code class="docutils literal"><span class="pre">merge_impute2</span></code></li>
<li><code class="docutils literal"><span class="pre">bgzip</span></code></li>
</ul>
<p>Some cluster doesn&#8217;t require any configuration at all. To skip configuration,
use the <code class="docutils literal"><span class="pre">main</span></code> section of the <em>ini</em> file as such:</p>
<div class="highlight-ini"><div class="highlight"><pre><span class="k">[main]</span>
<span class="na">skip_drmaa_config</span> <span class="o">=</span> <span class="s">yes</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Keep in mind that lines starting with a <code class="docutils literal"><span class="pre">#</span></code> are comments and are not used
in the DRMAA configuration. This is useful to describe what parameters are
used for each step.</p>
</div>
</div>
<div class="section" id="preamble-optional">
<h4>Preamble (optional)<a class="headerlink" href="#preamble-optional" title="Permalink to this headline">¶</a></h4>
<p>When using the <code class="docutils literal"><span class="pre">--use-drmaa</span></code> option, the pipeline creates <em>bash</em> script that
are launched on the computing cluster. Some clusters require module to be
loaded and the python virtual environment to be loaded before executing a
script. This is done using the preamble file (the <code class="docutils literal"><span class="pre">--preamble</span></code> option).</p>
<p>The content of the file will be added between the first line of the temporary
<em>bash</em> script (the <em>shebang</em>) and the actual command. For example, the
following file will load the gcc module (version 4.8.2) and the python virtual
environment before launching the task.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># Loading the required module</span>
module load gcc/4.8.2

<span class="c"># The python virtual environment</span>
<span class="nb">source</span> <span class="nv">$HOME</span>/softwares/python_env/bin/activate
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The preamble file is system dependent, but you should always at least
activate the virtual python environment so that the tools provided by
<a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> are automatically in the system path.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The preamble will be added <strong>as-is</strong> in the <em>bash</em> script that will be
executed. Hence, always be careful of what is included in the preamble.</p>
</div>
</div>
<div class="section" id="input-file-summary">
<h4>Input file summary<a class="headerlink" href="#input-file-summary" title="Permalink to this headline">¶</a></h4>
<p>You should have the following directory structure:</p>
<div class="highlight-text"><div class="highlight"><pre>$HOME/genipe_tutorial/
│
├── 1000GP_Phase3/
│   ├── 1000GP_Phase3_chr1.hap.gz
│   ├── 1000GP_Phase3_chr2.hap.gz
│   ├── ...
│   ├── 1000GP_Phase3_chr1.legend.gz
│   ├── 1000GP_Phase3_chr2.legend.gz
│   ├── ...
│   ├── 1000GP_Phase3.sample
│   ├── genetic_map_chr1_combined_b37.txt
│   ├── genetic_map_chr2_combined_b37.txt
│   └── ...
│
├── bin/
│   ├── impute2
│   ├── plink
│   └── shapeit
│
├── data/
│   ├── hapmap_CEU_r23a_hg19.bed
│   ├── hapmap_CEU_r23a_hg19.bim
│   └── hapmap_CEU_r23a_hg19.fam
│
├── genipe_config.ini  # OPTIONAL (--use-drmaa, --drmaa-config)
│
├── hg19/
│   ├── hg19.fasta
│   └── hg19.fasta.fai
│
└── preamble.txt     # OPTIONAL (--use-drmaa, --preamble)
</pre></div>
</div>
</div>
</div>
<div class="section" id="executing-the-pipeline">
<span id="genipe-tut-execute"></span><h3>Executing the pipeline<a class="headerlink" href="#executing-the-pipeline" title="Permalink to this headline">¶</a></h3>
<p>Once all the input files are ready for analysis, you can finally execute the
pipeline. Make sure that the virtual Python environment was properly activated
(see <a class="reference internal" href="../installation.html#genipe-pyvenv-activation"><span>Virtual environment activation</span></a> for more details).</p>
<p>When in the correct working directory, the following command should execute the
genome-wide imputation of the <em>HapMap</em> CEU dataset.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial

genipe-launcher <span class="se">\</span>
    --bfile data/hapmap_CEU_r23a_hg19 <span class="se">\</span>
    --shapeit-bin bin/shapeit <span class="se">\</span>
    --impute2-bin bin/impute2 <span class="se">\</span>
    --plink-bin bin/plink <span class="se">\</span>
    --reference hg19/hg19.fasta <span class="se">\</span>
    --hap-template 1000GP_Phase3/1000GP_Phase3_chr<span class="o">{</span>chrom<span class="o">}</span>.hap.gz <span class="se">\</span>
    --legend-template 1000GP_Phase3/1000GP_Phase3_chr<span class="o">{</span>chrom<span class="o">}</span>.legend.gz <span class="se">\</span>
    --map-template 1000GP_Phase3/genetic_map_chr<span class="o">{</span>chrom<span class="o">}</span>_combined_b37.txt <span class="se">\</span>
    --sample-file 1000GP_Phase3/1000GP_Phase3.sample <span class="se">\</span>
    --filtering-rules <span class="s1">&#39;ALL&lt;0.01&#39;</span> <span class="s1">&#39;ALL&gt;0.99&#39;</span> <span class="se">\</span>
    --bgzip <span class="se">\</span>
    --thread <span class="m">4</span> <span class="se">\</span>
    --report-title <span class="s2">&quot;Tutorial&quot;</span> <span class="se">\</span>
    --report-number <span class="s2">&quot;Test Report&quot;</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>In the previous command, the <code class="docutils literal"><span class="pre">--reference</span></code> and <code class="docutils literal"><span class="pre">--bgzip</span></code> options are
optional and might be skipped.</p>
<p class="last">It is possible to skip (in the previous command) the <code class="docutils literal"><span class="pre">--shapeit-bin</span></code>, the
<code class="docutils literal"><span class="pre">--impute2-bin</span></code> and/or the <code class="docutils literal"><span class="pre">--plink-bin</span></code> options if the SHAPEIT, IMPUTE2
and/or the Plink binaries are in  the <code class="docutils literal"><span class="pre">PATH</span></code> variable.</p>
</div>
<p>The following table describes the option used by <a class="reference internal" href="../module_content/genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> in the
previous command (see the <a class="reference internal" href="../index.html#genipe-usage"><span>Usage</span></a> section for a full list):</p>
<table border="1" class="docutils">
<colgroup>
<col width="32%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--bfile</span></code></td>
<td>The genotypes of the study cohort to be imputed.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--reference</span></code></td>
<td>The <em>fasta</em> file containing the reference genome
for initial strand verification (<em>optional</em>).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--hap-template</span></code></td>
<td>The template for <em>IMPUTE2</em>&#8216;s reference haplotype
files (<code class="docutils literal"><span class="pre">{chrom}</span></code> will be replaced by the
chromosome number).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--legend-template</span></code></td>
<td>The template for <em>IMPUTE2</em>&#8216;s reference legend
files (<code class="docutils literal"><span class="pre">{chrom}</span></code> will be replaced by the
chromosome number).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--map-template</span></code></td>
<td>The template for <em>IMPUTE2</em>&#8216;s reference map
files (<code class="docutils literal"><span class="pre">{chrom}</span></code> will be replaced by the
chromosome number).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--sample-file</span></code></td>
<td>The name of <em>IMPUTE2</em>&#8216;s reference sample file.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--filtering-rules</span></code></td>
<td>Rules used by <em>IMPUTE2</em> to exclude sites from
its reference files (using the legend files).
Each terms are joined using a logical <em>OR</em>.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--bgzip</span></code></td>
<td>The final (merged) <em>IMPUTE2</em> output files will
be compressed using the <code class="docutils literal"><span class="pre">bgzip</span></code> software (the
latter must be in the <code class="docutils literal"><span class="pre">path</span></code>). This will not
compress the segment files.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--thread</span></code></td>
<td>The number of thread to use for the analysis.
When using <em>DRMAA</em>, this will be the number of
simultaneous tasks.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">--report-title</span></code></td>
<td>The title of the automatic report.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">--report-number</span></code></td>
<td>The number of the report (will appear as
sub-title and in the footer of the automatic
report).</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>If the pipeline fails (<em>e.g.</em> not enough memory or the walltime exceeded),
re-running the pipeline (with different number of thread or different
walltime) will only launch the task that were not completed.</p>
<p class="last">The pipeline checks if output files are missing. If an output file is
deleted, the step producing this file will be run again (but not the
subsequent steps).</p>
</div>
</div>
<div class="section" id="compiling-the-report">
<span id="genipe-tut-compile-report"></span><h3>Compiling the report<a class="headerlink" href="#compiling-the-report" title="Permalink to this headline">¶</a></h3>
<p>A report containing useful information (such as quality metrics and execution
time, among others) is automatically generated once the imputation process is
completed. To compile the report, perform the following commands:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/genipe/report

make <span class="o">&amp;&amp;</span> make clean
</pre></div>
</div>
<p>This will generate the following
<a class="reference external" href="http://pgxcentre.github.io/genipe/_static/tutorial/report.pdf">PDF report</a>
(which is named <code class="docutils literal"><span class="pre">report.pdf</span></code>). It is always possible to modify the original
<code class="docutils literal"><span class="pre">report.tex</span></code> file to include analysis specific details (<em>e.g.</em> cohort
description).</p>
</div>
<div class="section" id="output-files">
<span id="genipe-tut-output-files"></span><h3>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h3>
<p>All results will be located in the <code class="docutils literal"><span class="pre">genipe</span></code> directory (or whatever
<code class="docutils literal"><span class="pre">--output-dir</span></code> links to). Here is the directory tree summarizing the output
files.</p>
<div class="highlight-text"><div class="highlight"><pre>genipe/
│
├── chr1/
│   ├── chr1.1_5000000.impute2
│   ├── chr1.1_5000000.impute2_info
│   ├── chr1.1_5000000.impute2_info_by_sample
│   ├── chr1.1_5000000.impute2_summary
│   ├── chr1.1_5000000.impute2_warnings
│   ├── ...
│   ├── chr1.final.bed
│   ├── chr1.final.bim
│   ├── chr1.final.fam
│   ├── chr1.final.log
│   ├── chr1.final.phased.haps
│   ├── chr1.final.phased.ind.me
│   ├── chr1.final.phased.ind.mm
│   ├── chr1.final.phased.log
│   ├── chr1.final.phased.sample
│   ├── chr1.final.phased.snp.me
│   ├── chr1.final.phased.snp.mm
│   ├── ...
│   │
│   └── final_impute2/
│       ├── chr1.imputed.alleles
│       ├── chr1.imputed.completion_rates
│       ├── chr1.imputed.good_sites
│       ├── chr1.imputed.impute2.gz
│       ├── chr1.imputed.impute2_info
│       ├── chr1.imputed.imputed_sites
│       ├── chr1.imputed.log
│       ├── chr1.imputed.maf
│       ├── chr1.imputed.map
│       └── chr1.imputed.sample
│
├── .../
│
├── chromosome_lengths.txt
├── frequency_pie.pdf
├── genipe.log
├── markers_to_exclude.txt
├── markers_to_flip.txt
│
├── missing
│   ├── missing.imiss
│   ├── missing.lmiss
│   └── missing.log
│
├── report
│   ├── frequency_pie.pdf
│   ├── Makefile
│   ├── references.bib
│   ├── references.bst
│   └── report.tex
│
└── tasks.db
</pre></div>
</div>
<div class="section" id="genipe-directory">
<h4><code class="docutils literal"><span class="pre">genipe</span></code> directory<a class="headerlink" href="#genipe-directory" title="Permalink to this headline">¶</a></h4>
<p>This directory contains all the chromosome specific analysis. The specific
directory content is describe below. The following files are created inside the
<code class="docutils literal"><span class="pre">genipe</span></code> directory:</p>
<table border="1" class="docutils">
<colgroup>
<col width="39%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">File</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">chromosome_lengths.txt</span></code></td>
<td>The length of each chromosome (this
information is fetched from Ensembl using
its REST API and saved to file).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">frequency_pie.pdf</span></code></td>
<td>This file contains a pie chart describing
the minor allele frequency distribution of
the imputed markers. This file is generated
only if the <code class="xref py py-mod docutils literal"><span class="pre">matplotlib</span></code> module is
installed.</td>
</tr>
<tr class="row-even"><td colspan="2"><code class="docutils literal"><span class="pre">genipe.log</span></code>               | The log file of the main pipeline.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">markers_to_exclude.txt</span></code></td>
<td>The list of markers to exclude prior to
phasing.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">markers_to_flip.txt</span></code></td>
<td>The list of markers to flip prior to
phasing.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">tasks.db</span></code></td>
<td>The <em>sqlite</em> database containing
information of all tasks (if it&#8217;s
completed, execution time, etc).</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genipe-chrn-directories">
<h4><code class="docutils literal"><span class="pre">genipe/chrN</span></code> directories<a class="headerlink" href="#genipe-chrn-directories" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal"><span class="pre">chrN</span></code> directories contain the intermediate files, created throughout the
pipeline. The most important files in these directories are the log files (for
errors and summary statistics). There will be one directory per autosomal
chromosomes.</p>
</div>
<div class="section" id="genipe-chrn-final-impute2-directories">
<span id="genipe-tut-output-files-final-impute2"></span><h4><code class="docutils literal"><span class="pre">genipe/chrN/final_impute2</span></code> directories<a class="headerlink" href="#genipe-chrn-final-impute2-directories" title="Permalink to this headline">¶</a></h4>
<p>These <code class="docutils literal"><span class="pre">final_impute2</span></code> directories (located in the <code class="docutils literal"><span class="pre">genipe/chrN</span></code>
directories) contain the final output files from the pipeline for each
autosomal chromosomes. They will contain the following files:</p>
<table border="1" class="docutils">
<colgroup>
<col width="43%" />
<col width="57%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Extension</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.imputed.alleles</span></code></td>
<td>Description of the reference and
alternative allele at each sites.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.imputed.completion_rates</span></code></td>
<td>Number of missing values and completion
rate for all sites (using the
probability threshold set by the user,
where the default is higher and equal
to 0.9).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.imputed.good_sites</span></code></td>
<td>List of sites which pass the completion
rate threshold (set by the user, where
the default is higher and equal to 0.98)
using the probability threshold (set by
the user, where the default is higher
and equal to 0.9).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.imputed.impute2</span></code> or
<code class="docutils literal"><span class="pre">.imputed.impute2.gz</span></code></td>
<td>Imputation results (merged from the
individual segment files. This file
might be compress (with the <code class="docutils literal"><span class="pre">.gz</span></code>
extension) if the <code class="docutils literal"><span class="pre">--bgzip</span></code> option was
used when launching the pipeline.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.imputed.impute2_info</span></code></td>
<td>Marker-wise information file with one
line per marker and a single header line
at the begening. It contains, among
others, the information value which is a
measure of the observed statistical
information associated with the allele
frequency estimate.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.imputed.imputed_sites</span></code></td>
<td>List of imputed sites (excluding sites
that were previously genotyped in the
study cohort).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.imputed.log</span></code></td>
<td>The log file of the merging step.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.imputed.maf</span></code></td>
<td>File containing the minor allele
frequency (along with minor allele
identification) for all sites using the
probabilitty threshold of 0.9. When no
genotypes are available (because they
are all below the threshold), the MAF is
<code class="docutils literal"><span class="pre">NA</span></code>.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.imputed.map</span></code></td>
<td>A <em>map</em> file describing the genomic
location of all sites.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.imputed.sample</span></code></td>
<td>The sample file generated by the phasing
step, which describe the sample ordering
in the IMPUTE2 files.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genipe-missing-directory">
<h4><code class="docutils literal"><span class="pre">genipe/missing</span></code> directory<a class="headerlink" href="#genipe-missing-directory" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal"><span class="pre">missing</span></code> directory contains the missing rates for both samples
(<code class="docutils literal"><span class="pre">missing.imiss</span></code>) and genotypes markers (<code class="docutils literal"><span class="pre">missing.lmiss</span></code>). Those files are
generated by Plink.</p>
</div>
<div class="section" id="genipe-report-directory">
<h4><code class="docutils literal"><span class="pre">genipe/report</span></code> directory<a class="headerlink" href="#genipe-report-directory" title="Permalink to this headline">¶</a></h4>
<p>This <code class="docutils literal"><span class="pre">report</span></code> directory contains the automatically generated report, which
provides valuable information about the imputation analysis. Such information
contains cross-validation statistics (as provided by IMPUTE2), frequency
statistics and completion rates according to user defined parameters.</p>
<p>The automatic report is generated in the <code class="docutils literal"><span class="pre">LaTeX</span></code> language (file
<code class="docutils literal"><span class="pre">report.tex</span></code>), and can be compile using the following command (as long as
<code class="docutils literal"><span class="pre">LaTeX</span></code> is installed).</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$HOME</span>/genipe_tutorial/genipe/report
make <span class="o">&amp;&amp;</span> make clean
</pre></div>
</div>
<p>This will generate the following
<a class="reference external" href="http://pgxcentre.github.io/genipe/_static/tutorial/report.pdf">PDF report</a>
(which is named <code class="docutils literal"><span class="pre">report.pdf</span></code>). It is always possible to modify the original
<code class="docutils literal"><span class="pre">report.tex</span></code> file to include analysis specific details (<em>e.g.</em> cohort
description).</p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_cox.html" title="Survival Tutorial"
             >next</a> |</li>
        <li class="right" >
          <a href="../tutorials.html" title="Short Tutorials"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.2.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorials.html" >Short Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>