

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Linear Tutorial &mdash; gwip 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/bizstyle.js"></script>
    <link rel="top" title="gwip 0.1 documentation" href="../index.html" />
    <link rel="up" title="Short Tutorials" href="../tutorial.html" />
    <link rel="next" title="Logistic Tutorial" href="tutorial_logistic.html" />
    <link rel="prev" title="SKAT Tutorial" href="tutorial_SKAT.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_logistic.html" title="Logistic Tutorial"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_SKAT.html" title="SKAT Tutorial"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">gwip 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" accesskey="U">Short Tutorials</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Linear Tutorial</a><ul>
<li><a class="reference internal" href="#quick-navigation">Quick navigation</a></li>
<li><a class="reference internal" href="#linear-regression">Linear regression</a><ul>
<li><a class="reference internal" href="#input-files">Input files</a><ul>
<li><a class="reference internal" href="#imputed-genotypes">Imputed genotypes</a></li>
<li><a class="reference internal" href="#samples-file">Samples file</a></li>
<li><a class="reference internal" href="#phenotype-file">Phenotype file</a></li>
<li><a class="reference internal" href="#sites-to-extract-optional">Sites to extract (optional)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#execution-time">Execution time</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_SKAT.html"
                        title="previous chapter">SKAT Tutorial</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_logistic.html"
                        title="next chapter">Logistic Tutorial</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/tutorial_linear.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="linear-tutorial">
<h1>Linear Tutorial<a class="headerlink" href="#linear-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Work on this tutorial is still in progress...</p>
</div>
<div class="section" id="quick-navigation">
<h2>Quick navigation<a class="headerlink" href="#quick-navigation" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><a class="reference internal" href="#lin-tut-input-files"><span>Input files</span></a></li>
<li><a class="reference internal" href="#lin-tut-usage"><span>Usage</span></a></li>
<li><a class="reference internal" href="#lin-tut-execution-time"><span>Execution time</span></a></li>
</ol>
</div>
<div class="section" id="linear-regression">
<h2>Linear regression<a class="headerlink" href="#linear-regression" title="Permalink to this headline">¶</a></h2>
<p>Linear regressions are commonly used to perform genome-wide association
studies. It is possible to perform such an analysis using imputation data as
dosage, where each imputed genotypes varies between 0 and 2 (inclusively).
Values close to 0 means that a homozygous genotypes of the most frequent allele
is the most probable. Values close to 2 means that a homozygous genotypes of
the rare allele is the most probable. Finally, values close to 1 means that a
heterozygous genotype is the most probable.</p>
<div class="section" id="input-files">
<span id="lin-tut-input-files"></span><h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h3>
<div class="section" id="imputed-genotypes">
<h4>Imputed genotypes<a class="headerlink" href="#imputed-genotypes" title="Permalink to this headline">¶</a></h4>
<p>After running the <a class="reference internal" href="../module_content/gwip.html#module-gwip" title="gwip"><code class="xref py py-mod docutils literal"><span class="pre">gwip</span></code></a> pipeline and has the <code class="docutils literal"><span class="pre">.impute2</span></code> or
<code class="docutils literal"><span class="pre">.impute2.gz</span></code> extension. There should be one <em>impute2</em> file per chromosome
(see the <a class="reference internal" href="tutorial_gwip.html#gwip-tut-output-files-final-impute2"><span>gwip/chrN/final_impute2 directories</span></a> section in the main
<a class="reference internal" href="../tutorial.html#gwip-tut"><span>Pipeline tutorial</span></a>). These files consist of the imputed genotypes required to
perform the analysis.</p>
<p>The general structure of the file contains the following columns (which are
space delimited): the chromosome, the name of the marker, its position and its
two alleles. The subsequent columns correspond to the probabilities of each
genotype (hence, there are three columns per sample). The first value
correspond to the probability of being homozygous of the first allele. The
second value correspond to the probability of being heterozygous. Finally, the
third value correspond to the probability of being homozygous of the second
allele. The following example shows two lines of the <em>impute2</em> file.</p>
<div class="highlight-text"><div class="highlight"><pre>21 rs376366718:10000302:A:G 10000302 A G 0.986 0.014 0 1 0 0 1 0 0 ...
21 21:10002805:C:T 10002805 C T 0.254 0.736 0.010 0.810 0.188 0.002 0.800 0.195 0.005 ...
</pre></div>
</div>
</div>
<div class="section" id="samples-file">
<h4>Samples file<a class="headerlink" href="#samples-file" title="Permalink to this headline">¶</a></h4>
<p>This file is generated by <a class="reference internal" href="../module_content/gwip.html#module-gwip" title="gwip"><code class="xref py py-mod docutils literal"><span class="pre">gwip</span></code></a> and has the <code class="docutils literal"><span class="pre">.sample</span></code> extension.
There should be one sample file per chromosome (see the
<a class="reference internal" href="tutorial_gwip.html#gwip-tut-output-files-final-impute2"><span>gwip/chrN/final_impute2 directories</span></a> section in the main
<a class="reference internal" href="../tutorial.html#gwip-tut"><span>Pipeline tutorial</span></a>). These files greatly resembles the <em>Plink</em> <code class="docutils literal"><span class="pre">fam</span></code> file.
Specifically, it contains the samples that are included in the <em>impute2</em> file
(with <strong>the same order</strong>). It is needed to correctly interpret the sample
described by the <em>impute2</em> file. The format is as follow:</p>
<div class="highlight-text"><div class="highlight"><pre>ID_1 ID_2 missing father mother sex plink_pheno
0 0 0 D D D B
1341 NA06985 0 0 0 2 -9
1341 NA06991 0 NA06993 NA06985 2 -9
1341 NA06993 0 0 0 1 -9
...
</pre></div>
</div>
<p>The first two rows are part of the format and should be as is.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The column <code class="docutils literal"><span class="pre">ID_2</span></code> should contain <strong>unique</strong> sample identification numbers,
since the analysis will only consider the <code class="docutils literal"><span class="pre">ID_2</span></code> (which correspond to the
sample ID in the <em>Plink</em> file) to correctly match the samples and the
imputed genotypes.</p>
</div>
</div>
<div class="section" id="phenotype-file">
<h4>Phenotype file<a class="headerlink" href="#phenotype-file" title="Permalink to this headline">¶</a></h4>
<p>This file describes the phenotype and variables used to perform the linear
regression. The file is <em>tab</em> separated and contains one row per sample, one
column per phenotype/variable.</p>
<p>The following is an example of a phenotype file:</p>
<div class="highlight-text"><div class="highlight"><pre>SampleID        Pheno1  Age     Var1    Gender
NA06985 58.6923101992   53      48.0104314206   2
NA06993 64.8981628053   47      23.7615117523   1
NA06994 129.562482664   48      20.2946857226   1
...
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Note that the gender is encoded such that males are <code class="docutils literal"><span class="pre">1</span></code> and females are
<code class="docutils literal"><span class="pre">2</span></code>. Missing gender is encoded as <code class="docutils literal"><span class="pre">0</span></code>.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The sample identification numbers should match the ones in the sample file
(see above). Those numbers should be unique for each sample. Only the
samples that are <strong>both</strong> in the sample and phenotype files will be kept for
analysis. The order of the samples in the phenotype file is not important.</p>
</div>
</div>
<div class="section" id="sites-to-extract-optional">
<h4>Sites to extract (optional)<a class="headerlink" href="#sites-to-extract-optional" title="Permalink to this headline">¶</a></h4>
<p>This file (which is optional) should contain a list of site (one identification
number per line) to keep for the analysis. This file might be the
<code class="docutils literal"><span class="pre">.good_sites</span></code> file automatically generated by <a class="reference internal" href="../module_content/gwip.html#module-gwip" title="gwip"><code class="xref py py-mod docutils literal"><span class="pre">gwip</span></code></a> (see the
<a class="reference internal" href="tutorial_gwip.html#gwip-tut-output-files-final-impute2"><span>gwip/chrN/final_impute2 directories</span></a> section in the main
<a class="reference internal" href="../tutorial.html#gwip-tut"><span>Pipeline tutorial</span></a>).</p>
</div>
</div>
<div class="section" id="usage">
<span id="lin-tut-usage"></span><h3>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h3>
<p>The following command will display the documentation for the linear regression
analysis in the console:</p>
<div class="highlight-console"><div class="highlight"><pre><span class="gp">$</span> imputed-stats linear --help
<span class="go">usage: imputed-stats linear [-h] [-v] [--debug] --impute2 FILE --sample FILE</span>
<span class="go">                            --pheno FILE [--extract-sites FILE] [--out FILE]</span>
<span class="go">                            [--nb-process INT] [--nb-lines INT] [--chrx]</span>
<span class="go">                            [--gender-column NAME] [--scale INT]</span>
<span class="go">                            [--prob FLOAT] [--maf FLOAT] [--covar NAME]</span>
<span class="go">                            [--missing-value NAME] [--sample-column NAME]</span>
<span class="go">                            [--interaction NAME] --pheno-name NAME</span>

<span class="go">Performs a linear regression (ordinary least squares) on imputed data. This</span>
<span class="go">script is part of the &#39;gwip&#39; package, version 0.1).</span>

<span class="go">optional arguments:</span>
<span class="go">  -h, --help            show this help message and exit</span>
<span class="go">  -v, --version         show program&#39;s version number and exit</span>
<span class="go">  --debug               set the logging level to debug</span>

<span class="go">Input Files:</span>
<span class="go">  --impute2 FILE        The output from IMPUTE2.</span>
<span class="go">  --sample FILE         The sample file (the order should be the same as in</span>
<span class="go">                        the IMPUTE2 files).</span>
<span class="go">  --pheno FILE          The file containing phenotypes and co variables.</span>
<span class="go">  --extract-sites FILE  A list of sites to extract for analysis (optional).</span>

<span class="go">Output Options:</span>
<span class="go">  --out FILE            The prefix for the output files. [imputed_stats]</span>

<span class="go">General Options:</span>
<span class="go">  --nb-process INT      The number of process to use. [1]</span>
<span class="go">  --nb-lines INT        The number of line to read at a time. [1000]</span>
<span class="go">  --chrx                The analysis is performed for the non pseudo-autosomal</span>
<span class="go">                        region of the chromosome X (male dosage will be</span>
<span class="go">                        divided by 2 to get values [0, 0.5] instead of [0, 1])</span>
<span class="go">                        (males are coded as 1 and option &#39;--gender-column&#39;</span>
<span class="go">                        should be used).</span>
<span class="go">  --gender-column NAME  The name of the gender column (use to exclude samples</span>
<span class="go">                        with unknown gender (i.e. not 1, male, or 2, female).</span>
<span class="go">                        If gender not available, use &#39;None&#39;. [Gender]</span>

<span class="go">Dosage Options:</span>
<span class="go">  --scale INT           Scale dosage so that values are in [0, n] (possible</span>
<span class="go">                        values are 1 (no scaling) or 2). [2]</span>
<span class="go">  --prob FLOAT          The minimal probability for which a genotype should be</span>
<span class="go">                        considered. [&gt;=0.9]</span>
<span class="go">  --maf FLOAT           Minor allele frequency threshold for which marker will</span>
<span class="go">                        be skipped. [&lt;0.01]</span>

<span class="go">Phenotype Options:</span>
<span class="go">  --covar NAME          The co variable names (in the phenotype file),</span>
<span class="go">                        separated by coma.</span>
<span class="go">  --missing-value NAME  The missing value in the phenotype file.</span>
<span class="go">  --sample-column NAME  The name of the sample ID column (in the phenotype</span>
<span class="go">                        file). [sample_id]</span>
<span class="go">  --interaction NAME    Add an interaction between the genotype and this</span>
<span class="go">                        variable.</span>

<span class="go">Linear Regression Options:</span>
<span class="go">  --pheno-name NAME     The phenotype.</span>
</pre></div>
</div>
</div>
<div class="section" id="execution-time">
<span id="lin-tut-execution-time"></span><h3>Execution time<a class="headerlink" href="#execution-time" title="Permalink to this headline">¶</a></h3>
<p>The following figure shows the approximate execution time for different number
of processes (the <code class="docutils literal"><span class="pre">--nb-process</span></code> option) with different installation methods
(<em>pyvenv</em> versus <em>miniconda</em>). This analysis was performed on a computer with
an <em>Intel(R) Core(TM) i7-3770 CPU &#64; 3.40GHz</em> (8 cores) and 16Go of RAM. The
analysis contained 30,000 imputed markers for 2,402 samples, and 6,000 lines
were processed at a time. The red line represent the execution time of the same
analysis using <em>Plink</em> (which is simple threaded).</p>
<div class="figure align-center" id="id1">
<span id="linear-exec-time"></span><a class="reference internal image-reference" href="../_images/Linear_Walltime.png"><img alt="Linear regression execution time vs number of processes." src="../_images/Linear_Walltime.png" style="width: 100%;" /></a>
<p class="caption"><span class="caption-text">Linear regression execution time vs number of processes.</span></p>
</div>
<p>Note that the linear regression from <em>Statsmodels</em> (at least when compiled on a
modern Linux system, <em>i.e.</em> when <a class="reference internal" href="../module_content/gwip.html#module-gwip" title="gwip"><code class="xref py py-mod docutils literal"><span class="pre">gwip</span></code></a> is installed using the <em>pyvenv</em>
method) uses more than 100% of each process, hence we recommend launching <em>n/2</em>
processes (where <em>n</em> is the number of processing cores on the machine). This is
not true when using a <em>miniconda</em> installation.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_logistic.html" title="Logistic Tutorial"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_SKAT.html" title="SKAT Tutorial"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">gwip 0.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../tutorial.html" >Short Tutorials</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>