

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>genipe.tools.imputed_stats &mdash; genipe 1.2.2 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.2.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/bizstyle.js"></script>
    <link rel="top" title="genipe 1.2.2 documentation" href="../../../index.html" />
    <link rel="up" title="genipe" href="../../genipe.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.2.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" accesskey="U">genipe</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for genipe.tools.imputed_stats</h1><div class="highlight"><pre>
<span class="c"># This file is part of genipe.</span>
<span class="c">#</span>
<span class="c"># This work is licensed under the Creative Commons Attribution-NonCommercial</span>
<span class="c"># 4.0 International License. To view a copy of this license, visit</span>
<span class="c"># http://creativecommons.org/licenses/by-nc/4.0/ or send a letter to Creative</span>
<span class="c"># Commons, PO Box 1866, Mountain View, CA 94042, USA.</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">platform</span>
<span class="kn">import</span> <span class="nn">traceback</span>
<span class="kn">from</span> <span class="nn">shutil</span> <span class="kn">import</span> <span class="n">which</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span>
<span class="kn">from</span> <span class="nn">subprocess</span> <span class="kn">import</span> <span class="n">Popen</span><span class="p">,</span> <span class="n">PIPE</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">namedtuple</span>

<span class="kn">import</span> <span class="nn">jinja2</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">__version__</span>
<span class="kn">from</span> <span class="nn">..formats.impute2</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">..error</span> <span class="kn">import</span> <span class="n">ProgramError</span>


<span class="n">__author__</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;Louis-Philippe Lemieux Perreault&quot;</span><span class="p">,</span> <span class="s">&quot;Marc-Andre Legault&quot;</span><span class="p">]</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s">&quot;Copyright 2014, Beaulieu-Saucier Pharmacogenomics Centre&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s">&quot;Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)&quot;</span>


<span class="c"># Check if lifelines is installed</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">lifelines</span> <span class="kn">import</span> <span class="n">CoxPHFitter</span>
    <span class="n">HAS_LIFELINES</span> <span class="o">=</span> <span class="bp">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">HAS_LIFELINES</span> <span class="o">=</span> <span class="bp">False</span>

<span class="c"># Check if statsmodels is installed</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="kn">as</span> <span class="nn">sm</span>
    <span class="kn">import</span> <span class="nn">statsmodels.formula.api</span> <span class="kn">as</span> <span class="nn">smf</span>
    <span class="n">HAS_STATSMODELS</span> <span class="o">=</span> <span class="bp">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">HAS_STATSMODELS</span> <span class="o">=</span> <span class="bp">False</span>


<span class="k">def</span> <span class="nf">_has_skat</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Checks if the SKAT R library is installed.</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if SKAT is installed, False otherwise.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">Popen</span><span class="p">(</span>
        <span class="p">[</span><span class="s">&quot;Rscript&quot;</span><span class="p">,</span> <span class="s">&quot;-e&quot;</span><span class="p">,</span> <span class="s">&#39;is.element(&quot;SKAT&quot;, installed.packages()[,1])&#39;</span><span class="p">],</span>
        <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">communicate</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">out</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&quot;TRUE&quot;</span><span class="p">)</span>


<span class="c"># Check if R and SKAT is installed</span>
<span class="n">HAS_R</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="s">&quot;Rscript&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span>
<span class="n">HAS_SKAT</span> <span class="o">=</span> <span class="n">_has_skat</span><span class="p">()</span> <span class="k">if</span> <span class="n">HAS_R</span> <span class="k">else</span> <span class="bp">False</span>


<span class="c"># An IMPUTE2 row to process</span>
<span class="n">_Row</span> <span class="o">=</span> <span class="n">namedtuple</span><span class="p">(</span><span class="s">&quot;_Row&quot;</span><span class="p">,</span> <span class="p">(</span><span class="s">&quot;row&quot;</span><span class="p">,</span> <span class="s">&quot;samples&quot;</span><span class="p">,</span> <span class="s">&quot;pheno&quot;</span><span class="p">,</span> <span class="s">&quot;pheno_name&quot;</span><span class="p">,</span> <span class="s">&quot;use_ml&quot;</span><span class="p">,</span>
                           <span class="s">&quot;categorical&quot;</span><span class="p">,</span> <span class="s">&quot;formula&quot;</span><span class="p">,</span> <span class="s">&quot;time_to_event&quot;</span><span class="p">,</span> <span class="s">&quot;event&quot;</span><span class="p">,</span>
                           <span class="s">&quot;inter_c&quot;</span><span class="p">,</span> <span class="s">&quot;is_chrx&quot;</span><span class="p">,</span> <span class="s">&quot;gender_c&quot;</span><span class="p">,</span> <span class="s">&quot;del_g&quot;</span><span class="p">,</span> <span class="s">&quot;scale&quot;</span><span class="p">,</span>
                           <span class="s">&quot;maf_t&quot;</span><span class="p">,</span> <span class="s">&quot;prob_t&quot;</span><span class="p">,</span> <span class="s">&quot;analysis_type&quot;</span><span class="p">,</span>
                           <span class="s">&quot;number_to_print&quot;</span><span class="p">))</span>

<span class="c"># The Cox&#39;s regression required values</span>
<span class="n">_COX_REQ_COLS</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;coef&quot;</span><span class="p">,</span> <span class="s">&quot;se(coef)&quot;</span><span class="p">,</span> <span class="s">&quot;lower 0.95&quot;</span><span class="p">,</span> <span class="s">&quot;upper 0.95&quot;</span><span class="p">,</span> <span class="s">&quot;z&quot;</span><span class="p">,</span> <span class="s">&quot;p&quot;</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.main">[docs]</a>    <span class="sd">&quot;&quot;&quot;The main function.</span>

<span class="sd">    Args:</span>
<span class="sd">        args (argparse.Namespace): the arguments to be parsed (if</span>
<span class="sd">                                   :py:func:`main` is called by another</span>
<span class="sd">                                   modulel)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Creating the option parser</span>
    <span class="n">desc</span> <span class="o">=</span> <span class="p">(</span><span class="s">&quot;Performs statistical analysis on imputed data (either SKAT &quot;</span>
            <span class="s">&quot;analysis, or linear, logistic or survival regression). This &quot;</span>
            <span class="s">&quot;script is part of the &#39;genipe&#39; package, version &quot;</span>
            <span class="s">&quot;{}).&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>

    <span class="c"># Files that need closing</span>
    <span class="n">logging_fh</span> <span class="o">=</span> <span class="bp">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c"># Parsing the options</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="c"># Getting the output directory (dirname of the output prefix</span>
        <span class="n">out_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

        <span class="c"># Adding the logging capability</span>
        <span class="n">log_file</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s">&quot;.log&quot;</span>
        <span class="n">logging_fh</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s">&quot;w&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
            <span class="n">format</span><span class="o">=</span><span class="s">&quot;[</span><span class="si">%(asctime)s</span><span class="s"> </span><span class="si">%(levelname)s</span><span class="s">] </span><span class="si">%(message)s</span><span class="s">&quot;</span><span class="p">,</span>
            <span class="n">datefmt</span><span class="o">=</span><span class="s">&quot;%Y-%m-</span><span class="si">%d</span><span class="s"> %H:%M:%S&quot;</span><span class="p">,</span>
            <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">debug</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
            <span class="n">handlers</span><span class="o">=</span><span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(),</span> <span class="n">logging_fh</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Logging everything into &#39;{}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">log_file</span><span class="p">))</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Program arguments: &#39;{}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])))</span>

        <span class="c"># Checking the options</span>
        <span class="n">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

        <span class="c"># Reading the phenotype file</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Reading phenotype file&quot;</span><span class="p">)</span>
        <span class="n">phenotypes</span><span class="p">,</span> <span class="n">remove_gender</span> <span class="o">=</span> <span class="n">read_phenotype</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
            <span class="n">args</span><span class="p">,</span>
            <span class="n">check_duplicated</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">!=</span> <span class="s">&quot;mixedlm&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} samples with phenotype&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">unique</span><span class="p">()),</span>
        <span class="p">))</span>

        <span class="c"># Reading the sample file</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Reading the sample file&quot;</span><span class="p">)</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="n">read_samples</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">sample</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} samples with imputation data&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span>
        <span class="p">))</span>

        <span class="c"># Reading the sites to extract (if required)</span>
        <span class="n">sites_to_extract</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">extract_sites</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Reading the sites to extract for analysis&quot;</span><span class="p">)</span>
            <span class="n">sites_to_extract</span> <span class="o">=</span> <span class="n">read_sites_to_extract</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">extract_sites</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - {:,d} sites to extract&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">sites_to_extract</span><span class="p">),</span>
            <span class="p">))</span>

        <span class="c"># Computing the statistics</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">!=</span> <span class="s">&quot;skat&quot;</span><span class="p">:</span>
            <span class="n">compute_statistics</span><span class="p">(</span>
                <span class="n">impute2_filename</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">,</span>
                <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                <span class="n">markers_to_extract</span><span class="o">=</span><span class="n">sites_to_extract</span><span class="p">,</span>
                <span class="n">phenotypes</span><span class="o">=</span><span class="n">phenotypes</span><span class="p">,</span>
                <span class="n">remove_gender</span><span class="o">=</span><span class="n">remove_gender</span><span class="p">,</span>
                <span class="n">out_prefix</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
                <span class="n">options</span><span class="o">=</span><span class="n">args</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">skat_parse_impute2</span><span class="p">(</span>
                <span class="n">impute2_filename</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">,</span>
                <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                <span class="n">markers_to_extract</span><span class="o">=</span><span class="n">sites_to_extract</span><span class="p">,</span>
                <span class="n">phenotypes</span><span class="o">=</span><span class="n">phenotypes</span><span class="p">,</span>
                <span class="n">remove_gender</span><span class="o">=</span><span class="n">remove_gender</span><span class="p">,</span>
                <span class="n">out_prefix</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
                <span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Analysis completed&quot;</span><span class="p">)</span>

    <span class="c"># Catching the Ctrl^C</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="c"># Catching the ProgramError</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">raise</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">logging_fh</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">logging_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">read_phenotype</span><span class="p">(</span><span class="n">i_filename</span><span class="p">,</span> <span class="n">opts</span><span class="p">,</span> <span class="n">check_duplicated</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span></div>
<div class="viewcode-block" id="read_phenotype"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.read_phenotype">[docs]</a>    <span class="sd">&quot;&quot;&quot;Reads the phenotype file.</span>

<span class="sd">    Args:</span>
<span class="sd">        i_filename (str): the name of the input file</span>
<span class="sd">        opts (argparse.Namespace): the options</span>
<span class="sd">        check_duplicated (bool): whether or not to check for duplicated samples</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: the phenotypes</span>

<span class="sd">    This file is expected to be a tab separated file of phenotypes and</span>
<span class="sd">    covariates. The columns to use will be determined by the</span>
<span class="sd">    ``--sample-column`` and the ``--covar`` options.</span>

<span class="sd">    For analysis including the X chromosome, the gender is automatically</span>
<span class="sd">    added as a covariate. The results are not shown to the user unless asked</span>
<span class="sd">    for.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Reading the data (and setting the index)</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">i_filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">na_values</span><span class="o">=</span><span class="n">opts</span><span class="o">.</span><span class="n">missing_value</span><span class="p">)</span>
    <span class="n">pheno</span><span class="p">[</span><span class="n">opts</span><span class="o">.</span><span class="n">sample_column</span><span class="p">]</span> <span class="o">=</span> <span class="n">pheno</span><span class="p">[</span><span class="n">opts</span><span class="o">.</span><span class="n">sample_column</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">pheno</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">opts</span><span class="o">.</span><span class="n">sample_column</span><span class="p">,</span>
                            <span class="n">verify_integrity</span><span class="o">=</span><span class="n">check_duplicated</span><span class="p">)</span>

    <span class="c"># Finding the required column</span>
    <span class="n">required_columns</span> <span class="o">=</span> <span class="n">opts</span><span class="o">.</span><span class="n">covar</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">opts</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;cox&quot;</span><span class="p">:</span>
        <span class="n">required_columns</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">opts</span><span class="o">.</span><span class="n">tte</span><span class="p">,</span> <span class="n">opts</span><span class="o">.</span><span class="n">event</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">required_columns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">opts</span><span class="o">.</span><span class="n">pheno_name</span><span class="p">)</span>

    <span class="c"># Interaction?</span>
    <span class="k">if</span> <span class="n">opts</span><span class="o">.</span><span class="n">interaction</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">opts</span><span class="o">.</span><span class="n">interaction</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">required_columns</span><span class="p">:</span>
            <span class="n">required_columns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">opts</span><span class="o">.</span><span class="n">interaction</span><span class="p">)</span>

    <span class="c"># Do we need the gender column?</span>
    <span class="n">remove_gender_column</span> <span class="o">=</span> <span class="bp">False</span>
    <span class="k">if</span> <span class="n">opts</span><span class="o">.</span><span class="n">chrx</span> <span class="ow">and</span> <span class="p">(</span><span class="n">opts</span><span class="o">.</span><span class="n">gender_column</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">required_columns</span><span class="p">):</span>
        <span class="c"># It&#39;s not already in the gender column, so we keep it, but we need to</span>
        <span class="c"># remove it later on...</span>
        <span class="n">required_columns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">opts</span><span class="o">.</span><span class="n">gender_column</span><span class="p">)</span>
        <span class="n">remove_gender_column</span> <span class="o">=</span> <span class="bp">True</span>

    <span class="c"># We need to exclude unknown gender only if gender was required</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">opts</span><span class="o">.</span><span class="n">gender_column</span> <span class="ow">in</span> <span class="n">required_columns</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="ow">not</span> <span class="n">remove_gender_column</span><span class="p">):</span>
        <span class="n">pheno</span> <span class="o">=</span> <span class="n">pheno</span><span class="p">[(</span><span class="n">pheno</span><span class="p">[</span><span class="n">opts</span><span class="o">.</span><span class="n">gender_column</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="o">|</span>
                      <span class="p">(</span><span class="n">pheno</span><span class="p">[</span><span class="n">opts</span><span class="o">.</span><span class="n">gender_column</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)]</span>

    <span class="c"># Extracting the required column</span>
    <span class="n">pheno</span> <span class="o">=</span> <span class="n">pheno</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">required_columns</span><span class="p">]</span>

    <span class="c"># Returning the phenotypes (dropping the nan values)</span>
    <span class="k">return</span> <span class="n">pheno</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">remove_gender_column</span>


<span class="k">def</span> <span class="nf">read_samples</span><span class="p">(</span><span class="n">i_filename</span><span class="p">):</span></div>
<div class="viewcode-block" id="read_samples"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.read_samples">[docs]</a>    <span class="sd">&quot;&quot;&quot;Reads the sample file (produced by SHAPEIT).</span>

<span class="sd">    Args:</span>
<span class="sd">        i_filename (str): the name of the input file</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: the list of samples</span>

<span class="sd">    This file contains the list of samples that are contained in the</span>
<span class="sd">    ``impute2`` file (with same order). The expected format for this file is a</span>
<span class="sd">    tab separated file with a first row containing the following columns: ::</span>

<span class="sd">        ID_1	ID_2	missing	father	mother	sex	plink_pheno</span>

<span class="sd">    The subsequent row will be discarded and should contain: ::</span>

<span class="sd">        0	0	0 D	D	D	B</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">        We are mostly interested in the sample IDs corresponding to the</span>
<span class="sd">        ``ID_2`` column. Their uniqueness is verified by pandas.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">i_filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot; &quot;</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">samples</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">samples</span><span class="p">[</span><span class="s">&quot;ID_2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="s">&quot;ID_2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">samples</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s">&quot;ID_2&quot;</span><span class="p">,</span> <span class="n">verify_integrity</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">skat_read_snp_set</span><span class="p">(</span><span class="n">i_filename</span><span class="p">):</span></div>
<div class="viewcode-block" id="skat_read_snp_set"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.skat_read_snp_set">[docs]</a>    <span class="sd">&quot;&quot;&quot;Reads the SKAT SNP set file.</span>

<span class="sd">    Args:</span>
<span class="sd">        i_filename (str): the name of the input file</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: the SNP set for the SKAT analysis</span>

<span class="sd">    This file has to be supplied by the user. The recognized columns are:</span>
<span class="sd">    ``variant``, ``snp_set`` and ``weight``. The ``weight`` column is optional</span>
<span class="sd">    and can be used to specify a custom weighting scheme for SKAT. If nothing</span>
<span class="sd">    is specified, the default Beta weights are used.</span>

<span class="sd">    The file has to be tab delimited.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">skat_info</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">i_filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">if</span> <span class="s">&quot;variant&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">skat_info</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;The SKAT SNP set file needs to have a &#39;variant&#39; &quot;</span>
                           <span class="s">&quot;column containing the ID of every variant of &quot;</span>
                           <span class="s">&quot;interest.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s">&quot;snp_set&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">skat_info</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;The SKAT SNP set file needs to have a &#39;snp_set&#39; &quot;</span>
                           <span class="s">&quot;column containing the SNP set ID for every &quot;</span>
                           <span class="s">&quot;variant. The user is free to choose the SNP ID&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">skat_info</span>


<span class="k">def</span> <span class="nf">read_sites_to_extract</span><span class="p">(</span><span class="n">i_filename</span><span class="p">):</span></div>
<div class="viewcode-block" id="read_sites_to_extract"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.read_sites_to_extract">[docs]</a>    <span class="sd">&quot;&quot;&quot;Reads the list of sites to extract.</span>

<span class="sd">    Args:</span>
<span class="sd">        i_filename (str): The input filename containing the IDs of the variants</span>
<span class="sd">                          to consider for the analysis.</span>

<span class="sd">    Returns:</span>
<span class="sd">        set: A set containing the variants.</span>

<span class="sd">    The expected file format is simply a list of variants. Every row should</span>
<span class="sd">    correspond to a single variant identifier. ::</span>

<span class="sd">        3:60069:t</span>
<span class="sd">        rs123456:A</span>
<span class="sd">        3:60322:A</span>

<span class="sd">    Typically, this is used to analyze only variants that passed some QC</span>
<span class="sd">    threshold. The :py:mod:`genipe` pipeline generates this file at the</span>
<span class="sd">    &#39;merge_impute2&#39; step.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">markers_to_extract</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">i_filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="n">markers_to_extract</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">i_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">markers_to_extract</span>


<span class="k">def</span> <span class="nf">skat_parse_impute2</span><span class="p">(</span><span class="n">impute2_filename</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">markers_to_extract</span><span class="p">,</span></div>
<div class="viewcode-block" id="skat_parse_impute2"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.skat_parse_impute2">[docs]</a>                       <span class="n">phenotypes</span><span class="p">,</span> <span class="n">remove_gender</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read the impute2 file and run the SKAT analysis.</span>

<span class="sd">    Args:</span>
<span class="sd">        impute2_filename (str): the name of the input file</span>
<span class="sd">        samples (pandas.DataFrame): the samples</span>
<span class="sd">        markers_to_extract (set): the set of markers to analyse</span>
<span class="sd">        phenotypes (pandas.DataFrame): the phenotypes</span>
<span class="sd">        remove_gender (bool): whether or not to remove the gender column</span>
<span class="sd">        out_prefix (str): the output prefix</span>
<span class="sd">        args (argparse.Namespace): the options</span>


<span class="sd">    This function does most of the &quot;dispatching&quot; to run SKAT. It writes the</span>
<span class="sd">    input files to the disk, runs the generated R scripts to do the actual</span>
<span class="sd">    analysis and then writes the results to disk.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c"># We keep track of the files that are generated because we will need the</span>
    <span class="c"># paths to generate the R script correctly.</span>
    <span class="n">r_files</span> <span class="o">=</span> <span class="p">{</span><span class="s">&quot;snp_sets&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s">&quot;covariates&quot;</span><span class="p">:</span> <span class="bp">None</span><span class="p">,</span> <span class="s">&quot;outcome&quot;</span><span class="p">:</span> <span class="bp">None</span><span class="p">,</span>
               <span class="s">&quot;weights&quot;</span><span class="p">:</span> <span class="bp">None</span><span class="p">}</span>

    <span class="c"># Read the SNP set and create the output files.</span>
    <span class="n">snp_set</span> <span class="o">=</span> <span class="n">skat_read_snp_set</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">snp_sets</span><span class="p">)</span>

    <span class="c"># We will use a temporary directory for our analysis.</span>
    <span class="n">dir_name</span> <span class="o">=</span> <span class="s">&quot;{}.{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span>
        <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">today</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s">&quot;skat.%Y.%m.</span><span class="si">%d</span><span class="s">&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">dir_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">dir_name</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">dir_name</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">dir_name</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c"># This should not happen because the folder name contains a timestamp</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;A folder named &#39;{}&#39; already exists.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">dir_name</span>
        <span class="p">))</span>

    <span class="c"># If weights were provided, we will write a weight vector to disk for R.</span>
    <span class="k">if</span> <span class="s">&quot;weight&quot;</span> <span class="ow">in</span> <span class="n">snp_set</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;SKAT will use the provided weights (from the SNP set &quot;</span>
                     <span class="s">&quot;file).&quot;</span><span class="p">)</span>
        <span class="n">weight_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="s">&quot;weights.csv&quot;</span><span class="p">)</span>
        <span class="n">snp_set</span><span class="p">[[</span><span class="s">&quot;weight&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">weight_filename</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
        <span class="n">r_files</span><span class="p">[</span><span class="s">&quot;weights&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">weight_filename</span>

    <span class="c"># Open genotype CSV files for every SNP set. Those CSV files will be</span>
    <span class="c"># read in R and used by SKAT.</span>
    <span class="n">genotype_files</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">snp_sets</span> <span class="o">=</span> <span class="n">snp_set</span><span class="p">[</span><span class="s">&quot;snp_set&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">set_id</span> <span class="ow">in</span> <span class="n">snp_sets</span><span class="p">:</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="s">&quot;{}.genotypes.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">set_id</span><span class="p">))</span>
        <span class="n">r_files</span><span class="p">[</span><span class="s">&quot;snp_sets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>  <span class="c"># Track the filenames.</span>
        <span class="n">genotype_files</span><span class="p">[</span><span class="n">set_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
        <span class="c"># We write the column headers for every file.</span>
        <span class="k">print</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="p">,</span> <span class="o">*</span><span class="n">samples</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">genotype_files</span><span class="p">[</span><span class="n">set_id</span><span class="p">])</span>

    <span class="c"># The markers of interest are the markers we want to include in the</span>
    <span class="c"># analysis. Concretely, they are the markers that were included in the snp</span>
    <span class="c"># sets. If a list of markers to extract is provided by the user, we also</span>
    <span class="c"># look at this to filter ou undesired markers.</span>
    <span class="c">#</span>
    <span class="c"># We also fill a set of written_markers to stop whenever all the markers</span>
    <span class="c"># were written to file.</span>
    <span class="n">markers_of_interest</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">snp_set</span><span class="p">[</span><span class="s">&quot;variant&quot;</span><span class="p">])</span>
    <span class="n">written_markers</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">markers_to_extract</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">markers_of_interest</span> <span class="o">=</span> <span class="n">markers_of_interest</span> <span class="o">&amp;</span> <span class="n">markers_to_extract</span>

    <span class="c"># Open the file. We use subprocess if it&#39;s gunzipped because it&#39;s faster.</span>
    <span class="c"># We use gzip -d -c instead of zcat because the default Mac OS zcat has</span>
    <span class="c"># weird behavior.</span>
    <span class="k">if</span> <span class="n">impute2_filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&quot;.gz&quot;</span><span class="p">):</span>
        <span class="n">proc</span> <span class="o">=</span> <span class="n">Popen</span><span class="p">([</span><span class="s">&quot;gzip&quot;</span><span class="p">,</span> <span class="s">&quot;-d&quot;</span><span class="p">,</span> <span class="s">&quot;-c&quot;</span><span class="p">,</span> <span class="n">impute2_filename</span><span class="p">],</span> <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">)</span>
        <span class="n">i_file</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">stdout</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">i_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">impute2_filename</span><span class="p">,</span> <span class="s">&quot;rb&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="c"># If we already found everything, we will stop here.</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">written_markers</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">markers_of_interest</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">written_markers</span> <span class="o">==</span> <span class="n">markers_of_interest</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Found all the necessary markers, skipping the &quot;</span>
                             <span class="s">&quot;rest of the file.&quot;</span><span class="p">)</span>
                <span class="k">break</span>

        <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s">&quot;ascii&quot;</span><span class="p">)</span>

        <span class="c"># TODO: Add the gender and do QC (men with hetero on chrX).</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">_skat_parse_line</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">markers_of_interest</span><span class="p">,</span> <span class="n">samples</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">line</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">name</span><span class="p">,</span> <span class="n">dosage</span> <span class="o">=</span> <span class="n">line</span>
            <span class="n">written_markers</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">_skat_write_marker</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">dosage</span><span class="p">,</span> <span class="n">snp_set</span><span class="p">,</span> <span class="n">genotype_files</span><span class="p">)</span>

    <span class="c"># We will warn the user if some variants were not found.</span>
    <span class="n">number_missing</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">markers_of_interest</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">written_markers</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">number_missing</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;{} markers of interest were not found in the Impute2 &quot;</span>
                        <span class="s">&quot;file.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">number_missing</span><span class="p">))</span>

    <span class="n">i_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c"># Close the genotype files.</span>
    <span class="k">for</span> <span class="n">file_handle</span> <span class="ow">in</span> <span class="n">genotype_files</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="n">file_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c"># Write the covariate file.</span>
    <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">samples</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">)</span>
    <span class="n">phenotype_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;sample&quot;</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">covar</span><span class="p">:</span>
        <span class="c"># Make sure the samples are consistent by merging phenotype and</span>
        <span class="c"># samples.</span>
        <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="s">&quot;covariates.csv&quot;</span><span class="p">)</span>
        <span class="n">phenotype_df</span><span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">covar</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">filename</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">r_files</span><span class="p">[</span><span class="s">&quot;covariates&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">filename</span>

    <span class="c"># Write the phenotype file.</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">dir_name</span><span class="p">,</span>
        <span class="s">&quot;{}.{}.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pheno_name</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">outcome_type</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">phenotype_df</span><span class="p">[[</span><span class="n">args</span><span class="o">.</span><span class="n">pheno_name</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="n">filename</span><span class="p">,</span>
        <span class="n">sep</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">r_files</span><span class="p">[</span><span class="s">&quot;outcome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">filename</span>

    <span class="n">r_scripts</span> <span class="o">=</span> <span class="n">_skat_generate_r_script</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="n">r_files</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

    <span class="c"># Run the SKAT analysis by calling Rscript either in different subprocesses</span>
    <span class="c"># or linearly.</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Launching SKAT using {} processes on {} SNP sets.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">nb_process</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">snp_sets</span><span class="p">)</span>
    <span class="p">))</span>

    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">nb_process</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">nb_process</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">_skat_run_job</span><span class="p">,</span> <span class="n">r_scripts</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">script</span> <span class="ow">in</span> <span class="n">r_scripts</span><span class="p">:</span>
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_skat_run_job</span><span class="p">(</span><span class="n">script</span><span class="p">))</span>

    <span class="c"># Finally, write the SKAT output to disk.</span>
    <span class="n">output_filename</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s">&quot;.skat.dosage&quot;</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;SKAT completed, writing the results to disk ({}).&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">output_filename</span>
    <span class="p">))</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="c"># Write the header.</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">skat_o</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="s">&quot;snp_set_id&quot;</span><span class="p">,</span> <span class="s">&quot;p_value&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="s">&quot;snp_set_id&quot;</span><span class="p">,</span> <span class="s">&quot;p_value&quot;</span><span class="p">,</span> <span class="s">&quot;q_value&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>

        <span class="c"># The order in the snp_sets list should have been consistent</span>
        <span class="c"># across the whole analysis. We just want to make sure that we</span>
        <span class="c"># actually have the right number of results.</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">snp_sets</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

        <span class="c"># Write a tab separated file containing the set and p-values.</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">p_value</span><span class="p">,</span> <span class="n">q_value</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">results</span><span class="p">):</span>
            <span class="n">set_id</span> <span class="o">=</span> <span class="n">snp_sets</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">skat_o</span><span class="p">:</span>
                <span class="k">print</span><span class="p">(</span><span class="n">set_id</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">print</span><span class="p">(</span><span class="n">set_id</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">q_value</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_skat_run_job</span><span class="p">(</span><span class="n">script_filename</span><span class="p">):</span></div>
    <span class="sd">&quot;&quot;&quot;Calls Rscript with the generated script and parses the results.</span>

<span class="sd">    Args:</span>
<span class="sd">        script_filename (str): the name of the script</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: two values: the *p-value* and the *q-value* (for SKAT-O, the</span>
<span class="sd">               *q-value* is set to None)</span>

<span class="sd">    The results should be somewhere in the standard output. The expected</span>
<span class="sd">    format is: ::</span>

<span class="sd">        _PYTHON_HOOK_QVAL:[0.123]</span>
<span class="sd">        _PYTHON_HOOK_PVAL:[0.123]</span>

<span class="sd">    If the template script is modified, this format should still be respected.</span>

<span class="sd">    It is also noteworthy that this function uses ``Rscript`` to run the</span>
<span class="sd">    analysis. Hence, it should be in the path when using the imputed_stats</span>
<span class="sd">    skat mode.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Parse the p-value.</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="n">Popen</span><span class="p">(</span>
        <span class="p">[</span><span class="s">&quot;Rscript&quot;</span><span class="p">,</span> <span class="n">script_filename</span><span class="p">],</span>
        <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span>
        <span class="n">stderr</span><span class="o">=</span><span class="n">PIPE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">out</span><span class="p">,</span> <span class="n">err</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">err</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;SKAT Warning: &quot;</span> <span class="o">+</span> <span class="n">err</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s">&quot;utf-8&quot;</span><span class="p">))</span>

    <span class="c"># Decoding</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s">&quot;utf-8&quot;</span><span class="p">)</span>

    <span class="c"># Getting the p value</span>
    <span class="n">p_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&quot;_PYTHON_HOOK_PVAL:\[(.+)\]&quot;</span><span class="p">,</span> <span class="n">out</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">p_match</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;SKAT did not return properly. See script &quot;</span>
                           <span class="s">&quot;&#39;{}&#39; for details.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">script_filename</span><span class="p">))</span>

    <span class="c"># Getting the Q value</span>
    <span class="n">q_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&quot;_PYTHON_HOOK_QVAL:\[(.+)\]&quot;</span><span class="p">,</span> <span class="n">out</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">q_match</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;SKAT did not return properly. See script &quot;</span>
                           <span class="s">&quot;&#39;{}&#39; for details.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">script_filename</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">q_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="s">&quot;NA&quot;</span><span class="p">:</span>
        <span class="c"># For SKAT-O, the Q will be set to NA.</span>
        <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">p_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)),</span> <span class="bp">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">p_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)),</span> <span class="nb">float</span><span class="p">(</span><span class="n">q_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">_skat_generate_r_script</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="n">r_files</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Uses jinja2 to generate an R script to do the SKAT analysis.</span>

<span class="sd">    Args:</span>
<span class="sd">        dir_name (str): the output directory name to write the scripts in</span>
<span class="sd">        r_files (dict): contains the different input files required by the R</span>
<span class="sd">                        script</span>
<span class="sd">        args (argparse.Namespace): the parsed arguments</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: the list of all script names</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">jinja_env</span> <span class="o">=</span> <span class="n">jinja2</span><span class="o">.</span><span class="n">Environment</span><span class="p">(</span>
        <span class="n">loader</span><span class="o">=</span><span class="n">jinja2</span><span class="o">.</span><span class="n">PackageLoader</span><span class="p">(</span><span class="s">&quot;genipe&quot;</span><span class="p">,</span> <span class="s">&quot;script_templates&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">template</span> <span class="o">=</span> <span class="n">jinja_env</span><span class="o">.</span><span class="n">get_template</span><span class="p">(</span><span class="s">&quot;run_skat.R&quot;</span><span class="p">)</span>

    <span class="n">scripts</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c"># Create one file per SNP set for parallelism.</span>
    <span class="k">for</span> <span class="n">snp_set_file</span> <span class="ow">in</span> <span class="n">r_files</span><span class="p">[</span><span class="s">&quot;snp_sets&quot;</span><span class="p">]:</span>
        <span class="n">rendered_script</span> <span class="o">=</span> <span class="n">template</span><span class="o">.</span><span class="n">render</span><span class="p">(</span>
            <span class="n">version</span><span class="o">=</span><span class="n">__version__</span><span class="p">,</span>

            <span class="n">snp_set_file</span><span class="o">=</span><span class="n">snp_set_file</span><span class="p">,</span>
            <span class="n">covariate_file</span><span class="o">=</span><span class="n">r_files</span><span class="p">[</span><span class="s">&quot;covariates&quot;</span><span class="p">],</span>
            <span class="n">outcome_file</span><span class="o">=</span><span class="n">r_files</span><span class="p">[</span><span class="s">&quot;outcome&quot;</span><span class="p">],</span>
            <span class="n">weights</span><span class="o">=</span><span class="n">r_files</span><span class="p">[</span><span class="s">&quot;weights&quot;</span><span class="p">],</span>

            <span class="n">outcome_type</span><span class="o">=</span><span class="s">&quot;C&quot;</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">outcome_type</span> <span class="o">==</span> <span class="s">&quot;continuous&quot;</span> <span class="k">else</span> <span class="s">&quot;D&quot;</span><span class="p">,</span>
            <span class="n">skat_o</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">skat_o</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c"># Write the rendered script to disk.</span>
        <span class="n">script_filename</span> <span class="o">=</span> <span class="s">&quot;run_skat_{}R&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="c"># We parse the set id name.</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">snp_set_file</span><span class="p">)[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="s">&quot;.genotype.csv&quot;</span><span class="p">)]</span>
        <span class="p">)</span>
        <span class="n">script_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_name</span><span class="p">,</span> <span class="n">script_filename</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">script_filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">rendered_script</span><span class="p">)</span>

        <span class="n">scripts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">script_filename</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">scripts</span>


<span class="k">def</span> <span class="nf">_skat_parse_line</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">markers_of_interest</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">gender</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parses a single line of the Impute2 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        line (str): a line from the Impute2 file</span>
<span class="sd">        markers_of_interest (set): a set of markers that are required for the</span>
<span class="sd">                                   analysis</span>
<span class="sd">        samples (pandas.DataFrame): contains the samples IDs (this is useful to</span>
<span class="sd">                                    make sure we return a dosage vector with</span>
<span class="sd">                                    the appropriate data)</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: Either None if the marker is not of interest or a tuple of</span>
<span class="sd">               ``(name , dosage_vector)`` where ``name`` is a ``str``</span>
<span class="sd">               representing the variant ID and ``dosage_vector`` is a numpy</span>
<span class="sd">               array containing the dosage values for every sample in the</span>
<span class="sd">               ``samples`` dataframe.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="p">)</span>
    <span class="c"># info_tuple contains: chrom, name, pos, a1, a2</span>
    <span class="c"># proba_matrix is a matrix of sample x (aa, ab, bb)</span>
    <span class="n">info_tuple</span><span class="p">,</span> <span class="n">proba_matrix</span> <span class="o">=</span> <span class="n">matrix_from_line</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

    <span class="n">chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span> <span class="o">=</span> <span class="n">info_tuple</span>

    <span class="c"># If this marker is not of interest, we don&#39;t bother continuing.</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">markers_of_interest</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">None</span>

    <span class="c"># We need to compute the dosage vector.</span>
    <span class="c"># This is given wrt to the minor and major alleles, so we need to get the</span>
    <span class="c"># MAF to identify those.</span>
    <span class="n">maf</span><span class="p">,</span> <span class="n">minor_i</span><span class="p">,</span> <span class="n">major_i</span> <span class="o">=</span> <span class="n">maf_from_probs</span><span class="p">(</span><span class="n">proba_matrix</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">gender</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>

    <span class="c"># We don&#39;t pass a scale parameter, because we want additive coding.</span>
    <span class="n">dosage</span> <span class="o">=</span> <span class="n">dosage_from_probs</span><span class="p">(</span>
        <span class="n">homo_probs</span><span class="o">=</span><span class="n">proba_matrix</span><span class="p">[:,</span> <span class="n">minor_i</span><span class="p">],</span>
        <span class="n">hetero_probs</span><span class="o">=</span><span class="n">proba_matrix</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">dosage</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_skat_write_marker</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">dosage</span><span class="p">,</span> <span class="n">snp_set</span><span class="p">,</span> <span class="n">genotype_files</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Write the dosage information to the appropriate genotype file.</span>

<span class="sd">    Args:</span>
<span class="sd">        name (str): the name of the marker</span>
<span class="sd">        dosage (numpy.array): the dosage vector</span>
<span class="sd">        snp_set (pandas.DataFrame: the dataframe that allows us to identify the</span>
<span class="sd">                                   correct SNP set for the specified variant</span>
<span class="sd">        genotype_files (dict): a dictionary containing the opened CSV files for</span>
<span class="sd">                               the genotypes</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Identify the correct SNP set.</span>
    <span class="n">this_snp_set</span> <span class="o">=</span> <span class="n">snp_set</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">snp_set</span><span class="p">[</span><span class="s">&quot;variant&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;snp_set&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">set_id</span> <span class="ow">in</span> <span class="n">this_snp_set</span><span class="p">:</span>
        <span class="n">file_object</span> <span class="o">=</span> <span class="n">genotype_files</span><span class="p">[</span><span class="n">set_id</span><span class="p">]</span>
        <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="o">*</span><span class="n">dosage</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">file_object</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">compute_statistics</span><span class="p">(</span><span class="n">impute2_filename</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">markers_to_extract</span><span class="p">,</span>
<div class="viewcode-block" id="compute_statistics"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.compute_statistics">[docs]</a>                       <span class="n">phenotypes</span><span class="p">,</span> <span class="n">remove_gender</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parses IMPUTE2 file while computing statistics.</span>

<span class="sd">    Args:</span>
<span class="sd">        impute2_filename (str): the name of the input file</span>
<span class="sd">        samples (pandas.DataFrame): the list of samples</span>
<span class="sd">        markers_to_extract (set): the set of markers to extract</span>
<span class="sd">        phenotypes (pandas.DataFrame): the phenotypes</span>
<span class="sd">        remove_gender (bool): whether or not to remove the gender column</span>
<span class="sd">        out_prefix (str): the output prefix</span>
<span class="sd">        options (argparse.Namespace): the options</span>

<span class="sd">    This function takes care of parallelism. It reads the Impute2 file and</span>
<span class="sd">    fills a queue that will trigger the analysis when full.</span>

<span class="sd">    If the number of process to launch is 1, the rows are analyzed as they</span>
<span class="sd">    come.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The name of the output file</span>
    <span class="n">o_name</span> <span class="o">=</span> <span class="s">&quot;{}.{}.dosage&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out_prefix</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">analysis_type</span><span class="p">)</span>

    <span class="c"># Do we need to create a formula?</span>
    <span class="n">formula</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">!=</span> <span class="s">&quot;cox&quot;</span><span class="p">:</span>
        <span class="n">formula</span> <span class="o">=</span> <span class="n">get_formula</span><span class="p">(</span>
            <span class="n">phenotype</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">pheno_name</span><span class="p">,</span>
            <span class="n">covars</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">covar</span><span class="p">,</span>
            <span class="n">interaction</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">interaction</span><span class="p">,</span>
            <span class="n">gender_c</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">gender_column</span><span class="p">,</span>
            <span class="n">categorical</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">categorical</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;{}: &#39;{}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">analysis_type</span><span class="p">,</span> <span class="n">formula</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;mixedlm&quot;</span> <span class="ow">and</span> <span class="n">options</span><span class="o">.</span><span class="n">use_ml</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - using ML&quot;</span><span class="p">)</span>

    <span class="c"># Reading the IMPUTE2 file one line (site) at a time, creating a subprocess</span>
    <span class="c"># if required</span>
    <span class="n">proc</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">i_file</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">o_name</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">pool</span> <span class="o">=</span> <span class="bp">None</span>

    <span class="c"># Multiprocessing?</span>
    <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">nb_process</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">pool</span> <span class="o">=</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">nb_process</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">impute2_filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&quot;.gz&quot;</span><span class="p">):</span>
            <span class="n">proc</span> <span class="o">=</span> <span class="n">Popen</span><span class="p">([</span><span class="s">&quot;gzip&quot;</span><span class="p">,</span> <span class="s">&quot;-d&quot;</span><span class="p">,</span> <span class="s">&quot;-c&quot;</span><span class="p">,</span> <span class="n">impute2_filename</span><span class="p">],</span> <span class="n">stdout</span><span class="o">=</span><span class="n">PIPE</span><span class="p">)</span>
            <span class="n">i_file</span> <span class="o">=</span> <span class="n">proc</span><span class="o">.</span><span class="n">stdout</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">i_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">impute2_filename</span><span class="p">,</span> <span class="s">&quot;rb&quot;</span><span class="p">)</span>

        <span class="c"># Printing the header of the output file</span>
        <span class="n">header</span> <span class="o">=</span> <span class="p">(</span><span class="s">&quot;chr&quot;</span><span class="p">,</span> <span class="s">&quot;pos&quot;</span><span class="p">,</span> <span class="s">&quot;snp&quot;</span><span class="p">,</span> <span class="s">&quot;major&quot;</span><span class="p">,</span> <span class="s">&quot;minor&quot;</span><span class="p">,</span> <span class="s">&quot;maf&quot;</span><span class="p">,</span> <span class="s">&quot;n&quot;</span><span class="p">,</span> <span class="s">&quot;coef&quot;</span><span class="p">,</span>
                  <span class="s">&quot;se&quot;</span><span class="p">,</span> <span class="s">&quot;lower&quot;</span><span class="p">,</span> <span class="s">&quot;upper&quot;</span><span class="p">,</span>
                  <span class="s">&quot;t&quot;</span> <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;linear&quot;</span> <span class="k">else</span> <span class="s">&quot;z&quot;</span><span class="p">,</span> <span class="s">&quot;p&quot;</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="o">*</span><span class="n">header</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>

        <span class="c"># The sites to process (if multiprocessing)</span>
        <span class="n">sites_to_process</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c"># Reading the file</span>
        <span class="n">nb_processed</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\r\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="p">)</span>

            <span class="c"># Is this site required?</span>
            <span class="k">if</span> <span class="n">markers_to_extract</span> <span class="ow">and</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">markers_to_extract</span><span class="p">):</span>
                <span class="k">continue</span>

            <span class="c"># Constructing the row object</span>
            <span class="n">site</span> <span class="o">=</span> <span class="n">_Row</span><span class="p">(</span>
                <span class="n">row</span><span class="o">=</span><span class="n">row</span><span class="p">,</span>
                <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                <span class="n">pheno</span><span class="o">=</span><span class="n">phenotypes</span><span class="p">,</span>
                <span class="n">use_ml</span><span class="o">=</span><span class="nb">vars</span><span class="p">(</span><span class="n">options</span><span class="p">)</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;use_ml&quot;</span><span class="p">,</span> <span class="bp">None</span><span class="p">),</span>
                <span class="n">pheno_name</span><span class="o">=</span><span class="nb">vars</span><span class="p">(</span><span class="n">options</span><span class="p">)</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;pheno_name&quot;</span><span class="p">,</span> <span class="bp">None</span><span class="p">),</span>
                <span class="n">formula</span><span class="o">=</span><span class="n">formula</span><span class="p">,</span>
                <span class="n">time_to_event</span><span class="o">=</span><span class="nb">vars</span><span class="p">(</span><span class="n">options</span><span class="p">)</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;tte&quot;</span><span class="p">,</span> <span class="bp">None</span><span class="p">),</span>
                <span class="n">event</span><span class="o">=</span><span class="nb">vars</span><span class="p">(</span><span class="n">options</span><span class="p">)</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;event&quot;</span><span class="p">,</span> <span class="bp">None</span><span class="p">),</span>
                <span class="n">inter_c</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">interaction</span><span class="p">,</span>
                <span class="n">is_chrx</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">chrx</span><span class="p">,</span>
                <span class="n">gender_c</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">gender_column</span><span class="p">,</span>
                <span class="n">del_g</span><span class="o">=</span><span class="n">remove_gender</span><span class="p">,</span>
                <span class="n">scale</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">scale</span><span class="p">,</span>
                <span class="n">maf_t</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">maf</span><span class="p">,</span>
                <span class="n">prob_t</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">prob</span><span class="p">,</span>
                <span class="n">analysis_type</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">analysis_type</span><span class="p">,</span>
                <span class="n">number_to_print</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">header</span><span class="p">),</span>
                <span class="n">categorical</span><span class="o">=</span><span class="n">options</span><span class="o">.</span><span class="n">categorical</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c"># Is there more than one process</span>
            <span class="k">if</span> <span class="n">options</span><span class="o">.</span><span class="n">nb_process</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c"># Saving this site to process later</span>
                <span class="n">sites_to_process</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>

                <span class="c"># Is there enough sites to process?</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sites_to_process</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">options</span><span class="o">.</span><span class="n">nb_lines</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_impute2_site</span><span class="p">,</span>
                                           <span class="n">sites_to_process</span><span class="p">):</span>
                        <span class="k">print</span><span class="p">(</span><span class="o">*</span><span class="n">result</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>

                    <span class="c"># Logging</span>
                    <span class="n">nb_processed</span> <span class="o">+=</span> <span class="n">options</span><span class="o">.</span><span class="n">nb_lines</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Processed {:,d} lines&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_processed</span><span class="p">))</span>

                    <span class="c"># Resetting the sites to process</span>
                    <span class="n">sites_to_process</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="c"># Processing this row</span>
                <span class="k">print</span><span class="p">(</span><span class="o">*</span><span class="n">process_impute2_site</span><span class="p">(</span><span class="n">site</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sites_to_process</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_impute2_site</span><span class="p">,</span> <span class="n">sites_to_process</span><span class="p">):</span>
                <span class="k">print</span><span class="p">(</span><span class="o">*</span><span class="n">result</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">o_file</span><span class="p">)</span>

            <span class="c"># Logging</span>
            <span class="n">nb_processed</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sites_to_process</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Processed {:,d} lines&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_processed</span><span class="p">))</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">pool</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">pool</span><span class="o">.</span><span class="n">terminate</span><span class="p">()</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">traceback</span><span class="o">.</span><span class="n">format_exc</span><span class="p">())</span>
        <span class="k">raise</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="c"># Closing the input file</span>
        <span class="n">i_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c"># Finishing the rows if required</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">options</span><span class="o">.</span><span class="n">nb_process</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">pool</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">):</span>
            <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c"># Closing the output file</span>
        <span class="n">o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c"># Closing the proc</span>
        <span class="k">if</span> <span class="n">proc</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">proc</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: problem while reading the GZ &quot;</span>
                                   <span class="s">&quot;file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">impute2_filename</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">process_impute2_site</span><span class="p">(</span><span class="n">site_info</span><span class="p">):</span></div>
<div class="viewcode-block" id="process_impute2_site"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.process_impute2_site">[docs]</a>    <span class="sd">&quot;&quot;&quot;Process an IMPUTE2 site (a line in an IMPUTE2 file).</span>

<span class="sd">    Args:</span>
<span class="sd">        site_info (list): the impute2 line (split by space)</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: the results of the analysis</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Getting the probability matrix and site information</span>
    <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">),</span> <span class="n">geno</span> <span class="o">=</span> <span class="n">matrix_from_line</span><span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">row</span><span class="p">)</span>

    <span class="c"># The name of the dosage column</span>
    <span class="n">dosage_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;_D1&quot;</span><span class="p">,</span> <span class="s">&quot;_D2&quot;</span><span class="p">,</span> <span class="s">&quot;_D3&quot;</span><span class="p">]</span>

    <span class="c"># Allele encoding</span>
    <span class="n">allele_encoding</span> <span class="o">=</span> <span class="p">{</span><span class="n">dosage_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="n">a1</span><span class="p">,</span> <span class="n">dosage_columns</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span> <span class="n">a2</span><span class="p">}</span>

    <span class="c"># Creating the sample data frame</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">site_info</span><span class="o">.</span><span class="n">samples</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">col_name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">dosage_columns</span><span class="p">):</span>
        <span class="n">samples</span><span class="p">[</span><span class="n">col_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span>

    <span class="c"># Merging with phenotypes</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">site_info</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
        <span class="n">samples</span><span class="p">,</span>
        <span class="n">how</span><span class="o">=</span><span class="s">&quot;inner&quot;</span><span class="p">,</span>
        <span class="n">left_index</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">right_index</span><span class="o">=</span><span class="bp">True</span>
    <span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)[</span><span class="nb">list</span><span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">pheno</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">+</span> <span class="n">dosage_columns</span><span class="p">]</span>

    <span class="c"># Keeping only good quality markers</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span>
        <span class="n">get_good_probs</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">dosage_columns</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">site_info</span><span class="o">.</span><span class="n">prob_t</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c"># FIXME: Quick and dirty fix for mixedlm...</span>
    <span class="c"># If the analysis type is not MixedLM, then t_data is just a pointer to the</span>
    <span class="c"># real data. Otherwise, t_data is the grouped values (first occurrence,</span>
    <span class="c"># which shouldn&#39;t cause a problem, since gender and probabilities are the</span>
    <span class="c"># same for each measurement).</span>
    <span class="n">t_data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="k">if</span> <span class="n">site_info</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;mixedlm&quot;</span><span class="p">:</span>
        <span class="n">t_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>

    <span class="c"># Checking gender if required</span>
    <span class="n">gender</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">site_info</span><span class="o">.</span><span class="n">is_chrx</span><span class="p">:</span>
        <span class="c"># We want to exclude males with heterozygous calls for the rest of the</span>
        <span class="c"># analysis</span>
        <span class="n">invalid_rows</span> <span class="o">=</span> <span class="n">samples_with_hetero_calls</span><span class="p">(</span>
            <span class="n">t_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">t_data</span><span class="p">[</span><span class="n">site_info</span><span class="o">.</span><span class="n">gender_c</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dosage_columns</span><span class="p">],</span>
            <span class="n">dosage_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">invalid_rows</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;There were {:,d} males with heterozygous &quot;</span>
                            <span class="s">&quot;calls for {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">invalid_rows</span><span class="p">),</span> <span class="n">name</span><span class="p">))</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">t_data</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">invalid_rows</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">t_data</span> <span class="o">=</span> <span class="n">t_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">invalid_rows</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">t_data</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

        <span class="c"># Getting the genders</span>
        <span class="n">gender</span> <span class="o">=</span> <span class="n">t_data</span><span class="p">[</span><span class="n">site_info</span><span class="o">.</span><span class="n">gender_c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="c"># Computing the frequency</span>
    <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span> <span class="o">=</span> <span class="n">maf_from_probs</span><span class="p">(</span><span class="n">t_data</span><span class="p">[</span><span class="n">dosage_columns</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                       <span class="n">dosage_columns</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">dosage_columns</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
                                       <span class="n">gender</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>

    <span class="c"># What we want to print</span>
    <span class="n">to_return</span> <span class="o">=</span> <span class="p">[</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">allele_encoding</span><span class="p">[</span><span class="n">major</span><span class="p">],</span>
                 <span class="n">allele_encoding</span><span class="p">[</span><span class="n">minor</span><span class="p">],</span> <span class="n">maf</span><span class="p">,</span> <span class="n">t_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

    <span class="c"># If the marker is too rare, we continue with the rest</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">maf</span> <span class="o">==</span> <span class="s">&quot;NA&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">maf</span> <span class="o">&lt;</span> <span class="n">site_info</span><span class="o">.</span><span class="n">maf_t</span><span class="p">):</span>
        <span class="n">to_return</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s">&quot;NA&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">number_to_print</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">to_return</span><span class="p">)))</span>
        <span class="k">return</span> <span class="n">to_return</span>

    <span class="c"># Computing the dosage on the minor allele</span>
    <span class="n">data</span><span class="p">[</span><span class="s">&quot;_GenoD&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dosage_from_probs</span><span class="p">(</span>
        <span class="n">homo_probs</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">minor</span><span class="p">],</span>
        <span class="n">hetero_probs</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">dosage_columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span>
        <span class="n">scale</span><span class="o">=</span><span class="n">site_info</span><span class="o">.</span><span class="n">scale</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># We might need to specifically add the interaction column (if Cox and if</span>
    <span class="c"># interaction)</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;cox&quot;</span><span class="p">):</span>
        <span class="n">data</span><span class="p">[</span><span class="s">&quot;_Inter&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">_GenoD</span> <span class="o">*</span> <span class="n">data</span><span class="p">[</span><span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span><span class="p">]</span>

    <span class="c"># Removing the unwanted columns</span>
    <span class="n">unwanted_columns</span> <span class="o">=</span> <span class="n">dosage_columns</span>
    <span class="k">if</span> <span class="n">site_info</span><span class="o">.</span><span class="n">del_g</span><span class="p">:</span>
        <span class="n">unwanted_columns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">gender_c</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">unwanted_columns</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c"># The column to get the result from</span>
    <span class="n">result_from_column</span> <span class="o">=</span> <span class="s">&quot;_GenoD&quot;</span>
    <span class="k">if</span> <span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">((</span><span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span> <span class="o">==</span> <span class="n">site_info</span><span class="o">.</span><span class="n">gender_c</span><span class="p">)</span> <span class="ow">or</span>
                <span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span> <span class="ow">in</span> <span class="n">site_info</span><span class="o">.</span><span class="n">categorical</span><span class="p">)):</span>
            <span class="n">result_from_column</span> <span class="o">=</span> <span class="s">&quot;_GenoD:C({})[T.{}]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span><span class="p">,</span>
                <span class="nb">sorted</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">result_from_column</span> <span class="o">=</span> <span class="s">&quot;_GenoD:&quot;</span> <span class="o">+</span> <span class="n">site_info</span><span class="o">.</span><span class="n">inter_c</span>

        <span class="c"># Cox is the _Inter column</span>
        <span class="k">if</span> <span class="n">site_info</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;cox&quot;</span><span class="p">:</span>
            <span class="n">result_from_column</span> <span class="o">=</span> <span class="s">&quot;_Inter&quot;</span>

    <span class="c"># Fitting</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">_fit_map</span><span class="p">[</span><span class="n">site_info</span><span class="o">.</span><span class="n">analysis_type</span><span class="p">](</span>
        <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
        <span class="n">groups</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
        <span class="n">time_to_event</span><span class="o">=</span><span class="n">site_info</span><span class="o">.</span><span class="n">time_to_event</span><span class="p">,</span>
        <span class="n">event</span><span class="o">=</span><span class="n">site_info</span><span class="o">.</span><span class="n">event</span><span class="p">,</span>
        <span class="n">formula</span><span class="o">=</span><span class="n">site_info</span><span class="o">.</span><span class="n">formula</span><span class="p">,</span>
        <span class="n">result_col</span><span class="o">=</span><span class="n">result_from_column</span><span class="p">,</span>
        <span class="n">use_ml</span><span class="o">=</span><span class="n">site_info</span><span class="o">.</span><span class="n">use_ml</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># Extending the list to return</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;NA&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">site_info</span><span class="o">.</span><span class="n">number_to_print</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">to_return</span><span class="p">))</span>
    <span class="n">to_return</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">to_return</span>


<span class="k">def</span> <span class="nf">samples_with_hetero_calls</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">hetero_c</span><span class="p">):</span></div>
<div class="viewcode-block" id="samples_with_hetero_calls"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.samples_with_hetero_calls">[docs]</a>    <span class="sd">&quot;&quot;&quot;Gets male and heterozygous calls.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (pandas.DataFrame): the probability matrix</span>
<span class="sd">        hetero_c (str): the name of the heterozygous column</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.Index: samples where call is heterozygous</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">idxmax</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="n">hetero_c</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>


<span class="k">def</span> <span class="nf">get_formula</span><span class="p">(</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">covars</span><span class="p">,</span> <span class="n">interaction</span><span class="p">,</span> <span class="n">gender_c</span><span class="p">,</span> <span class="n">categorical</span><span class="p">):</span></div>
<div class="viewcode-block" id="get_formula"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.get_formula">[docs]</a>    <span class="sd">&quot;&quot;&quot;Creates the linear/logistic regression formula (for statsmodel).</span>

<span class="sd">    Args:</span>
<span class="sd">        phenotype (str): the phenotype column</span>
<span class="sd">        covars (list): the list of co variable columns</span>
<span class="sd">        interaction (str): the interaction column</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the formula for the statistical analysis</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        The phenotype column needs to be specified. The list of co variables</span>
<span class="sd">        might be empty (if no co variables are necessary). The interaction</span>
<span class="sd">        column can be set to ``None`` if there is no interaction.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        The gender column should be categorical (hence, the formula requires</span>
<span class="sd">        the gender to be included into ``C()``, *e.g.* ``C(Gender)``).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The phenotype and genetics</span>
    <span class="n">formula</span> <span class="o">=</span> <span class="s">&quot;{} ~ _GenoD&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">phenotype</span><span class="p">)</span>

    <span class="c"># Are there any covars?</span>
    <span class="k">for</span> <span class="n">covar</span> <span class="ow">in</span> <span class="n">covars</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">covar</span> <span class="o">==</span> <span class="n">gender_c</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">covar</span> <span class="ow">in</span> <span class="n">categorical</span><span class="p">):</span>
            <span class="n">covar</span> <span class="o">=</span> <span class="s">&quot;C({})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">covar</span><span class="p">)</span>
        <span class="n">formula</span> <span class="o">+=</span> <span class="s">&quot; + &quot;</span> <span class="o">+</span> <span class="n">covar</span>

    <span class="c"># Is there an interaction term?</span>
    <span class="k">if</span> <span class="n">interaction</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">interaction</span> <span class="o">==</span> <span class="n">gender_c</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">interaction</span> <span class="ow">in</span> <span class="n">categorical</span><span class="p">):</span>
            <span class="n">interaction</span> <span class="o">=</span> <span class="s">&quot;C({})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">interaction</span><span class="p">)</span>
        <span class="n">formula</span> <span class="o">+=</span> <span class="s">&quot; + _GenoD*{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">interaction</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">formula</span>


<span class="k">def</span> <span class="nf">fit_cox</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">time_to_event</span><span class="p">,</span> <span class="n">event</span><span class="p">,</span> <span class="n">result_col</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span></div>
<div class="viewcode-block" id="fit_cox"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.fit_cox">[docs]</a>    <span class="sd">&quot;&quot;&quot;Fit a Cox&#39; proportional hazard to the data.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (pandas.DataFrame): the data to analyse</span>
<span class="sd">        time_to_event (str): the time to event column for the survival analysis</span>
<span class="sd">        event (str): the event column for the survival analysis</span>
<span class="sd">        result_col (str): the column that will contain the results</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.array: the results from the survival analysis</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        The tie method used is ``Efron``. Normalization is set to ``False``.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cf</span> <span class="o">=</span> <span class="n">CoxPHFitter</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">tie_method</span><span class="o">=</span><span class="s">&quot;Efron&quot;</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">cf</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">duration_col</span><span class="o">=</span><span class="n">time_to_event</span><span class="p">,</span> <span class="n">event_col</span><span class="o">=</span><span class="n">event</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">cf</span><span class="o">.</span><span class="n">summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result_col</span><span class="p">,</span> <span class="n">_COX_REQ_COLS</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>


<span class="k">def</span> <span class="nf">fit_linear</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">formula</span><span class="p">,</span> <span class="n">result_col</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span></div>
<div class="viewcode-block" id="fit_linear"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.fit_linear">[docs]</a>    <span class="sd">&quot;&quot;&quot;Fit a linear regression to the data.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (pandas.DataFrame): the data to analyse</span>
<span class="sd">        formula (str): the formula for the linear regression</span>
<span class="sd">        result_col (str): the column that will contain the results</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: the results from the linear regression</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_get_result_from_linear_logistic_mixedlm</span><span class="p">(</span>
        <span class="n">smf</span><span class="o">.</span><span class="n">ols</span><span class="p">(</span><span class="n">formula</span><span class="o">=</span><span class="n">formula</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(),</span>
        <span class="n">result_col</span><span class="o">=</span><span class="n">result_col</span><span class="p">,</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">fit_logistic</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">formula</span><span class="p">,</span> <span class="n">result_col</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span></div>
<div class="viewcode-block" id="fit_logistic"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.fit_logistic">[docs]</a>    <span class="sd">&quot;&quot;&quot;Fit a logistic regression to the data.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (pandas.DataFrame): the data to analyse</span>
<span class="sd">        formula (str): the formula for the logistic regression</span>
<span class="sd">        result_col (str): the column that will contain the results</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: the results from the logistic regression</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_get_result_from_linear_logistic_mixedlm</span><span class="p">(</span>
        <span class="n">smf</span><span class="o">.</span><span class="n">glm</span><span class="p">(</span><span class="n">formula</span><span class="o">=</span><span class="n">formula</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
                <span class="n">family</span><span class="o">=</span><span class="n">sm</span><span class="o">.</span><span class="n">families</span><span class="o">.</span><span class="n">Binomial</span><span class="p">())</span><span class="o">.</span><span class="n">fit</span><span class="p">(),</span>
        <span class="n">result_col</span><span class="o">=</span><span class="n">result_col</span><span class="p">,</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">fit_mixedlm</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">formula</span><span class="p">,</span> <span class="n">use_ml</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="n">result_col</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span></div>
<div class="viewcode-block" id="fit_mixedlm"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.fit_mixedlm">[docs]</a>    <span class="sd">&quot;&quot;&quot;Fit a linear mixed effects model to the data.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (pandas.DataFrame): the data to analyse</span>
<span class="sd">        formula (str): the formula for the linear mixed effects model</span>
<span class="sd">        use_ml (bool): whether to use ML instead of REML</span>
<span class="sd">        groups (str): the column containing the groups</span>
<span class="sd">        result_col (str): the column that will contain the results</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: the results from the linear mixed effects model</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_get_result_from_linear_logistic_mixedlm</span><span class="p">(</span>
        <span class="n">smf</span><span class="o">.</span><span class="n">mixedlm</span><span class="p">(</span><span class="n">formula</span><span class="o">=</span><span class="n">formula</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
                    <span class="n">groups</span><span class="o">=</span><span class="n">groups</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">reml</span><span class="o">=</span><span class="ow">not</span> <span class="n">use_ml</span><span class="p">),</span>
        <span class="n">result_col</span><span class="o">=</span><span class="n">result_col</span><span class="p">,</span>
    <span class="p">)</span>


<span class="n">_fit_map</span> <span class="o">=</span> <span class="p">{</span></div>
    <span class="s">&quot;cox&quot;</span><span class="p">:</span> <span class="n">fit_cox</span><span class="p">,</span>
    <span class="s">&quot;linear&quot;</span><span class="p">:</span> <span class="n">fit_linear</span><span class="p">,</span>
    <span class="s">&quot;logistic&quot;</span><span class="p">:</span> <span class="n">fit_logistic</span><span class="p">,</span>
    <span class="s">&quot;mixedlm&quot;</span><span class="p">:</span> <span class="n">fit_mixedlm</span><span class="p">,</span>
<span class="p">}</span>


<span class="k">def</span> <span class="nf">_get_result_from_linear_logistic_mixedlm</span><span class="p">(</span><span class="n">fit_result</span><span class="p">,</span> <span class="n">result_col</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Gets results from either a linear, a logistic or a mixedlm regression.</span>

<span class="sd">    Args:</span>
<span class="sd">        fit_result (RegressionResults): the results from the regression</span>
<span class="sd">        result_col (str): the name of the result column</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: the regression results</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">conf_int</span> <span class="o">=</span> <span class="n">fit_result</span><span class="o">.</span><span class="n">conf_int</span><span class="p">()</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result_col</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">values</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">conf_int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span>
    <span class="k">return</span> <span class="p">[</span>
        <span class="n">fit_result</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="n">result_col</span><span class="p">],</span>
        <span class="n">fit_result</span><span class="o">.</span><span class="n">bse</span><span class="p">[</span><span class="n">result_col</span><span class="p">],</span>
        <span class="n">conf_int</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
        <span class="n">conf_int</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
        <span class="n">fit_result</span><span class="o">.</span><span class="n">tvalues</span><span class="p">[</span><span class="n">result_col</span><span class="p">],</span>
        <span class="n">fit_result</span><span class="o">.</span><span class="n">pvalues</span><span class="p">[</span><span class="n">result_col</span><span class="p">],</span>
    <span class="p">]</span>


<span class="k">def</span> <span class="nf">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<div class="viewcode-block" id="check_args"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.check_args">[docs]</a>    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    Args:</span>
<span class="sd">        args (argparse.Namespace): the options to verify</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        If there is a problem, a :py:class:`genipe.error.ProgramError` is</span>
<span class="sd">        raised.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Checking if we have the requirements</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;cox&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">HAS_LIFELINES</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;missing optional module: lifelines&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span> <span class="ow">in</span> <span class="p">{</span><span class="s">&quot;linear&quot;</span><span class="p">,</span> <span class="s">&quot;logistic&quot;</span><span class="p">,</span> <span class="s">&quot;mixedlm&quot;</span><span class="p">}:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">HAS_STATSMODELS</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;missing optional module: statsmodels&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;skat&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">HAS_R</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;R is not installed&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">HAS_SKAT</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;R library missing: SKAT&quot;</span><span class="p">)</span>

    <span class="c"># Checking the required input files</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">sample</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">]:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>

    <span class="c"># Checking the optional input files</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">extract_sites</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
                <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>

    <span class="c"># Checking the number of process</span>
    <span class="n">on_mac_os</span> <span class="o">=</span> <span class="n">platform</span><span class="o">.</span><span class="n">system</span><span class="p">()</span> <span class="o">==</span> <span class="s">&quot;Darwin&quot;</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">nb_process</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: invalid number of &quot;</span>
                           <span class="s">&quot;processes&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">nb_process</span><span class="p">))</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">nb_process</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">on_mac_os</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">!=</span> <span class="s">&quot;skat&quot;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;multiprocessing is not supported on Mac OS when &quot;</span>
                           <span class="s">&quot;using linear regression, logistic regression or &quot;</span>
                           <span class="s">&quot;Cox.&quot;</span><span class="p">)</span>

    <span class="c"># Checking the number of lines to read</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">nb_lines</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: invalid number of lines to &quot;</span>
                           <span class="s">&quot;read&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">nb_lines</span><span class="p">))</span>

    <span class="c"># Checking the MAF threshold</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">maf</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">maf</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: invalid MAF&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">maf</span><span class="p">))</span>

    <span class="c"># Checking the probability threshold</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">prob</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">prob</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: invalid probability &quot;</span>
                           <span class="s">&quot;threshold&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">prob</span><span class="p">))</span>

    <span class="c"># Reading all the variables in the phenotype file</span>
    <span class="n">header</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
        <span class="n">header</span> <span class="o">=</span> <span class="p">{</span><span class="n">name</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">i_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)}</span>

    <span class="c"># Checking the restricted columns</span>
    <span class="n">restricted_columns</span> <span class="o">=</span> <span class="p">{</span><span class="s">&quot;_D1&quot;</span><span class="p">,</span> <span class="s">&quot;_D2&quot;</span><span class="p">,</span> <span class="s">&quot;_D3&quot;</span><span class="p">,</span> <span class="s">&quot;_MaxD&quot;</span><span class="p">,</span> <span class="s">&quot;_GenoD&quot;</span><span class="p">,</span> <span class="s">&quot;_Inter&quot;</span><span class="p">}</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">restricted_columns</span> <span class="o">&amp;</span> <span class="n">header</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: restricted variables&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
            <span class="n">restricted_columns</span> <span class="o">&amp;</span> <span class="n">header</span><span class="p">,</span>
        <span class="p">))</span>

    <span class="c"># Checking the required columns</span>
    <span class="n">variables_to_check</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span> <span class="o">==</span> <span class="s">&quot;cox&quot;</span><span class="p">:</span>
        <span class="n">variables_to_check</span> <span class="o">=</span> <span class="p">{</span><span class="n">args</span><span class="o">.</span><span class="n">tte</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">event</span><span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">variables_to_check</span> <span class="o">=</span> <span class="p">{</span><span class="n">args</span><span class="o">.</span><span class="n">pheno_name</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">variable</span> <span class="ow">in</span> <span class="n">variables_to_check</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">variable</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: missing variable for {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                <span class="n">variable</span><span class="p">,</span>
                <span class="n">args</span><span class="o">.</span><span class="n">analysis_type</span><span class="p">,</span>
            <span class="p">))</span>

    <span class="c"># The categorical variables</span>
    <span class="n">categorical_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">categorical</span> <span class="o">!=</span> <span class="s">&quot;&quot;</span><span class="p">:</span>
        <span class="n">categorical_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">categorical</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;,&quot;</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">categorical_set</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: missing categorical value&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                <span class="n">name</span><span class="p">,</span>
            <span class="p">))</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">pheno_name</span> <span class="ow">in</span> <span class="n">categorical_set</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: should not be in categorical &quot;</span>
                               <span class="s">&quot;list&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">pheno_name</span><span class="p">))</span>
    <span class="n">args</span><span class="o">.</span><span class="n">categorical</span> <span class="o">=</span> <span class="n">categorical_set</span>

    <span class="c"># Checking the co-variables</span>
    <span class="n">covar_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">covar</span> <span class="o">!=</span> <span class="s">&quot;&quot;</span><span class="p">:</span>
        <span class="n">covar_list</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">covar</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;,&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">covar</span> <span class="ow">in</span> <span class="n">covar_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">covar</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: missing co-variable&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                <span class="n">covar</span><span class="p">,</span>
            <span class="p">))</span>
    <span class="n">args</span><span class="o">.</span><span class="n">covar</span> <span class="o">=</span> <span class="n">covar_list</span>

    <span class="c"># Checking the sample column</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">sample_column</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: {}: no such column (--sample-column)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
            <span class="n">args</span><span class="o">.</span><span class="n">sample_column</span><span class="p">,</span>
        <span class="p">))</span>

    <span class="c"># Checking the gender column (only if required)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gender_column</span> <span class="o">!=</span> <span class="s">&quot;None&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gender_column</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span>
                <span class="s">&quot;{}: {}: no such column (--gender-column)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                    <span class="n">args</span><span class="o">.</span><span class="n">gender_column</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="c"># Checking the interaction column (if required)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">interaction</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">interaction</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">header</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span>
                <span class="s">&quot;{}: {}: no such column (--interaction)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">args</span><span class="o">.</span><span class="n">pheno</span><span class="p">,</span>
                    <span class="n">args</span><span class="o">.</span><span class="n">interaction</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">interaction</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">categorical</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;{}: interaction term is categorical: the last &quot;</span>
                            <span class="s">&quot;category will used in the rsults&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="bp">True</span>


<span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span></div>
<div class="viewcode-block" id="parse_args"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.imputed_stats.parse_args">[docs]</a>    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        parser (argparse.ArgumentParser): the argument parser</span>
<span class="sd">        args (list): the list of arguments (if not taken from ``sys.argv``)</span>

<span class="sd">    Returns:</span>
<span class="sd">        argparse.Namespace: the list of options and arguments</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        The only check that is done here is by the parser itself. Values are</span>
<span class="sd">        verified later by the :py:func:`check_args` function.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Adding the version option for the main parser</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;-v&quot;</span><span class="p">,</span>
        <span class="s">&quot;--version&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;version&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%(prog)s</span><span class="s">, part of genipe version {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="c"># Creating a parent parser (for common options between analysis type)</span>
    <span class="n">p_parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">add_help</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

    <span class="c"># The parser object</span>
    <span class="n">p_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;-v&quot;</span><span class="p">,</span>
        <span class="s">&quot;--version&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;version&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%(prog)s</span><span class="s"> (part of genipe version {})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">p_parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--debug&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;set the logging level to debug&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The input files</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">p_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Input Files&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--impute2&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The output from IMPUTE2.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--sample&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The sample file (the order should be the same as in the IMPUTE2 &quot;</span>
             <span class="s">&quot;files).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--pheno&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The file containing phenotypes and co variables.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--extract-sites&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;A list of sites to extract for analysis (optional).&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The output files</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">p_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Output Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--out&quot;</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;imputed_stats&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The prefix for the output files. [</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># General options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">p_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;General Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--nb-process&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;INT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The number of process to use. [</span><span class="si">%(default)d</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--nb-lines&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;INT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The number of line to read at a time. [</span><span class="si">%(default)d</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--chrx&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The analysis is performed for the non pseudo-autosomal region &quot;</span>
             <span class="s">&quot;of the chromosome X (male dosage will be divided by 2 to get &quot;</span>
             <span class="s">&quot;values [0, 0.5] instead of [0, 1]) (males are coded as 1 and &quot;</span>
             <span class="s">&quot;option &#39;--gender-column&#39; should be used).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--gender-column&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;Gender&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The name of the gender column (use to exclude samples with &quot;</span>
             <span class="s">&quot;unknown gender (i.e. not 1, male, or 2, female). If gender not &quot;</span>
             <span class="s">&quot;available, use &#39;None&#39;. [</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The dosage options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">p_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Dosage Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--scale&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;INT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Scale dosage so that values are in [0, n] (possible values are &quot;</span>
             <span class="s">&quot;1 (no scaling) or 2). [</span><span class="si">%(default)d</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--prob&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The minimal probability for which a genotype should be &quot;</span>
             <span class="s">&quot;considered. [&gt;=</span><span class="si">%(default).1f</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--maf&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Minor allele frequency threshold for which marker will be &quot;</span>
             <span class="s">&quot;skipped. [&lt;</span><span class="si">%(default).2f</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The general phenotype options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">p_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Phenotype Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--covar&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The co variable names (in the phenotype file), separated by &quot;</span>
             <span class="s">&quot;coma.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--categorical&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The name of the variables that are categorical (note that the &quot;</span>
             <span class="s">&quot;gender is always categorical). The variables are separated by &quot;</span>
             <span class="s">&quot;coma.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--missing-value&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The missing value in the phenotype file.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--sample-column&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;sample_id&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The name of the sample ID column (in the phenotype file). &quot;</span>
             <span class="s">&quot;[</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--interaction&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Add an interaction between the genotype and this variable.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># Sub parsers</span>
    <span class="n">subparsers</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_subparsers</span><span class="p">(</span>
        <span class="n">title</span><span class="o">=</span><span class="s">&quot;Statistical Analysis Type&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s">&quot;The type of statistical analysis to be performed on the &quot;</span>
                    <span class="s">&quot;imputed data.&quot;</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s">&quot;analysis_type&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">subparsers</span><span class="o">.</span><span class="n">required</span> <span class="o">=</span> <span class="bp">True</span>

    <span class="c"># The Cox parser</span>
    <span class="n">cox_parser</span> <span class="o">=</span> <span class="n">subparsers</span><span class="o">.</span><span class="n">add_parser</span><span class="p">(</span>
        <span class="s">&quot;cox&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Cox&#39;s proportional hazard model (survival regression).&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s">&quot;Performs a survival regression on imputed data using &quot;</span>
                    <span class="s">&quot;Cox&#39;s proportional hazard model. This script is part of &quot;</span>
                    <span class="s">&quot;the &#39;genipe&#39; package, version {}).&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
        <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">p_parser</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="n">group</span> <span class="o">=</span> <span class="n">cox_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Cox&#39;s Proportional Hazard Model &quot;</span>
                                          <span class="s">&quot;Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--time-to-event&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s">&quot;tte&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The time to event variable (in the pheno file).&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--event&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The event variable (1 if observed, 0 if not observed)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The linear parser</span>
    <span class="n">lin_parser</span> <span class="o">=</span> <span class="n">subparsers</span><span class="o">.</span><span class="n">add_parser</span><span class="p">(</span>
        <span class="s">&quot;linear&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Linear regression (ordinary least squares).&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s">&quot;Performs a linear regression (ordinary least squares) on &quot;</span>
                    <span class="s">&quot;imputed data. This script is part of the &#39;genipe&#39; &quot;</span>
                    <span class="s">&quot;package, version {}).&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
        <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">p_parser</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="c"># The phenotype options for linear regression</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">lin_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Linear Regression Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--pheno-name&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The phenotype.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The logistic parser</span>
    <span class="n">logit_parser</span> <span class="o">=</span> <span class="n">subparsers</span><span class="o">.</span><span class="n">add_parser</span><span class="p">(</span>
        <span class="s">&quot;logistic&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Logistic regression (GLM with binomial distribution).&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s">&quot;Performs a logistic regression on imputed data using a &quot;</span>
                    <span class="s">&quot;GLM with a binomial distribution. This script is part of &quot;</span>
                    <span class="s">&quot;the &#39;genipe&#39; package, version {}).&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
        <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">p_parser</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="c"># The phenotype options for logistic regression</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">logit_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Logistic Regression Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--pheno-name&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The phenotype.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The mixed effect model parser</span>
    <span class="n">mixedlm_parser</span> <span class="o">=</span> <span class="n">subparsers</span><span class="o">.</span><span class="n">add_parser</span><span class="p">(</span>
        <span class="s">&quot;mixedlm&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Linear mixed effect model (random intercept).&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s">&quot;Performs a linear mixed effects regression on imputed &quot;</span>
                    <span class="s">&quot;data using a random intercept for each group. This &quot;</span>
                    <span class="s">&quot;script is part of the &#39;genipe&#39; package, version &quot;</span>
                    <span class="s">&quot;{}).&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
        <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">p_parser</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="c"># The phenotype options for the mixed effect model</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">mixedlm_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Linear Mixed Effects Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--pheno-name&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The phenotype.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--use-ml&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Fit the standard likelihood using maximum likelihood (ML) &quot;</span>
             <span class="s">&quot;estimation instead of REML (default is REML).&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The SKAT parser.</span>
    <span class="n">skat_parser</span> <span class="o">=</span> <span class="n">subparsers</span><span class="o">.</span><span class="n">add_parser</span><span class="p">(</span>
        <span class="s">&quot;skat&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;SKAT analysis.&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s">&quot;Uses the SKAT R package to analyze user defined gene &quot;</span>
                    <span class="s">&quot;sets. This script is part of the &#39;genipe&#39; package, &quot;</span>
                    <span class="s">&quot;version {}).&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
        <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">p_parser</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="c"># Additional options for SKAT analysis.</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">skat_parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;SKAT Options&quot;</span><span class="p">)</span>

    <span class="c"># The SNP set file.</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--snp-sets&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;A file indicating a snp_set and an optional weight for every &quot;</span>
             <span class="s">&quot;variant.&quot;</span>
    <span class="p">)</span>

    <span class="c"># The outcome type: discrete or continuous.</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--outcome-type&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">(</span><span class="s">&quot;continuous&quot;</span><span class="p">,</span> <span class="s">&quot;discrete&quot;</span><span class="p">),</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The variable type for the outcome. This will be passed to SKAT.&quot;</span>
    <span class="p">)</span>

    <span class="c"># SKAT-O flag.</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--skat-o&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;By default, the regular SKAT is used. Setting this flag will &quot;</span>
             <span class="s">&quot;use the SKAT-O algorithm instead.&quot;</span>
    <span class="p">)</span>

    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--pheno-name&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;NAME&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The phenotype.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>


<span class="c"># Calling the main, if necessary</span>
<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span></div>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.2.2 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" >genipe</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>