

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>genipe.tools.impute2_merger &mdash; genipe 1.1.0 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/bizstyle.js"></script>
    <link rel="top" title="genipe 1.1.0 documentation" href="../../../index.html" />
    <link rel="up" title="genipe" href="../../genipe.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.1.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" accesskey="U">genipe</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for genipe.tools.impute2_merger</h1><div class="highlight"><pre>
<span class="c"># This file is part of genipe.</span>
<span class="c">#</span>
<span class="c"># This work is licensed under the Creative Commons Attribution-NonCommercial</span>
<span class="c"># 4.0 International License. To view a copy of this license, visit</span>
<span class="c"># http://creativecommons.org/licenses/by-nc/4.0/ or send a letter to Creative</span>
<span class="c"># Commons, PO Box 1866, Mountain View, CA 94042, USA.</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">__version__</span>
<span class="kn">from</span> <span class="nn">..formats.impute2</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">..error</span> <span class="kn">import</span> <span class="n">ProgramError</span>


<span class="n">__author__</span> <span class="o">=</span> <span class="s">&quot;Louis-Philippe Lemieux Perreault&quot;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s">&quot;Copyright 2014, Beaulieu-Saucier Pharmacogenomics Centre&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s">&quot;Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)&quot;</span>


<span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_merger.main">[docs]</a>    <span class="sd">&quot;&quot;&quot;The main function.</span>

<span class="sd">    Args:</span>
<span class="sd">        args (argparse.Namespace): the arguments to be parsed (if</span>
<span class="sd">                                   :py:func:`main` is called by another</span>
<span class="sd">                                   modulel)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Creating the option parser</span>
    <span class="n">desc</span> <span class="o">=</span> <span class="p">(</span><span class="s">&quot;Concatenate IMPUTE2 output files and retrieve some &quot;</span>
            <span class="s">&quot;statistics. This script is part of the &#39;genipe&#39; package, &quot;</span>
            <span class="s">&quot;version {}.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="n">desc</span><span class="p">)</span>

    <span class="c"># Files that need closing</span>
    <span class="n">logging_fh</span> <span class="o">=</span> <span class="bp">None</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c"># Parsing the options</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

        <span class="c"># Getting the output directory (dirname of the output prefix</span>
        <span class="n">out_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">prefix</span><span class="p">)</span>

        <span class="c"># Adding the logging capability</span>
        <span class="n">log_file</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&quot;.log&quot;</span>
        <span class="n">logging_fh</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">FileHandler</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s">&quot;w&quot;</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
            <span class="n">format</span><span class="o">=</span><span class="s">&quot;[</span><span class="si">%(asctime)s</span><span class="s"> </span><span class="si">%(levelname)s</span><span class="s">] </span><span class="si">%(message)s</span><span class="s">&quot;</span><span class="p">,</span>
            <span class="n">datefmt</span><span class="o">=</span><span class="s">&quot;%Y-%m-</span><span class="si">%d</span><span class="s"> %H:%M:%S&quot;</span><span class="p">,</span>
            <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">debug</span> <span class="k">else</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
            <span class="n">handlers</span><span class="o">=</span><span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(),</span> <span class="n">logging_fh</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Logging everything into &#39;{}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">log_file</span><span class="p">))</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Program arguments: &#39;{}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:])))</span>

        <span class="c"># Checking the options</span>
        <span class="n">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

        <span class="c"># Concatenate and extract information</span>
        <span class="n">concatenate_files</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

    <span class="c"># Catching the Ctrl^C</span>
    <span class="k">except</span> <span class="ne">KeyboardInterrupt</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Cancelled by user&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="c"># Catching the ProgramError</span>
    <span class="k">except</span> <span class="n">ProgramError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">raise</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">logging_fh</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">logging_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">concatenate_files</span><span class="p">(</span><span class="n">i_filenames</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">,</span> <span class="n">real_chrom</span><span class="p">,</span> <span class="n">options</span><span class="p">):</span></div>
<div class="viewcode-block" id="concatenate_files"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_merger.concatenate_files">[docs]</a>    <span class="sd">&quot;&quot;&quot;Concatenates and extracts information from IMPUTE2 GEN file(s).</span>

<span class="sd">    Args:</span>
<span class="sd">        i_filenames (list): the list of input filenames (to concatenate)</span>
<span class="sd">        out_prefix (str): the output prefix for the output files</span>
<span class="sd">        real_chrom (str): the chromosome contained in all the input files</span>
<span class="sd">        options (argparse.Namespace): the options</span>

<span class="sd">    This function will create the following seven files:</span>

<span class="sd">    +-----------------------+-------------------------------------------------+</span>
<span class="sd">    | File name             | Description                                     |</span>
<span class="sd">    +=======================+=================================================+</span>
<span class="sd">    | ``.impute2``          | Imputation results (merged from all the input   |</span>
<span class="sd">    |                       | files).                                         |</span>
<span class="sd">    +-----------------------+-------------------------------------------------+</span>
<span class="sd">    | ``.alleles``          | Description of the reference and alternative    |</span>
<span class="sd">    |                       | allele at each sites.                           |</span>
<span class="sd">    +-----------------------+-------------------------------------------------+</span>
<span class="sd">    | ``.imputed_sites``    | List of imputed sites (excluding sites that     |</span>
<span class="sd">    |                       | were previously genotyped in the study cohort). |</span>
<span class="sd">    +-----------------------+-------------------------------------------------+</span>
<span class="sd">    | ``.completion_rates`` | Number of missing values and completion rate    |</span>
<span class="sd">    |                       | for all sites (using the probability threshold  |</span>
<span class="sd">    |                       | set by the user, where the default is higher    |</span>
<span class="sd">    |                       | and equal to 0.9).                              |</span>
<span class="sd">    +-----------------------+-------------------------------------------------+</span>
<span class="sd">    | ``.good_sites``       | List of sites which pass the completion rate    |</span>
<span class="sd">    |                       | threshold (set by the user, where the default   |</span>
<span class="sd">    |                       | is higher and equal to 0.98) using the          |</span>
<span class="sd">    |                       | probability threshold (set by the user, where   |</span>
<span class="sd">    |                       | the default is higher and equal to 0.9).        |</span>
<span class="sd">    +-----------------------+-------------------------------------------------+</span>
<span class="sd">    | ``.map``              | A map file describing the genomic location of   |</span>
<span class="sd">    |                       | all sites.                                      |</span>
<span class="sd">    +-----------------------+-------------------------------------------------+</span>
<span class="sd">    | ``.maf``              | File containing the minor allele frequency      |</span>
<span class="sd">    |                       | (along with minor allele identification) for    |</span>
<span class="sd">    |                       | all sites using the probabilitty threshold of   |</span>
<span class="sd">    |                       | 0.9. When no genotypes are available (because   |</span>
<span class="sd">    |                       | they are all below the threshold), the MAF is   |</span>
<span class="sd">    |                       | ``NA``.                                         |</span>
<span class="sd">    +-----------------------+-------------------------------------------------+</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Opening output files</span>
    <span class="n">impute2_o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.impute2&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">alleles_o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.alleles&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">imputed_sites_o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.imputed_sites&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">completion_o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.completion_rates&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">good_sites_o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.good_sites&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">map_o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.map&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">maf_o_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s">&quot;.maf&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>

    <span class="c"># Printing the headers</span>
    <span class="k">print</span><span class="p">(</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;nb_missing&quot;</span><span class="p">,</span> <span class="s">&quot;completion_rate&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span>
          <span class="nb">file</span><span class="o">=</span><span class="n">completion_o_file</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;a1&quot;</span><span class="p">,</span> <span class="s">&quot;a2&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">alleles_o_file</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s">&quot;name&quot;</span><span class="p">,</span> <span class="s">&quot;major&quot;</span><span class="p">,</span> <span class="s">&quot;minor&quot;</span><span class="p">,</span> <span class="s">&quot;maf&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">maf_o_file</span><span class="p">)</span>

    <span class="c"># The markers that were already seen</span>
    <span class="n">already_seen</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="c"># Opening the input file(s) one by one</span>
    <span class="n">chr23_par_already_warned</span> <span class="o">=</span> <span class="bp">False</span>
    <span class="k">for</span> <span class="n">i_filename</span> <span class="ow">in</span> <span class="n">i_filenames</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Working with {}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i_filename</span><span class="p">))</span>

        <span class="c"># Getting the expected number of lines from summary file</span>
        <span class="n">summary</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">i_filename</span> <span class="o">+</span> <span class="s">&quot;_summary&quot;</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="n">summary</span> <span class="o">=</span> <span class="n">i_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&quot;-Output file\n --\d+ type 0 SNPs\n --\d+ type 1 SNPs&quot;</span>
                      <span class="s">r&quot;\n --\d+ type 2 SNPs\n --\d+ type 3 SNPs\n&quot;</span>
                      <span class="s">r&quot; --(\d+) total SNPs&quot;</span><span class="p">,</span> <span class="n">summary</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">r</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: unknown &quot;</span>
                               <span class="s">&quot;format&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i_filename</span> <span class="o">+</span> <span class="s">&quot;_summary&quot;</span><span class="p">))</span>
        <span class="n">nb_expected</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;  - expecting {:,d} lines&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_expected</span><span class="p">))</span>

        <span class="c"># The input file</span>
        <span class="n">i_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">i_filename</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span>

        <span class="n">nb_line</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i_file</span><span class="p">:</span>
            <span class="c"># The number of line</span>
            <span class="n">nb_line</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c"># Splitting the line</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\r\n</span><span class="s">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot; &quot;</span><span class="p">)</span>

            <span class="c"># Gathering genotypes</span>
            <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">),</span> <span class="n">geno</span> <span class="o">=</span> <span class="n">matrix_from_line</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>

            <span class="c"># Checking the name of the marker</span>
            <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s">&quot;.&quot;</span><span class="p">:</span>
                <span class="n">name</span> <span class="o">=</span> <span class="s">&quot;{}:{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">real_chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">)</span>

            <span class="c"># Have we seen this marker?</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">already_seen</span><span class="p">:</span>
                <span class="n">new_name</span> <span class="o">=</span> <span class="s">&quot;{}_{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">already_seen</span><span class="p">[</span><span class="n">name</span><span class="p">])</span>
                <span class="k">while</span> <span class="n">new_name</span> <span class="ow">in</span> <span class="n">already_seen</span><span class="p">:</span>
                    <span class="n">already_seen</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">new_name</span> <span class="o">=</span> <span class="s">&quot;{}_{}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">already_seen</span><span class="p">[</span><span class="n">name</span><span class="p">])</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">new_name</span>
            <span class="n">already_seen</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c"># Checking the chromosome</span>
            <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s">&quot;---&quot;</span><span class="p">:</span>
                <span class="c"># This site is imputed</span>
                <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">imputed_sites_o_file</span><span class="p">)</span>

            <span class="k">elif</span> <span class="n">chrom</span> <span class="o">!=</span> <span class="n">real_chrom</span><span class="p">:</span>
                <span class="c"># The chromosome in the impute2 file is not the same as</span>
                <span class="c"># the one required in the options</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">chrom</span> <span class="o">==</span> <span class="s">&quot;23&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">real_chrom</span> <span class="o">==</span> <span class="s">&quot;25&quot;</span><span class="p">):</span>
                    <span class="c"># This might be that we are in the PAR</span>
                    <span class="c"># (pseudo-autosomal) region, so we print a warning</span>
                    <span class="c"># and continue</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">chr23_par_already_warned</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;WARNING: asked for chromosome 25, &quot;</span>
                                        <span class="s">&quot;but in chromosome 23: be sure to be &quot;</span>
                                        <span class="s">&quot;in the pseudo-autosomal region&quot;</span><span class="p">)</span>
                        <span class="n">chr23_par_already_warned</span> <span class="o">=</span> <span class="bp">True</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="c"># This is a problem, so we quit</span>
                    <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{} != {}: not same &quot;</span>
                                       <span class="s">&quot;chromosome&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">real_chrom</span><span class="p">))</span>

            <span class="c"># Adding the information to the coding file</span>
            <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">alleles_o_file</span><span class="p">)</span>

            <span class="c"># Checking the completion rate and saving it</span>
            <span class="n">good_calls</span> <span class="o">=</span> <span class="n">get_good_probs</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">options</span><span class="o">.</span><span class="n">probability</span><span class="p">)</span>
            <span class="n">nb</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">good_calls</span><span class="p">)</span>
            <span class="n">comp</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">comp</span> <span class="o">=</span> <span class="n">nb</span> <span class="o">/</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">nb</span><span class="p">,</span> <span class="n">comp</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span>
                  <span class="nb">file</span><span class="o">=</span><span class="n">completion_o_file</span><span class="p">)</span>

            <span class="c"># Computing the MAF and saving it</span>
            <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span> <span class="o">=</span> <span class="n">maf_from_probs</span><span class="p">(</span><span class="n">geno</span><span class="p">[</span><span class="n">good_calls</span><span class="p">],</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">)</span>
            <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">major</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">maf</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">maf_o_file</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">comp</span> <span class="o">&gt;=</span> <span class="n">options</span><span class="o">.</span><span class="n">completion</span><span class="p">:</span>
                <span class="c"># The completion is over the threshold</span>
                <span class="k">print</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">good_sites_o_file</span><span class="p">)</span>

            <span class="c"># Saving the map file</span>
            <span class="k">print</span><span class="p">(</span><span class="n">real_chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;0&quot;</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">map_o_file</span><span class="p">)</span>

            <span class="c"># Saving the data</span>
            <span class="k">print</span><span class="p">(</span><span class="n">real_chrom</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="mi">5</span><span class="p">:],</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot; &quot;</span><span class="p">,</span>
                  <span class="nb">file</span><span class="o">=</span><span class="n">impute2_o_file</span><span class="p">)</span>

        <span class="c"># Closing the input file</span>
        <span class="n">i_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">nb_line</span> <span class="o">!=</span> <span class="n">nb_expected</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;  - number of lines ({:,d}) is not as expected &quot;</span>
                        <span class="s">&quot;({:,d})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">nb_line</span><span class="p">,</span> <span class="n">nb_expected</span><span class="p">))</span>

    <span class="c"># Closing output files</span>
    <span class="n">impute2_o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">alleles_o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">imputed_sites_o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">completion_o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">good_sites_o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">map_o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">maf_o_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">check_args</span><span class="p">(</span><span class="n">args</span><span class="p">):</span></div>
<div class="viewcode-block" id="check_args"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_merger.check_args">[docs]</a>    <span class="sd">&quot;&quot;&quot;Checks the arguments and options.</span>

<span class="sd">    Args:</span>
<span class="sd">        args (argparse.Namespace): the options to verify</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        If there is a problem, a :py:class:`genipe.error.ProgramError` is</span>
<span class="sd">        raised.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Checking the input files</span>
    <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">impute2</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>

        <span class="n">summary_file</span> <span class="o">=</span> <span class="n">filename</span> <span class="o">+</span> <span class="s">&quot;_summary&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">summary_file</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: no such file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">summary_file</span><span class="p">))</span>

    <span class="c"># Checking the chromosome</span>
    <span class="n">valid_chromosome</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">24</span><span class="p">)]</span>
    <span class="n">valid_chromosome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&quot;25&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">chrom</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_chromosome</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">ProgramError</span><span class="p">(</span><span class="s">&quot;{}: invalid chromosome&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">chrom</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">chrom</span> <span class="o">==</span> <span class="s">&quot;23&quot;</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;MAF computation is wrong for chromosome 23 (males &quot;</span>
                        <span class="s">&quot;have 2 alleles in the computation, instead of 1)...&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="bp">True</span>


<span class="k">def</span> <span class="nf">parse_args</span><span class="p">(</span><span class="n">parser</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span></div>
<div class="viewcode-block" id="parse_args"><a class="viewcode-back" href="../../../module_content/genipe.tools.html#genipe.tools.impute2_merger.parse_args">[docs]</a>    <span class="sd">&quot;&quot;&quot;Parses the command line options and arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        parser (argparse.ArgumentParser): the argument parser</span>
<span class="sd">        args (list): the list of arguments (if not taken from ``sys.argv``)</span>

<span class="sd">    Returns:</span>
<span class="sd">        argparse.Namespace: the list of options and arguments</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>
<span class="sd">        The only check that is done here is by the parser itself. Values are</span>
<span class="sd">        verified later by the :py:func:`check_args` function.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># The parser object</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--version&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;version&quot;</span><span class="p">,</span>
        <span class="n">version</span><span class="o">=</span><span class="s">&quot;</span><span class="si">%(prog)s</span><span class="s"> (part of genipe version {})&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--debug&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;set the logging level to debug&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The input files</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Input Files&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;-i&quot;</span><span class="p">,</span>
        <span class="s">&quot;--impute2&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;GEN&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">nargs</span><span class="o">=</span><span class="s">&quot;+&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;IMPUTE2 file(s) to merge.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The options</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Options&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--chr&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;CHR&quot;</span><span class="p">,</span>
        <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">dest</span><span class="o">=</span><span class="s">&quot;chrom&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The chromosome on witch the imputation was made.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--probability&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The probability threshold for no calls. [&lt;</span><span class="si">%(default).1f</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--completion&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FLOAT&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.98</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The completion rate threshold for site exclusion. &quot;</span>
             <span class="s">&quot;[&lt;</span><span class="si">%(default).2f</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c"># The output files</span>
    <span class="n">group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s">&quot;Output Files&quot;</span><span class="p">)</span>
    <span class="n">group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s">&quot;--prefix&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">metavar</span><span class="o">=</span><span class="s">&quot;FILE&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s">&quot;imputed&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s">&quot;The prefix for the output files. [</span><span class="si">%(default)s</span><span class="s">]&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>


<span class="c"># Calling the main, if necessary</span>
<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span></div>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">genipe 1.1.0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="../../genipe.html" >genipe</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>