

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>genipe package &mdash; genipe 1.2.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.2.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/bizstyle.js"></script>
    <link rel="top" title="genipe 1.2.3 documentation" href="../index.html" />
    <link rel="up" title="genipe" href="modules.html" />
    <link rel="next" title="genipe.config package" href="genipe.config.html" />
    <link rel="prev" title="genipe" href="modules.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="genipe.config.html" title="genipe.config package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="modules.html" title="genipe"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.2.3 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../implementation.html" >Implementation</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="modules.html" accesskey="U">genipe</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">genipe package</a><ul>
<li><a class="reference internal" href="#module-genipe">Module contents</a></li>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-genipe.error">genipe.error module</a></li>
<li><a class="reference internal" href="#module-genipe.pipeline">genipe.pipeline module</a></li>
<li><a class="reference internal" href="#module-genipe.version">genipe.version module</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="modules.html"
                        title="previous chapter">genipe</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="genipe.config.html"
                        title="next chapter">genipe.config package</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/module_content/genipe.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="genipe-package">
<h1>genipe package<a class="headerlink" href="#genipe-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-genipe">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-genipe" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="genipe.test">
<code class="descclassname">genipe.</code><code class="descname">test</code><span class="sig-paren">(</span><em>verbosity=1</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe.html#test"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.test" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes all the tests for genipe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>verbosity</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#int" title="(in Jinja2 v2.8-dev)"><em>int</em></a>) &#8211; the verbosity level</td>
</tr>
</tbody>
</table>
<p>Just set <code class="docutils literal"><span class="pre">verbosity</span></code> to an integer higher than 1 to have more information
about the tests.</p>
</dd></dl>

</div>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="genipe.config.html">genipe.config package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genipe.config.html#module-genipe.config">Module contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.config.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.config.html#module-genipe.config.parser">genipe.config.parser module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="genipe.db.html">genipe.db package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genipe.db.html#module-genipe.db">Module contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.db.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.db.html#module-genipe.db.utils">genipe.db.utils module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="genipe.formats.html">genipe.formats package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genipe.formats.html#module-genipe.formats">Module contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.formats.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.formats.html#module-genipe.formats.impute2">genipe.formats.impute2 module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.formats.html#module-genipe.formats.index">genipe.formats.index module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="genipe.reporting.html">genipe.reporting package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genipe.reporting.html#module-genipe.reporting">Module contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.reporting.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.reporting.html#module-genipe.reporting.autoreport">genipe.reporting.autoreport module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.reporting.html#module-genipe.reporting.utils">genipe.reporting.utils module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="genipe.task.html">genipe.task package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genipe.task.html#module-genipe.task">Module contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.task.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.task.html#module-genipe.task.launcher">genipe.task.launcher module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="genipe.tests.html">genipe.tests package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests">Module contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.python_3_3_compatibility">genipe.tests.python_3_3_compatibility module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_db">genipe.tests.test_db module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_formats">genipe.tests.test_formats module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_impute2_extractor">genipe.tests.test_impute2_extractor module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_impute2_merger">genipe.tests.test_impute2_merger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_imputed_stats">genipe.tests.test_imputed_stats module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_launcher">genipe.tests.test_launcher module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_main_pipeline">genipe.tests.test_main_pipeline module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tests.html#module-genipe.tests.test_reporting_utils">genipe.tests.test_reporting_utils module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="genipe.tools.html">genipe.tools package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="genipe.tools.html#module-genipe.tools">Module contents</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tools.html#module-genipe.tools.impute2_extractor">genipe.tools.impute2_extractor module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tools.html#module-genipe.tools.impute2_merger">genipe.tools.impute2_merger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tools.html#module-genipe.tools.imputed_stats">genipe.tools.imputed_stats module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genipe.tools.html#module-genipe.tools.genipe_tutorial">genipe.tools.genipe_tutorial</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-genipe.error">
<span id="genipe-error-module"></span><h2>genipe.error module<a class="headerlink" href="#module-genipe.error" title="Permalink to this headline">¶</a></h2>
<dl class="exception">
<dt id="genipe.error.ProgramError">
<em class="property">exception </em><code class="descclassname">genipe.error.</code><code class="descname">ProgramError</code><span class="sig-paren">(</span><em>msg</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/error.html#ProgramError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.error.ProgramError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.4/library/exceptions.html#Exception" title="(in Python v3.4)"><code class="xref py py-class docutils literal"><span class="pre">Exception</span></code></a></p>
<p>An Exception raised in case of a problem.</p>
</dd></dl>

</div>
<div class="section" id="module-genipe.pipeline">
<span id="genipe-pipeline-module"></span><h2>genipe.pipeline module<a class="headerlink" href="#module-genipe.pipeline" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="genipe.pipeline.check_args">
<code class="descclassname">genipe.pipeline.</code><code class="descname">check_args</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#check_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.check_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the arguments and options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the arguments and options</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><cite>True</cite> if everything is OK</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/functions.html#bool" title="(in Python v3.4)">bool</a></td>
</tr>
</tbody>
</table>
<p>If an option is invalid, a <a class="reference internal" href="#genipe.error.ProgramError" title="genipe.error.ProgramError"><code class="xref py py-class docutils literal"><span class="pre">genipe.error.ProgramError</span></code></a> is raised.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.check_strand">
<code class="descclassname">genipe.pipeline.</code><code class="descname">check_strand</code><span class="sig-paren">(</span><em>prefix</em>, <em>id_suffix</em>, <em>db_name</em>, <em>options</em>, <em>exclude=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#check_strand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.check_strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the strand using SHAPEIT2.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix template of the input files</li>
<li><strong>id_suffix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the suffix of the task</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
<li><strong>exclude</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/functions.html#bool" title="(in Python v3.4)"><em>bool</em></a>) &#8211; should markers be excluded or flipped (default is
flipped)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><dl class="docutils">
<dt>statistics about the task (number of markers that will be flipped</dt>
<dd><p class="first last">or excluded)</p>
</dd>
</dl>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></p>
</td>
</tr>
</tbody>
</table>
<p>A template contains the string <code class="docutils literal"><span class="pre">{chrom}</span></code>, which will be replaced by the
chromosome number (e.g. <code class="docutils literal"><span class="pre">genipe/chr{chrom}/chr{chrom}.final</span></code> will be
replaced by <code class="docutils literal"><span class="pre">genipe/chr1/chr1.final</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.compress_impute2_files">
<code class="descclassname">genipe.pipeline.</code><code class="descname">compress_impute2_files</code><span class="sig-paren">(</span><em>filename_template</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#compress_impute2_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.compress_impute2_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges impute2 files.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>filename_template</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the template for the final IMPUTE2 file</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>A template contains the string <code class="docutils literal"><span class="pre">{chrom}</span></code>, which will be replaced by the
chromosome number (e.g. <code class="docutils literal"><span class="pre">genipe/chr{chrom}/chr{chrom}.final</span></code> will be
replaced by <code class="docutils literal"><span class="pre">genipe/chr1/chr1.final</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.compute_marker_missing_rate">
<code class="descclassname">genipe.pipeline.</code><code class="descname">compute_marker_missing_rate</code><span class="sig-paren">(</span><em>prefix</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#compute_marker_missing_rate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.compute_marker_missing_rate" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute (using Plink) marker missing rate.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix of the input file</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the missing rate for each site (results from Plink)</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.18.0)">pandas.DataFrame</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.exclude_markers_before_phasing">
<code class="descclassname">genipe.pipeline.</code><code class="descname">exclude_markers_before_phasing</code><span class="sig-paren">(</span><em>prefix</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#exclude_markers_before_phasing"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.exclude_markers_before_phasing" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds and excludes ambiguous markers (A/T and G/C) or duplicated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix of the input files</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">information about the data set</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></p>
</td>
</tr>
</tbody>
</table>
<p>If required, the function uses <a class="reference internal" href="#genipe.pipeline.is_reversed" title="genipe.pipeline.is_reversed"><code class="xref py py-func docutils literal"><span class="pre">is_reversed()</span></code></a> to check if a marker
is on the reverse strand and needs flipping.</p>
<p>The information returned includes the initial number of markers and
samples, the number of ambiguous, duplicated and non-autosomal markers,
along with the number of markers to flip if the reference was checked.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.file_sorter">
<code class="descclassname">genipe.pipeline.</code><code class="descname">file_sorter</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#file_sorter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.file_sorter" title="Permalink to this definition">¶</a></dt>
<dd><p>Helps in filename sorting.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the file to compare while sorting</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="docutils">
<dt>a tuple containing three elements: chromosome (int), start (int)</dt>
<dd>and end (int)</dd>
</dl>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#tuple" title="(in Python v3.4)">tuple</a></td>
</tr>
</tbody>
</table>
<p>Using a regular expression, finds the chromosome along with the starting
and ending position of an imputed segment. The file
<code class="docutils literal"><span class="pre">chr22.1_50000.impute2</span></code> should return <code class="docutils literal"><span class="pre">(22,</span> <span class="pre">1,</span> <span class="pre">50000)</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.final_exclusion">
<code class="descclassname">genipe.pipeline.</code><code class="descname">final_exclusion</code><span class="sig-paren">(</span><em>prefix</em>, <em>to_exclude</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#final_exclusion"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.final_exclusion" title="Permalink to this definition">¶</a></dt>
<dd><p>Flip markers.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix template of the input files</li>
<li><strong>to_exclude</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the file containing the markers to
exclude</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the number of remaining markers for phasing</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></p>
</td>
</tr>
</tbody>
</table>
<p>A template contains the string <code class="docutils literal"><span class="pre">{chrom}</span></code>, which will be replaced by the
chromosome number (e.g. <code class="docutils literal"><span class="pre">genipe/chr{chrom}/chr{chrom}.final</span></code> will be
replaced by <code class="docutils literal"><span class="pre">genipe/chr1/chr1.final</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.flip_markers">
<code class="descclassname">genipe.pipeline.</code><code class="descname">flip_markers</code><span class="sig-paren">(</span><em>prefix</em>, <em>to_flip</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#flip_markers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.flip_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Flip markers.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix template of the input files</li>
<li><strong>to_flip</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the file containing markers to flip</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>A template contains the string <code class="docutils literal"><span class="pre">{chrom}</span></code>, which will be replaced by the
chromosome number (e.g. <code class="docutils literal"><span class="pre">genipe/chr{chrom}/chr{chrom}.final</span></code> will be
replaced by <code class="docutils literal"><span class="pre">genipe/chr1/chr1.final</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.gather_execution_time">
<code class="descclassname">genipe.pipeline.</code><code class="descname">gather_execution_time</code><span class="sig-paren">(</span><em>db_name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#gather_execution_time"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.gather_execution_time" title="Permalink to this definition">¶</a></dt>
<dd><p>Gather all the execution times.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the execution time for all tasks</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.gather_imputation_stats">
<code class="descclassname">genipe.pipeline.</code><code class="descname">gather_imputation_stats</code><span class="sig-paren">(</span><em>prob_t</em>, <em>completion_t</em>, <em>info_t</em>, <em>nb_samples</em>, <em>missing</em>, <em>o_dir</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#gather_imputation_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.gather_imputation_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Gathers imputation statistics from the merged dataset.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prob_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the probability threshold (&gt;= t)</li>
<li><strong>completion_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the completion threshold (&gt;= t)</li>
<li><strong>info_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the information threshold (&gt;= t)</li>
<li><strong>nb_samples</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#int" title="(in Jinja2 v2.8-dev)"><em>int</em></a>) &#8211; the number of samples</li>
<li><strong>missing</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.18.0)"><em>pandas.DataFrame</em></a>) &#8211; the missing rate</li>
<li><strong>o_dir</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the output directory</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">imputation statistics</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></p>
</td>
</tr>
</tbody>
</table>
<p>The information returned includes the following (all values are formated in
strings):</p>
<table border="1" class="docutils">
<colgroup>
<col width="44%" />
<col width="56%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Information</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">prob_threshold</span></code></td>
<td>the probability threshold used (&gt;= t)</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_imputed</span></code></td>
<td>the number of imputed sites</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">average_comp_rate</span></code></td>
<td>the average completion rate</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">rate_threshold</span></code></td>
<td>the completion rate threshold (&gt;= t)</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">info_threshold</span></code></td>
<td>the information threshold (&gt;= t)</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_good_sites</span></code></td>
<td>the number of &#8220;good&#8221; sites (that pass
the different probability and
completion thresholds)</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">pct_good_sites</span></code></td>
<td>the percentage of &#8220;good&#8221; sites (that
pass the different probability and
completion thresholds)</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">average_comp_rate_cleaned</span></code></td>
<td>the average completion rate of &#8220;good&#8221;
sites</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">mean_missing</span></code></td>
<td>the mean number of missing genotypes
per sample (according to the
probability and completion thresholds)</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_samples</span></code></td>
<td>the total number of samples</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">nb_genotyped</span></code></td>
<td>the number of genotyped sites in the
original dataset that are included in
the imputed dataset</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_genotyped_not_complete</span></code></td>
<td>The number of original markers with a
completion rate lower than 100%</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">pct_genotyped_not_complete</span></code></td>
<td>The percentage of original markers
with a completion rate lower than 100%</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_geno_now_complete</span></code></td>
<td>The number of original (genotyped)
markers which are now 100% complete
after imputation</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">nb_missing_geno</span></code></td>
<td>The number of missing genotypes</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">pct_geno_now_complete</span></code></td>
<td>The percentage of now complete
genotypes (according to the number of
previously missing ones)</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">nb_site_now_complete</span></code></td>
<td>The number of previously genotyped
markers that are now 100% complete
after imputation</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.gather_maf_stats">
<code class="descclassname">genipe.pipeline.</code><code class="descname">gather_maf_stats</code><span class="sig-paren">(</span><em>o_dir</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#gather_maf_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.gather_maf_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Gather minor allele frequencies from imputation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>o_dir</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the output directory</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">frequency information</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></td>
</tr>
</tbody>
</table>
<p>The following information about minor allele frequencies (MAF) are
returned (all values are formatted in strings):</p>
<table border="1" class="docutils">
<colgroup>
<col width="36%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Information</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">nb_maf_nan</span></code></td>
<td>the number of markers with invalid MAF</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_marker_with_maf</span></code></td>
<td>the number of markers with valid MAF</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">nb_maf_geq_01</span></code></td>
<td>the number of markers with MAF &gt;= 1%</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_maf_geq_05</span></code></td>
<td>the number of markers with MAF &gt;= 5%</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">nb_maf_lt_05</span></code></td>
<td>the number of markers with MAF &lt; 5%</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">nb_maf_lt_01</span></code></td>
<td>the number of markers with MAF &lt; 1%</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">nb_maf_geq_01_lt_05</span></code></td>
<td>the number of markers with 1% &gt;= MAF &lt; 5%</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">pct_maf_geq_01</span></code></td>
<td>the percentage of markers with MAF &gt;= 1%</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">pct_maf_geq_05</span></code></td>
<td>the percentage of markers with MAF &gt;= 5%</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">pct_maf_lt_05</span></code></td>
<td>the percentage of markers with MAF &lt; 5%</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">pct_maf_lt_01</span></code></td>
<td>the percentage of markers with MAF &lt; 1%</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">pct_maf_geq_01_lt_05</span></code></td>
<td>the percentage of markers with
1% &gt;= MAF &lt; 5%</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">frequency_barh</span></code></td>
<td>the name of the file containing the bar plot
(if it was created)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.get_chrom_encoding">
<code class="descclassname">genipe.pipeline.</code><code class="descname">get_chrom_encoding</code><span class="sig-paren">(</span><em>reference</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#get_chrom_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.get_chrom_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the chromosome&#8217;s encoding (e.g. 1 vs chr1, X vs 23, etc).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>reference</strong> (<em>pyfaidx.Fasta</em>) &#8211; the reference</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the chromosome encoding</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></td>
</tr>
</tbody>
</table>
<p>The encoding is a dictionary where numerical chromosomes from 1 to 22 are
the keys, and the encoded chromosomes (the one present in the reference)
are the values. An example would be <code class="docutils literal"><span class="pre">1</span> <span class="pre">-&gt;</span> <span class="pre">chr1</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.get_chromosome_length">
<code class="descclassname">genipe.pipeline.</code><code class="descname">get_chromosome_length</code><span class="sig-paren">(</span><em>out_dir</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#get_chromosome_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.get_chromosome_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the chromosome length from Ensembl REST API.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>out_dir</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the output directory</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the chromosome length (chr -&gt; length)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.get_cross_validation_results">
<code class="descclassname">genipe.pipeline.</code><code class="descname">get_cross_validation_results</code><span class="sig-paren">(</span><em>glob_pattern</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#get_cross_validation_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.get_cross_validation_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a weighted mean for each chromosome for cross-validation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>glob_pattern</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the pattern used to find files using <a class="reference external" href="https://docs.python.org/3.4/library/glob.html#module-glob" title="(in Python v3.4)"><code class="xref py py-mod docutils literal"><span class="pre">glob</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">weighted mean by chromosome (cross-validation)</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></td>
</tr>
</tbody>
</table>
<p>The returned information includes the total number of genotyped tested, the
total number of genotyped by chromosome, the two summary tables produced by
IMPUTE2 (but using weighted means) for all chromosomes, and the two summary
tables for each chromosome.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.get_impute2_version">
<code class="descclassname">genipe.pipeline.</code><code class="descname">get_impute2_version</code><span class="sig-paren">(</span><em>binary</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#get_impute2_version"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.get_impute2_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the IMPUTE2 version from the binary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>binary</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the IMPUTE2 binary</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the version of the IMPUTE2 software</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)">str</a></td>
</tr>
</tbody>
</table>
<p>This function uses <a class="reference external" href="https://docs.python.org/3.4/library/subprocess.html#subprocess.Popen" title="(in Python v3.4)"><code class="xref py py-class docutils literal"><span class="pre">subprocess.Popen</span></code></a> to gather the version of
the IMPUTE2 binary. Since executing the software to gather the version
creates output files, they are deleted.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This function only works as long as the version is returned as
<code class="docutils literal"><span class="pre">IMPUTE</span> <span class="pre">version</span> <span class="pre">NNN</span></code> (where <code class="docutils literal"><span class="pre">NNN</span></code> is the version), since we use
regular expression to extract the version number from the standard
output of the software.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.get_plink_version">
<code class="descclassname">genipe.pipeline.</code><code class="descname">get_plink_version</code><span class="sig-paren">(</span><em>binary</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#get_plink_version"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.get_plink_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the Plink version from the binary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>binary</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the Plink binary</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the version of the Plink software</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)">str</a></td>
</tr>
</tbody>
</table>
<p>This function uses <a class="reference external" href="https://docs.python.org/3.4/library/subprocess.html#subprocess.Popen" title="(in Python v3.4)"><code class="xref py py-class docutils literal"><span class="pre">subprocess.Popen</span></code></a> to gather the version of
the Plink binary. Since executing the software to gather the version
creates an output file, it is deleted.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This function only works as long as the version is returned as
<code class="docutils literal"><span class="pre">|</span> <span class="pre">PLINK!</span> <span class="pre">|</span> <span class="pre">NNN</span> <span class="pre">|</span></code> (where, <code class="docutils literal"><span class="pre">NNN</span></code> is the version), since we use
regular expresion to extract the version number from the standard
output of the software.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.get_shapeit_version">
<code class="descclassname">genipe.pipeline.</code><code class="descname">get_shapeit_version</code><span class="sig-paren">(</span><em>binary</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#get_shapeit_version"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.get_shapeit_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the SHAPEIT version from the binary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>binary</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the SHAPEIT binary</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the version of the SHAPEIT software</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)">str</a></td>
</tr>
</tbody>
</table>
<p>This function uses <a class="reference external" href="https://docs.python.org/3.4/library/subprocess.html#subprocess.Popen" title="(in Python v3.4)"><code class="xref py py-class docutils literal"><span class="pre">subprocess.Popen</span></code></a> to gather the version of
the SHAPEIT binary.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This function only works as long as the version is returned as
<code class="docutils literal"><span class="pre">Version:</span> <span class="pre">NNN</span></code> (where <code class="docutils literal"><span class="pre">NNN</span></code> is the version), since we use regular
expression to extract the version number from the standard output of
the software.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.impute_markers">
<code class="descclassname">genipe.pipeline.</code><code class="descname">impute_markers</code><span class="sig-paren">(</span><em>phased_haplotypes</em>, <em>out_prefix</em>, <em>chrom_length</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#impute_markers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.impute_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Imputes the markers using IMPUTE2.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>phased_haplotypes</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the template for the haplotype files</li>
<li><strong>out_prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix template of the output files</li>
<li><strong>chrom_length</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)"><em>dict</em></a>) &#8211; the length of each chromosome</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>A template contains the string <code class="docutils literal"><span class="pre">{chrom}</span></code>, which will be replaced by the
chromosome number (e.g. <code class="docutils literal"><span class="pre">genipe/chr{chrom}/chr{chrom}.final</span></code> will be
replaced by <code class="docutils literal"><span class="pre">genipe/chr1/chr1.final</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.is_reversed">
<code class="descclassname">genipe.pipeline.</code><code class="descname">is_reversed</code><span class="sig-paren">(</span><em>chrom</em>, <em>pos</em>, <em>a1</em>, <em>a2</em>, <em>reference</em>, <em>encoding</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#is_reversed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.is_reversed" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the strand using a reference, returns False if problem.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chrom</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the chromosome</li>
<li><strong>pos</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#int" title="(in Jinja2 v2.8-dev)"><em>int</em></a>) &#8211; the position</li>
<li><strong>a1</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the first allele</li>
<li><strong>a2</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the second allele</li>
<li><strong>reference</strong> (<em>pyfaidx.Fasta</em>) &#8211; the reference</li>
<li><strong>encoding</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)"><em>dict</em></a>) &#8211; the chromosome encoding in the reference</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><dl class="docutils">
<dt><code class="docutils literal"><span class="pre">True</span></code> if it&#8217;s the complement, <code class="docutils literal"><span class="pre">False</span></code> otherwise (also</dt>
<dd><p class="first last">returns <code class="docutils literal"><span class="pre">False</span></code> if there was a problem)</p>
</dd>
</dl>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.4/library/functions.html#bool" title="(in Python v3.4)">bool</a></p>
</td>
</tr>
</tbody>
</table>
<p>The encoding (used to encode chromosome to search in the reference) is the
dictionary returned by the <a class="reference internal" href="#genipe.pipeline.get_chrom_encoding" title="genipe.pipeline.get_chrom_encoding"><code class="xref py py-func docutils literal"><span class="pre">get_chrom_encoding()</span></code></a> function.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.main">
<code class="descclassname">genipe.pipeline.</code><code class="descname">main</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.main" title="Permalink to this definition">¶</a></dt>
<dd><p>The main function of the genome-wide imputation pipeline.</p>
<p>These are the pipeline steps:</p>
<ol class="arabic simple">
<li>Find markers with strand issues and flip them (Plink)</li>
<li>Exclude markers that have ambiguous alleles [A/T or C/G] (Plink)</li>
<li>Exclude duplicated markers (only keep one) (Plink)</li>
<li>Split the dataset by chromosome (Plink)</li>
<li>Find markers that need to be flipped (strand problem) (SHAPEIT)</li>
<li>Flip those markers (Plink)</li>
<li>Find markers with strand problem (SHAPEIT)</li>
<li>Exclude markers with strand problem (Plink)</li>
<li>Phase using SHAPEIT</li>
<li>Impute using IMPUTE2</li>
<li>Merge IMPUTE2 files</li>
<li>If asked by the user, compress the final IMPUTE2 files (bgzip)</li>
</ol>
<p>At the end of the pipeline, a report (LaTeX) is automatically generated. It
includes different quality statistics and imputation metrics.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.merge_impute2_files">
<code class="descclassname">genipe.pipeline.</code><code class="descname">merge_impute2_files</code><span class="sig-paren">(</span><em>in_glob</em>, <em>o_prefix</em>, <em>probability_t</em>, <em>completion_t</em>, <em>info_t</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#merge_impute2_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.merge_impute2_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges impute2 files.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>in_glob</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the template that will be used to find files with the
<a class="reference external" href="https://docs.python.org/3.4/library/glob.html#module-glob" title="(in Python v3.4)"><code class="xref py py-mod docutils literal"><span class="pre">glob</span></code></a> module</li>
<li><strong>o_prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix template of the output files</li>
<li><strong>probability_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the probability threshold to use</li>
<li><strong>completion_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the completion threshold to use</li>
<li><strong>info_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the info threshold to use</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>A template contains the string <code class="docutils literal"><span class="pre">{chrom}</span></code>, which will be replaced by the
chromosome number (e.g. <code class="docutils literal"><span class="pre">genipe/chr{chrom}/chr{chrom}.final</span></code> will be
replaced by <code class="docutils literal"><span class="pre">genipe/chr1/chr1.final</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.parse_args">
<code class="descclassname">genipe.pipeline.</code><code class="descname">parse_args</code><span class="sig-paren">(</span><em>parser</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#parse_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.parse_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the command line options and arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.ArgumentParser" title="(in Python v3.4)"><em>argparse.ArgumentParser</em></a>) &#8211; the parser object</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the parsed options and arguments</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)">argparse.Namespace</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.phase_markers">
<code class="descclassname">genipe.pipeline.</code><code class="descname">phase_markers</code><span class="sig-paren">(</span><em>prefix</em>, <em>o_prefix</em>, <em>db_name</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#phase_markers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.phase_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Phase markers using shapeit.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix template of the input files</li>
<li><strong>o_prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the prefix template of the output files</li>
<li><strong>db_name</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the DB saving tasks&#8217; information</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the pipeline options</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the list of all samples that were phased</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)">list</a></p>
</td>
</tr>
</tbody>
</table>
<p>A template contains the string <code class="docutils literal"><span class="pre">{chrom}</span></code>, which will be replaced by the
chromosome number (e.g. <code class="docutils literal"><span class="pre">genipe/chr{chrom}/chr{chrom}.final</span></code> will be
replaced by <code class="docutils literal"><span class="pre">genipe/chr1/chr1.final</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.pipeline.read_preamble">
<code class="descclassname">genipe.pipeline.</code><code class="descname">read_preamble</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/pipeline.html#read_preamble"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.pipeline.read_preamble" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the preamble file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the preamble file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the preamble</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-genipe.version">
<span id="genipe-version-module"></span><h2>genipe.version module<a class="headerlink" href="#module-genipe.version" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="genipe.config.html" title="genipe.config package"
             >next</a> |</li>
        <li class="right" >
          <a href="modules.html" title="genipe"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.2.3 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../implementation.html" >Implementation</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="modules.html" >genipe</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.
    </div>
  </body>
</html>