

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>genipe.tools package &mdash; genipe 1.2.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/bizstyle.js"></script>
    <link rel="top" title="genipe 1.2.1 documentation" href="../index.html" />
    <link rel="up" title="genipe package" href="genipe.html" />
    <link rel="prev" title="genipe.tests package" href="genipe.tests.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="genipe.tests.html" title="genipe.tests package"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.2.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../implementation.html" >Implementation</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="modules.html" >genipe</a> &raquo;</li>
          <li class="nav-item nav-item-3"><a href="genipe.html" accesskey="U">genipe package</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">genipe.tools package</a><ul>
<li><a class="reference internal" href="#module-genipe.tools">Module contents</a></li>
<li><a class="reference internal" href="#module-genipe.tools.impute2_extractor">genipe.tools.impute2_extractor module</a></li>
<li><a class="reference internal" href="#module-genipe.tools.impute2_merger">genipe.tools.impute2_merger module</a></li>
<li><a class="reference internal" href="#module-genipe.tools.imputed_stats">genipe.tools.imputed_stats module</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="genipe.tests.html"
                        title="previous chapter">genipe.tests package</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/module_content/genipe.tools.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="genipe-tools-package">
<h1>genipe.tools package<a class="headerlink" href="#genipe-tools-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-genipe.tools">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-genipe.tools" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-genipe.tools.impute2_extractor">
<span id="genipe-tools-impute2-extractor-module"></span><h2>genipe.tools.impute2_extractor module<a class="headerlink" href="#module-genipe.tools.impute2_extractor" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="genipe.tools.impute2_extractor.check_args">
<code class="descclassname">genipe.tools.impute2_extractor.</code><code class="descname">check_args</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_extractor.html#check_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_extractor.check_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the arguments and options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the options to verify</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If there is a problem, a <a class="reference internal" href="genipe.html#genipe.error.ProgramError" title="genipe.error.ProgramError"><code class="xref py py-class docutils literal"><span class="pre">genipe.error.ProgramError</span></code></a> is
raised.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_extractor.extract_markers">
<code class="descclassname">genipe.tools.impute2_extractor.</code><code class="descname">extract_markers</code><span class="sig-paren">(</span><em>i_filenames</em>, <em>to_extract</em>, <em>out_prefix</em>, <em>out_format</em>, <em>prob_t</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_extractor.html#extract_markers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_extractor.extract_markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts according to names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>i_filenames</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the list of input file names</li>
<li><strong>to_extract</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)"><em>dict</em></a>) &#8211; the list of markers to extract for each input file</li>
<li><strong>out_prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the output prefix</li>
<li><strong>out_format</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the output format(s)</li>
<li><strong>prob_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the probability threshold</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_extractor.gather_extraction">
<code class="descclassname">genipe.tools.impute2_extractor.</code><code class="descname">gather_extraction</code><span class="sig-paren">(</span><em>i_filenames</em>, <em>maf</em>, <em>rate</em>, <em>info</em>, <em>extract_filename</em>, <em>genomic_range</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_extractor.html#gather_extraction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_extractor.gather_extraction" title="Permalink to this definition">¶</a></dt>
<dd><p>Gather positions that are required.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>i_filenames</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the list of input files</li>
<li><strong>maf</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the minor allele frequency threshold (might be <code class="docutils literal"><span class="pre">None</span></code>)</li>
<li><strong>rate</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the call rate threshold (might be <code class="docutils literal"><span class="pre">None</span></code>)</li>
<li><strong>info</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the marker information value threshold (might be
<code class="docutils literal"><span class="pre">None</span></code>)</li>
<li><strong>extract_filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the file containing marker names to
extract (might be <code class="docutils literal"><span class="pre">None</span></code>)</li>
<li><strong>genomic_range</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the genomic range for extraction</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the list of markers to extract for each input file</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)">dict</a></p>
</td>
</tr>
</tbody>
</table>
<p>If extraction by marker name is required, only those markers will be
extracted. Otherwise, <code class="docutils literal"><span class="pre">maf</span></code>, <code class="docutils literal"><span class="pre">rate</span></code>, <code class="docutils literal"><span class="pre">info</span></code> or <code class="docutils literal"><span class="pre">genomic_range</span></code> can
be specified (alone or together) to extract markers according to minor
allele frequency, call rate and genomic location.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_extractor.get_file_prefix">
<code class="descclassname">genipe.tools.impute2_extractor.</code><code class="descname">get_file_prefix</code><span class="sig-paren">(</span><em>fn</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_extractor.html#get_file_prefix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_extractor.get_file_prefix" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the filename prefix.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fn</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the file from which the prefix is required</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the prefix of the file</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)">str</a></td>
</tr>
</tbody>
</table>
<p>This function removes the extension from the file name, and return its
prefix (<em>e.g.</em> <code class="docutils literal"><span class="pre">test.impute2</span></code> returns <code class="docutils literal"><span class="pre">test</span></code>, and
<code class="docutils literal"><span class="pre">../test.impute2.gz</span></code> returns <code class="docutils literal"><span class="pre">../test</span></code>).</p>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_extractor.main">
<code class="descclassname">genipe.tools.impute2_extractor.</code><code class="descname">main</code><span class="sig-paren">(</span><em>args=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_extractor.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_extractor.main" title="Permalink to this definition">¶</a></dt>
<dd><p>The main function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the arguments to be parsed (if
<a class="reference internal" href="#genipe.tools.impute2_extractor.main" title="genipe.tools.impute2_extractor.main"><code class="xref py py-func docutils literal"><span class="pre">main()</span></code></a> is called by another
modulel)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_extractor.parse_args">
<code class="descclassname">genipe.tools.impute2_extractor.</code><code class="descname">parse_args</code><span class="sig-paren">(</span><em>parser</em>, <em>args=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_extractor.html#parse_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_extractor.parse_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the command line options and arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.ArgumentParser" title="(in Python v3.4)"><em>argparse.ArgumentParser</em></a>) &#8211; the argument parser</li>
<li><strong>args</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the list of arguments (if not taken from <code class="docutils literal"><span class="pre">sys.argv</span></code>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the list of options and arguments</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)">argparse.Namespace</a></p>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The only check that is done here is by the parser itself. Values are
verified later by the <a class="reference internal" href="#genipe.tools.impute2_extractor.check_args" title="genipe.tools.impute2_extractor.check_args"><code class="xref py py-func docutils literal"><span class="pre">check_args()</span></code></a> function.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_extractor.print_data">
<code class="descclassname">genipe.tools.impute2_extractor.</code><code class="descname">print_data</code><span class="sig-paren">(</span><em>o_files</em>, <em>prob_t</em>, <em>*</em>, <em>line=None</em>, <em>row=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_extractor.html#print_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_extractor.print_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints an impute2 line.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>o_files</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#dict" title="(in Jinja2 v2.8-dev)"><em>dict</em></a>) &#8211; the output files</li>
<li><strong>prob_t</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#float" title="(in Jinja2 v2.8-dev)"><em>float</em></a>) &#8211; the probability threshold</li>
<li><strong>line</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the impute2 line</li>
<li><strong>row</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the impute2 line, split by spaces</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-genipe.tools.impute2_merger">
<span id="genipe-tools-impute2-merger-module"></span><h2>genipe.tools.impute2_merger module<a class="headerlink" href="#module-genipe.tools.impute2_merger" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="genipe.tools.impute2_merger.check_args">
<code class="descclassname">genipe.tools.impute2_merger.</code><code class="descname">check_args</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_merger.html#check_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_merger.check_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the arguments and options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the options to verify</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If there is a problem, a <a class="reference internal" href="genipe.html#genipe.error.ProgramError" title="genipe.error.ProgramError"><code class="xref py py-class docutils literal"><span class="pre">genipe.error.ProgramError</span></code></a> is
raised.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_merger.concatenate_files">
<code class="descclassname">genipe.tools.impute2_merger.</code><code class="descname">concatenate_files</code><span class="sig-paren">(</span><em>i_filenames</em>, <em>out_prefix</em>, <em>real_chrom</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_merger.html#concatenate_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_merger.concatenate_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Concatenates and extracts information from IMPUTE2 GEN file(s).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>i_filenames</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the list of input filenames (to concatenate)</li>
<li><strong>out_prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the output prefix for the output files</li>
<li><strong>real_chrom</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the chromosome contained in all the input files</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the options</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>This function will create the following seven files:</p>
<table border="1" class="docutils">
<colgroup>
<col width="32%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">File name</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.impute2</span></code></td>
<td>Imputation results (merged from all the input
files).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.alleles</span></code></td>
<td>Description of the reference and alternative
allele at each sites.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.imputed_sites</span></code></td>
<td>List of imputed sites (excluding sites that
were previously genotyped in the study cohort).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.impute2_info</span></code></td>
<td>SNP-wise information file with one line per SNP
and a single header line at the beginning.</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.completion_rates</span></code></td>
<td>Number of missing values and completion rate
for all sites (using the probability threshold
set by the user, where the default is higher
and equal to 0.9).</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.good_sites</span></code></td>
<td>List of sites which pass the completion rate
threshold (set by the user, where the default
is higher and equal to 0.98) using the
probability threshold (set by the user, where
the default is higher and equal to 0.9).</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">.map</span></code></td>
<td>A map file describing the genomic location of
all sites.</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">.maf</span></code></td>
<td>File containing the minor allele frequency
(along with minor allele identification) for
all sites using the probabilitty threshold of
0.9. When no genotypes are available (because
they are all below the threshold), the MAF is
<code class="docutils literal"><span class="pre">NA</span></code>.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_merger.main">
<code class="descclassname">genipe.tools.impute2_merger.</code><code class="descname">main</code><span class="sig-paren">(</span><em>args=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_merger.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_merger.main" title="Permalink to this definition">¶</a></dt>
<dd><p>The main function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the arguments to be parsed (if
<a class="reference internal" href="#genipe.tools.impute2_merger.main" title="genipe.tools.impute2_merger.main"><code class="xref py py-func docutils literal"><span class="pre">main()</span></code></a> is called by another
modulel)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.impute2_merger.parse_args">
<code class="descclassname">genipe.tools.impute2_merger.</code><code class="descname">parse_args</code><span class="sig-paren">(</span><em>parser</em>, <em>args=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/impute2_merger.html#parse_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.impute2_merger.parse_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the command line options and arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.ArgumentParser" title="(in Python v3.4)"><em>argparse.ArgumentParser</em></a>) &#8211; the argument parser</li>
<li><strong>args</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the list of arguments (if not taken from <code class="docutils literal"><span class="pre">sys.argv</span></code>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the list of options and arguments</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)">argparse.Namespace</a></p>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The only check that is done here is by the parser itself. Values are
verified later by the <a class="reference internal" href="#genipe.tools.impute2_merger.check_args" title="genipe.tools.impute2_merger.check_args"><code class="xref py py-func docutils literal"><span class="pre">check_args()</span></code></a> function.</p>
</div>
</dd></dl>

</div>
<div class="section" id="module-genipe.tools.imputed_stats">
<span id="genipe-tools-imputed-stats-module"></span><h2>genipe.tools.imputed_stats module<a class="headerlink" href="#module-genipe.tools.imputed_stats" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="genipe.tools.imputed_stats.check_args">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">check_args</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#check_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.check_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the arguments and options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the options to verify</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If there is a problem, a <a class="reference internal" href="genipe.html#genipe.error.ProgramError" title="genipe.error.ProgramError"><code class="xref py py-class docutils literal"><span class="pre">genipe.error.ProgramError</span></code></a> is
raised.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.compute_statistics">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">compute_statistics</code><span class="sig-paren">(</span><em>impute2_filename</em>, <em>samples</em>, <em>markers_to_extract</em>, <em>phenotypes</em>, <em>remove_gender</em>, <em>out_prefix</em>, <em>options</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#compute_statistics"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.compute_statistics" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses IMPUTE2 file while computing statistics.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>impute2_filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the input file</li>
<li><strong>samples</strong> (<em>pandas.DataFrame</em>) &#8211; the list of samples</li>
<li><strong>markers_to_extract</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#set" title="(in Python v3.4)"><em>set</em></a>) &#8211; the set of markers to extract</li>
<li><strong>phenotypes</strong> (<em>pandas.DataFrame</em>) &#8211; the phenotypes</li>
<li><strong>remove_gender</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/functions.html#bool" title="(in Python v3.4)"><em>bool</em></a>) &#8211; whether or not to remove the gender column</li>
<li><strong>out_prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the output prefix</li>
<li><strong>options</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the options</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>This function takes care of parallelism. It reads the Impute2 file and
fills a queue that will trigger the analysis when full.</p>
<p>If the number of process to launch is 1, the rows are analyzed as they
come.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.fit_cox">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">fit_cox</code><span class="sig-paren">(</span><em>data</em>, <em>time_to_event</em>, <em>event</em>, <em>result_col</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#fit_cox"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.fit_cox" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a Cox&#8217; proportional hazard to the data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>pandas.DataFrame</em>) &#8211; the data to analyse</li>
<li><strong>time_to_event</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the time to event column for the survival analysis</li>
<li><strong>event</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the event column for the survival analysis</li>
<li><strong>result_col</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the column that will contain the results</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the results from the survival analysis</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html#numpy.array" title="(in NumPy v1.9)">numpy.array</a></p>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The tie method used is <code class="docutils literal"><span class="pre">Efron</span></code>. Normalization is set to <code class="docutils literal"><span class="pre">False</span></code>.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.fit_linear">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">fit_linear</code><span class="sig-paren">(</span><em>data</em>, <em>formula</em>, <em>result_col</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#fit_linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.fit_linear" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a linear regression to the data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>pandas.DataFrame</em>) &#8211; the data to analyse</li>
<li><strong>formula</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the formula for the linear regression</li>
<li><strong>result_col</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the column that will contain the results</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the results from the linear regression</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)">list</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.fit_logistic">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">fit_logistic</code><span class="sig-paren">(</span><em>data</em>, <em>formula</em>, <em>result_col</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#fit_logistic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.fit_logistic" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a logistic regression to the data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>pandas.DataFrame</em>) &#8211; the data to analyse</li>
<li><strong>formula</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the formula for the logistic regression</li>
<li><strong>result_col</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the column that will contain the results</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the results from the logistic regression</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)">list</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.fit_mixedlm">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">fit_mixedlm</code><span class="sig-paren">(</span><em>data</em>, <em>formula</em>, <em>use_ml</em>, <em>groups</em>, <em>result_col</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#fit_mixedlm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.fit_mixedlm" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a linear mixed effects model to the data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>pandas.DataFrame</em>) &#8211; the data to analyse</li>
<li><strong>formula</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the formula for the linear mixed effects model</li>
<li><strong>use_ml</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/functions.html#bool" title="(in Python v3.4)"><em>bool</em></a>) &#8211; whether to use ML instead of REML</li>
<li><strong>groups</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the column containing the groups</li>
<li><strong>result_col</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the column that will contain the results</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the results from the linear mixed effects model</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)">list</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.get_formula">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">get_formula</code><span class="sig-paren">(</span><em>phenotype</em>, <em>covars</em>, <em>interaction</em>, <em>gender_c</em>, <em>categorical</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#get_formula"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.get_formula" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates the linear/logistic regression formula (for statsmodel).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>phenotype</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the phenotype column</li>
<li><strong>covars</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the list of co variable columns</li>
<li><strong>interaction</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the interaction column</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the formula for the statistical analysis</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)">str</a></p>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The phenotype column needs to be specified. The list of co variables
might be empty (if no co variables are necessary). The interaction
column can be set to <code class="docutils literal"><span class="pre">None</span></code> if there is no interaction.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The gender column should be categorical (hence, the formula requires
the gender to be included into <code class="docutils literal"><span class="pre">C()</span></code>, <em>e.g.</em> <code class="docutils literal"><span class="pre">C(Gender)</span></code>).</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.main">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">main</code><span class="sig-paren">(</span><em>args=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.main" title="Permalink to this definition">¶</a></dt>
<dd><p>The main function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the arguments to be parsed (if
<a class="reference internal" href="#genipe.tools.imputed_stats.main" title="genipe.tools.imputed_stats.main"><code class="xref py py-func docutils literal"><span class="pre">main()</span></code></a> is called by another
modulel)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.parse_args">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">parse_args</code><span class="sig-paren">(</span><em>parser</em>, <em>args=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#parse_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.parse_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses the command line options and arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>parser</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.ArgumentParser" title="(in Python v3.4)"><em>argparse.ArgumentParser</em></a>) &#8211; the argument parser</li>
<li><strong>args</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the list of arguments (if not taken from <code class="docutils literal"><span class="pre">sys.argv</span></code>)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the list of options and arguments</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)">argparse.Namespace</a></p>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The only check that is done here is by the parser itself. Values are
verified later by the <a class="reference internal" href="#genipe.tools.imputed_stats.check_args" title="genipe.tools.imputed_stats.check_args"><code class="xref py py-func docutils literal"><span class="pre">check_args()</span></code></a> function.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.process_impute2_site">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">process_impute2_site</code><span class="sig-paren">(</span><em>site_info</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#process_impute2_site"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.process_impute2_site" title="Permalink to this definition">¶</a></dt>
<dd><p>Process an IMPUTE2 site (a line in an IMPUTE2 file).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>site_info</strong> (<a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)"><em>list</em></a>) &#8211; the impute2 line (split by space)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the results of the analysis</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="http://jinja.pocoo.org/docs/templates/#list" title="(in Jinja2 v2.8-dev)">list</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.read_phenotype">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">read_phenotype</code><span class="sig-paren">(</span><em>i_filename</em>, <em>opts</em>, <em>check_duplicated=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#read_phenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.read_phenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the phenotype file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>i_filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the input file</li>
<li><strong>opts</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the options</li>
<li><strong>check_duplicated</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/functions.html#bool" title="(in Python v3.4)"><em>bool</em></a>) &#8211; whether or not to check for duplicated samples</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the phenotypes</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pandas.DataFrame</p>
</td>
</tr>
</tbody>
</table>
<p>This file is expected to be a tab separated file of phenotypes and
covariates. The columns to use will be determined by the
<code class="docutils literal"><span class="pre">--sample-column</span></code> and the <code class="docutils literal"><span class="pre">--covar</span></code> options.</p>
<p>For analysis including the X chromosome, the gender is automatically
added as a covariate. The results are not shown to the user unless asked
for.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.read_samples">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">read_samples</code><span class="sig-paren">(</span><em>i_filename</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#read_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.read_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the sample file (produced by SHAPEIT).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i_filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the input file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the list of samples</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pandas.DataFrame</td>
</tr>
</tbody>
</table>
<p>This file contains the list of samples that are contained in the
<code class="docutils literal"><span class="pre">impute2</span></code> file (with same order). The expected format for this file is a
tab separated file with a first row containing the following columns:</p>
<div class="highlight-python"><div class="highlight"><pre>ID_1    ID_2    missing father  mother  sex     plink_pheno
</pre></div>
</div>
<p>The subsequent row will be discarded and should contain:</p>
<div class="highlight-python"><div class="highlight"><pre>0       0       0 D     D       D       B
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>We are mostly interested in the sample IDs corresponding to the
<code class="docutils literal"><span class="pre">ID_2</span></code> column. Their uniqueness is verified by pandas.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.read_sites_to_extract">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">read_sites_to_extract</code><span class="sig-paren">(</span><em>i_filename</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#read_sites_to_extract"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.read_sites_to_extract" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the list of sites to extract.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i_filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; The input filename containing the IDs of the variants
to consider for the analysis.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">A set containing the variants.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#set" title="(in Python v3.4)">set</a></td>
</tr>
</tbody>
</table>
<p>The expected file format is simply a list of variants. Every row should
correspond to a single variant identifier.</p>
<div class="highlight-python"><div class="highlight"><pre>3:60069:t
rs123456:A
3:60322:A
</pre></div>
</div>
<p>Typically, this is used to analyze only variants that passed some QC
threshold. The <a class="reference internal" href="genipe.html#module-genipe" title="genipe"><code class="xref py py-mod docutils literal"><span class="pre">genipe</span></code></a> pipeline generates this file at the
&#8216;merge_impute2&#8217; step.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.samples_with_hetero_calls">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">samples_with_hetero_calls</code><span class="sig-paren">(</span><em>data</em>, <em>hetero_c</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#samples_with_hetero_calls"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.samples_with_hetero_calls" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets male and heterozygous calls.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>pandas.DataFrame</em>) &#8211; the probability matrix</li>
<li><strong>hetero_c</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the heterozygous column</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">samples where call is heterozygous</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pandas.Index</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.skat_parse_impute2">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">skat_parse_impute2</code><span class="sig-paren">(</span><em>impute2_filename</em>, <em>samples</em>, <em>markers_to_extract</em>, <em>phenotypes</em>, <em>remove_gender</em>, <em>out_prefix</em>, <em>args</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#skat_parse_impute2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.skat_parse_impute2" title="Permalink to this definition">¶</a></dt>
<dd><p>Read the impute2 file and run the SKAT analysis.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>impute2_filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the input file</li>
<li><strong>samples</strong> (<em>pandas.DataFrame</em>) &#8211; the samples</li>
<li><strong>markers_to_extract</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#set" title="(in Python v3.4)"><em>set</em></a>) &#8211; the set of markers to analyse</li>
<li><strong>phenotypes</strong> (<em>pandas.DataFrame</em>) &#8211; the phenotypes</li>
<li><strong>remove_gender</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/functions.html#bool" title="(in Python v3.4)"><em>bool</em></a>) &#8211; whether or not to remove the gender column</li>
<li><strong>out_prefix</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the output prefix</li>
<li><strong>args</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/argparse.html#argparse.Namespace" title="(in Python v3.4)"><em>argparse.Namespace</em></a>) &#8211; the options</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>This function does most of the &#8220;dispatching&#8221; to run SKAT. It writes the
input files to the disk, runs the generated R scripts to do the actual
analysis and then writes the results to disk.</p>
</dd></dl>

<dl class="function">
<dt id="genipe.tools.imputed_stats.skat_read_snp_set">
<code class="descclassname">genipe.tools.imputed_stats.</code><code class="descname">skat_read_snp_set</code><span class="sig-paren">(</span><em>i_filename</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genipe/tools/imputed_stats.html#skat_read_snp_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#genipe.tools.imputed_stats.skat_read_snp_set" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the SKAT SNP set file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>i_filename</strong> (<a class="reference external" href="https://docs.python.org/3.4/library/stdtypes.html#str" title="(in Python v3.4)"><em>str</em></a>) &#8211; the name of the input file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the SNP set for the SKAT analysis</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pandas.DataFrame</td>
</tr>
</tbody>
</table>
<p>This file has to be supplied by the user. The recognized columns are:
<code class="docutils literal"><span class="pre">variant</span></code>, <code class="docutils literal"><span class="pre">snp_set</span></code> and <code class="docutils literal"><span class="pre">weight</span></code>. The <code class="docutils literal"><span class="pre">weight</span></code> column is optional
and can be used to specify a custom weighting scheme for SKAT. If nothing
is specified, the default Beta weights are used.</p>
<p>The file has to be tab delimited.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="genipe.tests.html" title="genipe.tests package"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">genipe 1.2.1 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../implementation.html" >Implementation</a> &raquo;</li>
          <li class="nav-item nav-item-2"><a href="modules.html" >genipe</a> &raquo;</li>
          <li class="nav-item nav-item-3"><a href="genipe.html" >genipe package</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Beaulieu-Saucier Pharmacogenomics Centre.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>